// shadcn-vue@2.7.4/dist/index.js
#!/usr/bin/env node
import { a as BASES, c as DEPRECATED_COMPONENTS, d as ICON_LIBRARIES$1, i as replaceIconLibraryInDependencies, m as STYLES, o as BASE_COLORS, p as SHADCN_VUE_URL, r as ICON_LIBRARIES, s as BUILTIN_REGISTRIES, u as FONTS } from "./transform-menu-ZTp9ez-j.js";
import { _ as registryItemSchema, a as rawConfigSchema, x as registrySchema } from "./schema-CjAXtlQ7.js";
import { $ as findExistingEnvFile, A as resolveRegistryItems, B as _getQuoteChar, C as getRegistryItems, D as getRegistryVisualStyle, E as getRegistryStyles, F as resolveInitUrl, G as clearRegistryContext, H as handleError, I as resolveRegistryBaseConfig, J as updateFiles, K as isUniversalRegistryItem, L as ensureRegistriesInConfig, M as DEFAULT_PRESETS, N as promptForBase, O as getRegistryVisualStyles, P as promptForPreset, Q as logger, R as resolveRegistryTree, S as getRegistryIcons, T as getRegistryPresets, Tt as highlighter, U as buildUrlAndHeadersForRegistryItem, V as updateTailwindConfig, W as validateRegistryConfigForItems, X as transformRtl, Y as transform$1, Z as spinner, _ as getRegistryBases, a as getProjectTailwindVersionFromConfig, at as DEFAULT_TAILWIND_CONFIG, b as getRegistryIconLibraries, c as getPackageInfo, ct as createConfig, d as getRegistriesConfig, dt as getConfig, et as getNewEnvKeys, f as getRegistriesIndex, ft as getWorkspaceConfig, g as getRegistryBaseColors, h as getRegistryBaseColor, i as getProjectInfo, it as DEFAULT_COMPONENTS, j as resolveTree, k as getShadcnRegistryIndex, l as fetchTree, lt as findCommonRoot, m as getRegistryBase, n as getProjectComponents, nt as composeStyleId, o as getPackageManager, ot as DEFAULT_TAILWIND_CSS, p as getRegistry, pt as resolveConfigPaths, q as isUrl, r as getProjectConfig, rt as configWithDefaults, s as updateDependencies, st as DEFAULT_UTILS, t as searchRegistries, tt as mergeEnvContent, u as getItemTargetPath, ut as findPackageRoot, v as getRegistryFont, w as getRegistryPreset, x as getRegistryIconLibrary, y as getRegistryFonts, z as _createSourceFile } from "./registry-BZ439r4u.js";
import { p as decodePreset, v as isPresetCode } from "./preset-ByCdjgEv.js";
import { t as server } from "./mcp-D65MaH3A.js";
import { Command } from "commander";
import path, { isAbsolute, join, normalize, resolve, sep } from "pathe";
import prompts from "prompts";
import z$1, { z } from "zod";
import { existsSync, promises } from "fs";
import deepmerge from "deepmerge";
import fsExtra from "fs-extra";
import { glob } from "tinyglobby";
import * as fs$1 from "fs/promises";
import { tmpdir } from "os";
import { Project, ScriptKind, SyntaxKind } from "ts-morph";
import { transform } from "vue-metamorph";
import consola from "consola";
import { randomBytes } from "crypto";
import * as path$2 from "path";
import { addDevDependency, installDependencies } from "nypm";
import postcss from "postcss";
import AtRule from "postcss/lib/at-rule";
import { downloadTemplate } from "giget";
import { x } from "tinyexec";
import { diffLines } from "diff";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";

//#region src/utils/errors.ts
const MISSING_DIR_OR_EMPTY_PROJECT = "1";
const MISSING_CONFIG = "3";
const TAILWIND_NOT_CONFIGURED = "5";
const IMPORT_ALIAS_MISSING = "6";
const UNSUPPORTED_FRAMEWORK = "7";
const BUILD_MISSING_REGISTRY_FILE = "13";

//#endregion
//#region src/preflights/preflight-init.ts
async function preFlightInit(options) {
	const errors = {};
	if (!fsExtra.existsSync(options.cwd) || !fsExtra.existsSync(path.resolve(options.cwd, "package.json"))) {
		errors[MISSING_DIR_OR_EMPTY_PROJECT] = true;
		return {
			errors,
			projectInfo: null
		};
	}
	const projectSpinner = spinner(`Preflight checks.`, { silent: options.silent }).start();
	if (fsExtra.existsSync(path.resolve(options.cwd, "components.json")) && !options.force) {
		projectSpinner?.fail();
		logger.break();
		logger.error(`A ${highlighter.info("components.json")} file already exists at ${highlighter.info(options.cwd)}.\nTo start over, remove the ${highlighter.info("components.json")} file and run ${highlighter.info("init")} again.`);
		logger.break();
		process.exit(1);
	}
	projectSpinner?.succeed();
	const frameworkSpinner = spinner(`Verifying framework.`, { silent: options.silent }).start();
	const projectInfo = await getProjectInfo(options.cwd);
	if (!projectInfo || projectInfo?.framework.name === "manual") {
		errors[UNSUPPORTED_FRAMEWORK] = true;
		frameworkSpinner?.fail();
		logger.break();
		if (projectInfo?.framework.links.installation) logger.error(`We could not detect a supported framework at ${highlighter.info(options.cwd)}.\nVisit ${highlighter.info(projectInfo?.framework.links.installation)} to manually configure your project.\nOnce configured, you can use the cli to add components.`);
		logger.break();
		process.exit(1);
	}
	frameworkSpinner?.succeed(`Verifying framework. Found ${highlighter.info(projectInfo.framework.label)}.`);
	let tailwindSpinnerMessage = "Validating Tailwind CSS.";
	if (projectInfo.tailwindVersion === "v4") tailwindSpinnerMessage = `Validating Tailwind CSS config. Found ${highlighter.info("v4")}.`;
	const tailwindSpinner = spinner(tailwindSpinnerMessage, { silent: options.silent }).start();
	if (projectInfo.tailwindVersion === "v3" && (!projectInfo?.tailwindConfigFile || !projectInfo?.tailwindCssFile)) {
		errors[TAILWIND_NOT_CONFIGURED] = true;
		tailwindSpinner?.fail();
	} else if (projectInfo.tailwindVersion === "v4" && !projectInfo?.tailwindCssFile) {
		errors[TAILWIND_NOT_CONFIGURED] = true;
		tailwindSpinner?.fail();
	} else if (!projectInfo.tailwindVersion) {
		errors[TAILWIND_NOT_CONFIGURED] = true;
		tailwindSpinner?.fail();
	} else tailwindSpinner?.succeed();
	const tsConfigSpinner = spinner(`Validating import alias.`, { silent: options.silent }).start();
	if (!projectInfo?.aliasPrefix) {
		errors[IMPORT_ALIAS_MISSING] = true;
		tsConfigSpinner?.fail();
	} else tsConfigSpinner?.succeed();
	if (Object.keys(errors).length > 0) {
		if (errors[TAILWIND_NOT_CONFIGURED]) {
			logger.break();
			logger.error(`No Tailwind CSS configuration found at ${highlighter.info(options.cwd)}.`);
			logger.error(`It is likely you do not have Tailwind CSS installed or have an invalid configuration.`);
			logger.error(`Install Tailwind CSS then try again.`);
			if (projectInfo?.framework.links.tailwind) logger.error(`Visit ${highlighter.info(projectInfo?.framework.links.tailwind)} to get started.`);
		}
		if (errors[IMPORT_ALIAS_MISSING]) {
			logger.break();
			logger.error(`No import alias found in your tsconfig.json file.`);
			if (projectInfo?.framework.links.installation) logger.error(`Visit ${highlighter.info(projectInfo?.framework.links.installation)} to learn how to set an import alias.`);
		}
		logger.break();
		process.exit(1);
	}
	return {
		errors,
		projectInfo
	};
}

//#endregion
//#region src/utils/fonts.ts
/**
* Get font configuration by name.
*/
function getFont(name) {
	return FONTS.find((font) => font.name === name);
}
/**
* Get font CSS import statement.
*/
function getFontImport(name) {
	return getFont(name)?.import ?? "";
}
/**
* Get CSS variable name for the font.
*/
function getFontVariable(name) {
	return getFont(name)?.variable ?? "--font-sans";
}

//#endregion
//#region src/utils/is-safe-target.ts
function isSafeTarget(targetPath, cwd) {
	if (targetPath.includes("\0")) return false;
	let decodedPath;
	try {
		decodedPath = targetPath;
		let prevPath = "";
		while (decodedPath !== prevPath && decodedPath.includes("%")) {
			prevPath = decodedPath;
			decodedPath = decodeURIComponent(decodedPath);
		}
	} catch {
		return false;
	}
	const normalizedTarget = normalize(decodedPath);
	const normalizedRoot = normalize(cwd);
	const hasPathTraversal = (path$3) => {
		return path$3.replace(/\[\.\.\..*?\]/g, "").includes("..");
	};
	if (hasPathTraversal(normalizedTarget) || hasPathTraversal(decodedPath) || hasPathTraversal(targetPath)) return false;
	const cleanPath = (path$3) => path$3.replace(/\[\.\.\..*?\]/g, "");
	const cleanTarget = cleanPath(targetPath);
	const cleanDecoded = cleanPath(decodedPath);
	if ([
		/\.\.[/\\]/,
		/[/\\]\.\./,
		/\.\./,
		/\.\.%/,
		/\0/,
		/[\x01-\x1F]/
	].some((pattern) => pattern.test(cleanTarget) || pattern.test(cleanDecoded))) return false;
	if ((targetPath.includes("~") || decodedPath.includes("~")) && (targetPath.includes("../") || decodedPath.includes("../"))) return false;
	if (/^[a-z]:[/\\]/i.test(decodedPath)) {
		if (process.platform === "win32") return decodedPath.toLowerCase().startsWith(cwd.toLowerCase());
		return false;
	}
	const absoluteTarget = isAbsolute(normalizedTarget) ? normalizedTarget : resolve(normalizedRoot, normalizedTarget);
	const safeRoot = normalizedRoot.endsWith(sep) ? normalizedRoot : normalizedRoot + sep;
	return absoluteTarget === normalizedRoot || absoluteTarget.startsWith(safeRoot);
}

//#endregion
//#region src/utils/updaters/update-css.ts
async function updateCss(css, config, options) {
	if (!config.resolvedPaths.tailwindCss || !css || Object.keys(css).length === 0) return;
	options = {
		silent: false,
		...options
	};
	const cssFilepath = config.resolvedPaths.tailwindCss;
	const cssFilepathRelative = path.relative(config.resolvedPaths.cwd, cssFilepath);
	const cssSpinner = spinner(`Updating ${highlighter.info(cssFilepathRelative)}`, { silent: options.silent }).start();
	let output = await transformCss(await promises.readFile(cssFilepath, "utf8"), css);
	await promises.writeFile(cssFilepath, output, "utf8");
	cssSpinner.succeed();
}
async function transformCss(input, css) {
	const result = await postcss([updateCssPlugin(css)]).process(input, { from: void 0 });
	let output = result.css;
	const root = result.root;
	if (root.nodes && root.nodes.length > 0) {
		const lastNode = root.nodes[root.nodes.length - 1];
		if (lastNode.type === "atrule" && !lastNode.nodes && !output.trimEnd().endsWith(";")) output = `${output.trimEnd()};`;
	}
	output = output.replace(/\/\* ---break--- \*\//g, "");
	output = output.replace(/(\n\s*\n)+/g, "\n\n");
	output = output.trimEnd();
	return output;
}
function updateCssPlugin(css) {
	return {
		postcssPlugin: "update-css",
		Once(root) {
			for (const [selector, properties] of Object.entries(css)) if (selector.startsWith("@")) {
				const atRuleMatch = selector.match(/@([a-z-]+)\s*(.*)/i);
				if (!atRuleMatch) continue;
				const [, name, params] = atRuleMatch;
				if (name === "import") {
					const normalizeImportParams = (p) => {
						const trimmed = p.trim();
						if (trimmed.startsWith("\"") && trimmed.endsWith("\"")) return trimmed.slice(1, -1);
						if (trimmed.startsWith("'") && trimmed.endsWith("'")) return trimmed.slice(1, -1);
						return trimmed;
					};
					if (!root.nodes?.find((node) => node.type === "atrule" && node.name === "import" && normalizeImportParams(node.params) === normalizeImportParams(params))) {
						const importRule = postcss.atRule({
							name: "import",
							params,
							raws: { semicolon: true }
						});
						const importNodes = root.nodes?.filter((node) => node.type === "atrule" && node.name === "import");
						if (importNodes && importNodes.length > 0) {
							const lastImport = importNodes[importNodes.length - 1];
							importRule.raws.before = "\n";
							root.insertAfter(lastImport, importRule);
						} else {
							if (!root.nodes || root.nodes.length === 0) importRule.raws.before = "";
							else importRule.raws.before = "";
							root.prepend(importRule);
						}
					}
				} else if (name === "plugin") {
					let quotedParams = params;
					if (params && !params.startsWith("\"") && !params.startsWith("'")) quotedParams = `"${params}"`;
					const normalizeParams = (p) => {
						if (p.startsWith("\"") && p.endsWith("\"")) return p.slice(1, -1);
						if (p.startsWith("'") && p.endsWith("'")) return p.slice(1, -1);
						return p;
					};
					if (!root.nodes?.find((node) => {
						if (node.type !== "atrule" || node.name !== "plugin") return false;
						return normalizeParams(node.params) === normalizeParams(params);
					})) {
						const pluginRule = postcss.atRule({
							name: "plugin",
							params: quotedParams,
							raws: {
								semicolon: true,
								before: "\n"
							}
						});
						const importNodes = root.nodes?.filter((node) => node.type === "atrule" && node.name === "import");
						const pluginNodes = root.nodes?.filter((node) => node.type === "atrule" && node.name === "plugin");
						if (pluginNodes && pluginNodes.length > 0) {
							const lastPlugin = pluginNodes[pluginNodes.length - 1];
							root.insertAfter(lastPlugin, pluginRule);
						} else if (importNodes && importNodes.length > 0) {
							const lastImport = importNodes[importNodes.length - 1];
							root.insertAfter(lastImport, pluginRule);
							root.insertBefore(pluginRule, postcss.comment({ text: "---break---" }));
							root.insertAfter(pluginRule, postcss.comment({ text: "---break---" }));
						} else {
							root.prepend(pluginRule);
							root.insertBefore(pluginRule, postcss.comment({ text: "---break---" }));
							root.insertAfter(pluginRule, postcss.comment({ text: "---break---" }));
						}
					}
				} else if (typeof properties === "object" && Object.keys(properties).length === 0) {
					if (!root.nodes?.find((node) => node.type === "atrule" && node.name === name && node.params === params)) {
						const newAtRule = postcss.atRule({
							name,
							params,
							raws: { semicolon: true }
						});
						root.append(newAtRule);
						root.insertBefore(newAtRule, postcss.comment({ text: "---break---" }));
					}
				} else if (name === "keyframes") {
					let themeInline = root.nodes?.find((node) => node.type === "atrule" && node.name === "theme" && node.params === "inline");
					if (!themeInline) {
						themeInline = postcss.atRule({
							name: "theme",
							params: "inline",
							raws: {
								semicolon: true,
								between: " ",
								before: "\n"
							}
						});
						root.append(themeInline);
						root.insertBefore(themeInline, postcss.comment({ text: "---break---" }));
					}
					const existingKeyframesRule = themeInline.nodes?.find((node) => node.type === "atrule" && node.name === "keyframes" && node.params === params);
					let keyframesRule;
					if (existingKeyframesRule) {
						keyframesRule = postcss.atRule({
							name: "keyframes",
							params,
							raws: {
								semicolon: true,
								between: " ",
								before: "\n  "
							}
						});
						existingKeyframesRule.replaceWith(keyframesRule);
					} else {
						keyframesRule = postcss.atRule({
							name: "keyframes",
							params,
							raws: {
								semicolon: true,
								between: " ",
								before: "\n  "
							}
						});
						themeInline.append(keyframesRule);
					}
					if (typeof properties === "object") for (const [step, stepProps] of Object.entries(properties)) processRule(keyframesRule, step, stepProps);
				} else if (name === "utility") {
					const utilityAtRule = root.nodes?.find((node) => node.type === "atrule" && node.name === name && node.params === params);
					if (!utilityAtRule) {
						const atRule = postcss.atRule({
							name,
							params,
							raws: {
								semicolon: true,
								between: " ",
								before: "\n"
							}
						});
						root.append(atRule);
						root.insertBefore(atRule, postcss.comment({ text: "---break---" }));
						if (typeof properties === "object") {
							for (const [prop, value] of Object.entries(properties)) if (typeof value === "string") {
								const decl = postcss.decl({
									prop,
									value,
									raws: {
										semicolon: true,
										before: "\n    "
									}
								});
								atRule.append(decl);
							} else if (typeof value === "object") processRule(atRule, prop, value);
						}
					} else if (typeof properties === "object") {
						for (const [prop, value] of Object.entries(properties)) if (typeof value === "string") {
							const existingDecl = utilityAtRule.nodes?.find((node) => node.type === "decl" && node.prop === prop);
							const decl = postcss.decl({
								prop,
								value,
								raws: {
									semicolon: true,
									before: "\n    "
								}
							});
							existingDecl ? existingDecl.replaceWith(decl) : utilityAtRule.append(decl);
						} else if (typeof value === "object") processRule(utilityAtRule, prop, value);
					}
				} else if (name === "property") processRule(root, selector, properties);
				else processAtRule(root, name, params, properties);
			} else processRule(root, selector, properties);
		}
	};
}
function processAtRule(root, name, params, properties) {
	let atRule = root.nodes?.find((node) => node.type === "atrule" && node.name === name && node.params === params);
	if (!atRule) {
		atRule = postcss.atRule({
			name,
			params,
			raws: {
				semicolon: true,
				between: " ",
				before: "\n"
			}
		});
		root.append(atRule);
		root.insertBefore(atRule, postcss.comment({ text: "---break---" }));
	}
	if (typeof properties === "object") for (const [childSelector, childProps] of Object.entries(properties)) if (childSelector.startsWith("@")) {
		const nestedMatch = childSelector.match(/@([a-z-]+)\s*(.*)/i);
		if (nestedMatch) {
			const [, nestedName, nestedParams] = nestedMatch;
			processAtRule(atRule, nestedName, nestedParams, childProps);
		}
	} else processRule(atRule, childSelector, childProps);
	else if (typeof properties === "string") try {
		const tempRule = postcss.parse(`.temp{${properties}}`).first;
		if (tempRule && tempRule.nodes) {
			const rule = postcss.rule({
				selector: "temp",
				raws: {
					semicolon: true,
					between: " ",
					before: "\n  "
				}
			});
			tempRule.nodes.forEach((node) => {
				if (node.type === "decl") {
					const clone = node.clone();
					clone.raws.before = "\n    ";
					rule.append(clone);
				}
			});
			if (rule.nodes?.length) atRule.append(rule);
		}
	} catch (error) {
		console.error("Error parsing at-rule content:", properties, error);
		throw error;
	}
}
function processRule(parent, selector, properties) {
	let rule = parent.nodes?.find((node) => node.type === "rule" && node.selector === selector);
	if (!rule) {
		rule = postcss.rule({
			selector,
			raws: {
				semicolon: true,
				between: " ",
				before: "\n  "
			}
		});
		parent.append(rule);
	}
	if (typeof properties === "object") {
		for (const [prop, value] of Object.entries(properties)) if (prop.startsWith("@") && typeof value === "object" && value !== null && Object.keys(value).length === 0) {
			const atRuleMatch = prop.match(/@([a-z-]+)\s*(.*)/i);
			if (atRuleMatch) {
				const [, atRuleName, atRuleParams] = atRuleMatch;
				if (!rule.nodes?.find((node) => node.type === "atrule" && node.name === atRuleName && node.params === atRuleParams)) {
					const atRule = postcss.atRule({
						name: atRuleName,
						params: atRuleParams,
						raws: {
							semicolon: true,
							before: "\n    "
						}
					});
					rule.append(atRule);
				}
			}
		} else if (typeof value === "string") {
			const decl = postcss.decl({
				prop,
				value,
				raws: {
					semicolon: true,
					before: "\n    "
				}
			});
			const existingDecl = rule.nodes?.find((node) => node.type === "decl" && node.prop === prop);
			existingDecl ? existingDecl.replaceWith(decl) : rule.append(decl);
		} else if (typeof value === "object") processRule(parent, prop.startsWith("&") ? selector.replace(/^([^:]+)/, `$1${prop.substring(1)}`) : prop, value);
	} else if (typeof properties === "string") try {
		const tempRule = postcss.parse(`.temp{${properties}}`).first;
		if (tempRule && tempRule.nodes) tempRule.nodes.forEach((node) => {
			if (node.type === "decl") {
				const clone = node.clone();
				clone.raws.before = "\n    ";
				rule?.append(clone);
			}
		});
	} catch (error) {
		console.error("Error parsing rule content:", selector, properties, error);
		throw error;
	}
}

//#endregion
//#region src/utils/updaters/update-css-vars.ts
async function updateCssVars(cssVars, config, options) {
	if (!config.resolvedPaths.tailwindCss || !Object.keys(cssVars ?? {}).length) return;
	options = {
		cleanupDefaultNextStyles: false,
		silent: false,
		tailwindVersion: "v3",
		overwriteCssVars: false,
		initIndex: true,
		...options
	};
	const cssFilepath = config.resolvedPaths.tailwindCss;
	const cssFilepathRelative = path.relative(config.resolvedPaths.cwd, cssFilepath);
	const cssVarsSpinner = spinner(`Updating CSS variables in ${highlighter.info(cssFilepathRelative)}`, { silent: options.silent }).start();
	const output = await transformCssVars(await promises.readFile(cssFilepath, "utf8"), cssVars ?? {}, config, {
		cleanupDefaultNextStyles: options.cleanupDefaultNextStyles,
		tailwindVersion: options.tailwindVersion,
		tailwindConfig: options.tailwindConfig,
		overwriteCssVars: options.overwriteCssVars,
		initIndex: options.initIndex,
		fontImports: options.fontImports
	});
	await promises.writeFile(cssFilepath, output, "utf8");
	cssVarsSpinner.succeed();
}
async function transformCssVars(input, cssVars, config, options = {
	cleanupDefaultNextStyles: false,
	tailwindVersion: "v3",
	tailwindConfig: void 0,
	overwriteCssVars: false,
	initIndex: true
}) {
	options = {
		cleanupDefaultNextStyles: false,
		tailwindVersion: "v3",
		tailwindConfig: void 0,
		overwriteCssVars: false,
		initIndex: true,
		...options
	};
	let plugins = [updateCssVarsPlugin(cssVars)];
	const fontImports = options.fontImports ?? [];
	if (fontImports.length > 0) plugins.unshift(addFontImportPlugin({ fontImports }));
	if (options.cleanupDefaultNextStyles) plugins.push(cleanupDefaultNextStylesPlugin());
	if (options.tailwindVersion === "v4") {
		plugins = [];
		if (fontImports.length > 0) plugins.push(addFontImportPlugin({ fontImports }));
		if (config.resolvedPaths?.cwd) {
			const packageInfo = getPackageInfo(config.resolvedPaths.cwd);
			if (!packageInfo?.dependencies?.["tailwindcss-animate"] && !packageInfo?.devDependencies?.["tailwindcss-animate"] && options.initIndex) plugins.push(addCustomImport({ params: "tw-animate-css" }));
		}
		plugins.push(addCustomVariant({ params: "dark (&:is(.dark *))" }));
		if (options.cleanupDefaultNextStyles) plugins.push(cleanupDefaultNextStylesPlugin());
		plugins.push(updateCssVarsPluginV4(cssVars, { overwriteCssVars: options.overwriteCssVars }));
		plugins.push(updateThemePlugin(cssVars));
		if (options.tailwindConfig) {
			plugins.push(updateTailwindConfigPlugin(options.tailwindConfig));
			plugins.push(updateTailwindConfigAnimationPlugin(options.tailwindConfig));
			plugins.push(updateTailwindConfigKeyframesPlugin(options.tailwindConfig));
		}
	}
	if (config.tailwind.cssVariables && options.initIndex) plugins.push(updateBaseLayerPlugin({ tailwindVersion: options.tailwindVersion }));
	let output = (await postcss(plugins).process(input, { from: void 0 })).css;
	output = output.replace(/\/\* ---break--- \*\//g, "");
	if (options.tailwindVersion === "v4") output = output.replace(/(\n\s*\n)+/g, "\n\n");
	return output;
}
function updateBaseLayerPlugin({ tailwindVersion }) {
	return {
		postcssPlugin: "update-base-layer",
		Once(root) {
			const requiredRules = [{
				selector: "*",
				apply: tailwindVersion === "v4" ? "border-border outline-ring/50" : "border-border"
			}, {
				selector: "body",
				apply: "bg-background text-foreground"
			}];
			let baseLayer = root.nodes.find((node) => node.type === "atrule" && node.name === "layer" && node.params === "base" && requiredRules.every(({ selector, apply: apply$1 }) => node.nodes?.some((rule) => rule.type === "rule" && rule.selector === selector && rule.nodes.some((applyRule) => applyRule.type === "atrule" && applyRule.name === "apply" && applyRule.params === apply$1))));
			if (!baseLayer) {
				baseLayer = postcss.atRule({
					name: "layer",
					params: "base",
					raws: {
						semicolon: true,
						between: " ",
						before: "\n"
					}
				});
				root.append(baseLayer);
				root.insertBefore(baseLayer, postcss.comment({ text: "---break---" }));
			}
			requiredRules.forEach(({ selector, apply: apply$1 }) => {
				if (!baseLayer?.nodes?.find((node) => node.type === "rule" && node.selector === selector)) baseLayer?.append(postcss.rule({
					selector,
					nodes: [postcss.atRule({
						name: "apply",
						params: apply$1,
						raws: {
							semicolon: true,
							before: "\n    "
						}
					})],
					raws: {
						semicolon: true,
						between: " ",
						before: "\n  "
					}
				}));
			});
		}
	};
}
function updateCssVarsPlugin(cssVars) {
	return {
		postcssPlugin: "update-css-vars",
		Once(root) {
			let baseLayer = root.nodes.find((node) => node.type === "atrule" && node.name === "layer" && node.params === "base");
			if (!(baseLayer instanceof AtRule)) {
				baseLayer = postcss.atRule({
					name: "layer",
					params: "base",
					nodes: [],
					raws: {
						semicolon: true,
						before: "\n",
						between: " "
					}
				});
				root.append(baseLayer);
				root.insertBefore(baseLayer, postcss.comment({ text: "---break---" }));
			}
			if (baseLayer !== void 0) Object.entries(cssVars).forEach(([key, vars]) => {
				const selector = key === "light" ? ":root" : `.${key}`;
				addOrUpdateVars(baseLayer, selector, vars);
			});
		}
	};
}
function removeConflictVars(root) {
	const rootRule = root.nodes.find((node) => node.type === "rule" && node.selector === ":root");
	if (rootRule) {
		const propsToRemove = ["--background", "--foreground"];
		rootRule.nodes.filter((node) => node.type === "decl" && propsToRemove.includes(node.prop)).forEach((node) => node.remove());
		if (rootRule.nodes.length === 0) rootRule.remove();
	}
}
function cleanupDefaultNextStylesPlugin() {
	return {
		postcssPlugin: "cleanup-default-next-styles",
		Once(root) {
			const bodyRule = root.nodes.find((node) => node.type === "rule" && node.selector === "body");
			if (bodyRule) {
				bodyRule.nodes.find((node) => node.type === "decl" && node.prop === "color" && ["rgb(var(--foreground-rgb))", "var(--foreground)"].includes(node.value))?.remove();
				bodyRule.nodes.find((node) => {
					return node.type === "decl" && node.prop === "background" && (node.value.startsWith("linear-gradient") || node.value === "var(--background)");
				})?.remove();
				bodyRule.nodes.find((node) => node.type === "decl" && node.prop === "font-family" && node.value === "Arial, Helvetica, sans-serif")?.remove();
				if (bodyRule.nodes.length === 0) bodyRule.remove();
			}
			removeConflictVars(root);
			const darkRootRule = root.nodes.find((node) => node.type === "atrule" && node.params === "(prefers-color-scheme: dark)");
			if (darkRootRule) {
				removeConflictVars(darkRootRule);
				if (darkRootRule.nodes.length === 0) darkRootRule.remove();
			}
		}
	};
}
function addOrUpdateVars(baseLayer, selector, vars) {
	let ruleNode = baseLayer.nodes?.find((node) => node.type === "rule" && node.selector === selector);
	if (!ruleNode) {
		if (Object.keys(vars).length > 0) {
			ruleNode = postcss.rule({
				selector,
				raws: {
					between: " ",
					before: "\n  "
				}
			});
			baseLayer.append(ruleNode);
		}
	}
	Object.entries(vars).forEach(([key, value]) => {
		const prop = `--${key.replace(/^--/, "")}`;
		const newDecl = postcss.decl({
			prop,
			value,
			raws: { semicolon: true }
		});
		const existingDecl = ruleNode?.nodes.find((node) => node.type === "decl" && node.prop === prop);
		existingDecl ? existingDecl.replaceWith(newDecl) : ruleNode?.append(newDecl);
	});
}
function updateCssVarsPluginV4(cssVars, options) {
	return {
		postcssPlugin: "update-css-vars-v4",
		Once(root) {
			Object.entries(cssVars).forEach(([key, vars]) => {
				let selector = key === "light" ? ":root" : `.${key}`;
				if (key === "theme") {
					selector = "@theme";
					const themeNode = upsertThemeNode(root);
					Object.entries(vars).forEach(([key$1, value]) => {
						const prop = `--${key$1.replace(/^--/, "")}`;
						const newDecl = postcss.decl({
							prop,
							value,
							raws: { semicolon: true }
						});
						const existingDecl = themeNode?.nodes?.find((node) => node.type === "decl" && node.prop === prop);
						if (options.overwriteCssVars) if (existingDecl) existingDecl.replaceWith(newDecl);
						else themeNode?.append(newDecl);
						else if (!existingDecl) themeNode?.append(newDecl);
					});
					return;
				}
				let ruleNode = root.nodes?.find((node) => node.type === "rule" && node.selector === selector);
				if (!ruleNode && Object.keys(vars).length > 0) {
					ruleNode = postcss.rule({
						selector,
						nodes: [],
						raws: {
							semicolon: true,
							between: " ",
							before: "\n"
						}
					});
					root.append(ruleNode);
					root.insertBefore(ruleNode, postcss.comment({ text: "---break---" }));
				}
				Object.entries(vars).forEach(([key$1, value]) => {
					let prop = `--${key$1.replace(/^--/, "")}`;
					if (prop === "--sidebar-background") prop = "--sidebar";
					if (isLocalHSLValue(value)) value = `hsl(${value})`;
					const newDecl = postcss.decl({
						prop,
						value,
						raws: { semicolon: true }
					});
					const existingDecl = ruleNode?.nodes.find((node) => node.type === "decl" && node.prop === prop);
					if (options.overwriteCssVars) if (existingDecl) existingDecl.replaceWith(newDecl);
					else ruleNode?.append(newDecl);
					else if (!existingDecl) ruleNode?.append(newDecl);
				});
			});
		}
	};
}
function updateThemePlugin(cssVars) {
	return {
		postcssPlugin: "update-theme",
		Once(root) {
			const variables = Array.from(new Set(Object.keys(cssVars).flatMap((key) => Object.keys(cssVars[key] || {}))));
			if (!variables.length) return;
			const themeNode = upsertThemeNode(root);
			const themeVarNodes = themeNode.nodes?.filter((node) => node.type === "decl" && node.prop.startsWith("--"));
			for (const variable of variables) {
				const value = Object.values(cssVars).find((vars) => vars[variable])?.[variable];
				if (!value) continue;
				if (variable === "radius") {
					for (const [key, value$1] of Object.entries({
						sm: "calc(var(--radius) - 4px)",
						md: "calc(var(--radius) - 2px)",
						lg: "var(--radius)",
						xl: "calc(var(--radius) + 4px)"
					})) {
						const cssVarNode$1 = postcss.decl({
							prop: `--radius-${key}`,
							value: value$1,
							raws: { semicolon: true }
						});
						if (themeNode?.nodes?.find((node) => node.type === "decl" && node.prop === cssVarNode$1.prop)) continue;
						themeNode?.append(cssVarNode$1);
					}
					continue;
				}
				let prop = isLocalHSLValue(value) || isColorValue(value) ? `--color-${variable.replace(/^--/, "")}` : `--${variable.replace(/^--/, "")}`;
				if (prop === "--color-sidebar-background") prop = "--color-sidebar";
				let propValue = `var(--${variable})`;
				if (prop === "--color-sidebar") propValue = "var(--sidebar)";
				const cssVarNode = postcss.decl({
					prop,
					value: propValue,
					raws: { semicolon: true }
				});
				if (!themeNode?.nodes?.find((node) => node.type === "decl" && node.prop === cssVarNode.prop)) if (themeVarNodes?.length) themeNode?.insertAfter(themeVarNodes[themeVarNodes.length - 1], cssVarNode);
				else themeNode?.append(cssVarNode);
			}
		}
	};
}
function upsertThemeNode(root) {
	let themeNode = root.nodes.find((node) => node.type === "atrule" && node.name === "theme" && node.params === "inline");
	if (!themeNode) {
		themeNode = postcss.atRule({
			name: "theme",
			params: "inline",
			nodes: [],
			raws: {
				semicolon: true,
				between: " ",
				before: "\n"
			}
		});
		root.append(themeNode);
		root.insertBefore(themeNode, postcss.comment({ text: "---break---" }));
	}
	return themeNode;
}
function addCustomVariant({ params }) {
	return {
		postcssPlugin: "add-custom-variant",
		Once(root) {
			if (!root.nodes.find((node) => node.type === "atrule" && node.name === "custom-variant")) {
				const importNodes = root.nodes.filter((node) => node.type === "atrule" && node.name === "import");
				const variantNode = postcss.atRule({
					name: "custom-variant",
					params,
					raws: {
						semicolon: true,
						before: "\n"
					}
				});
				if (importNodes.length > 0) {
					const lastImport = importNodes[importNodes.length - 1];
					root.insertAfter(lastImport, variantNode);
				} else root.insertAfter(root.nodes[0], variantNode);
				root.insertBefore(variantNode, postcss.comment({ text: "---break---" }));
			}
		}
	};
}
function addCustomImport({ params }) {
	return {
		postcssPlugin: "add-custom-import",
		Once(root) {
			const importNodes = root.nodes.filter((node) => node.type === "atrule" && node.name === "import");
			const customVariantNode = root.nodes.find((node) => node.type === "atrule" && node.name === "custom-variant");
			if (!importNodes.some((node) => node.params.replace(/["']/g, "") === params)) {
				const importNode = postcss.atRule({
					name: "import",
					params: `"${params}"`,
					raws: {
						semicolon: true,
						before: "\n"
					}
				});
				if (importNodes.length > 0) {
					const lastImport = importNodes[importNodes.length - 1];
					root.insertAfter(lastImport, importNode);
				} else if (customVariantNode) {
					root.insertBefore(customVariantNode, importNode);
					root.insertBefore(customVariantNode, postcss.comment({ text: "---break---" }));
				} else {
					root.prepend(importNode);
					root.insertAfter(importNode, postcss.comment({ text: "---break---" }));
				}
			}
		}
	};
}
/**
* Sync Google Fonts `@import` statements at the top of the CSS file with the
* active font set (body + heading). The full set of target URLs is passed in
* each call — any existing `fonts.googleapis.com` imports not in the set are
* removed, any missing ones are added. This is replace-all semantics so that
* swapping to a new preset actually reflects the new fonts instead of
* accumulating dead imports from previous runs.
*/
function addFontImportPlugin({ fontImports }) {
	return {
		postcssPlugin: "add-font-import",
		Once(root) {
			if (!fontImports || fontImports.length === 0) return;
			const targetUrls = [];
			for (const fontImport of fontImports) {
				const match = fontImport.match(/url\(['"]?([^'"]+)['"]?\)/);
				if (match && !targetUrls.includes(match[1])) targetUrls.push(match[1]);
			}
			if (targetUrls.length === 0) return;
			const existingGoogleFontsImports = root.nodes.filter((node) => node.type === "atrule" && node.name === "import" && node.params.includes("fonts.googleapis.com"));
			const keptByUrl = /* @__PURE__ */ new Map();
			for (const node of existingGoogleFontsImports) {
				const url = targetUrls.find((u) => node.params.includes(u));
				if (url && !keptByUrl.has(url)) keptByUrl.set(url, node);
				else node.remove();
			}
			const missingUrls = targetUrls.filter((url) => !keptByUrl.has(url));
			if (missingUrls.length === 0) return;
			const firstKept = existingGoogleFontsImports.find((node) => keptByUrl.get(targetUrls.find((u) => node.params.includes(u))) === node);
			const newNodes = missingUrls.map((url) => postcss.atRule({
				name: "import",
				params: `url('${url}')`,
				raws: {
					semicolon: true,
					before: "\n"
				}
			}));
			if (firstKept) {
				let anchor = firstKept;
				for (const node of newNodes) {
					root.insertAfter(anchor, node);
					anchor = node;
				}
				return;
			}
			for (let i = newNodes.length - 1; i >= 0; i--) root.prepend(newNodes[i]);
			root.insertAfter(newNodes[newNodes.length - 1], postcss.comment({ text: "---break---" }));
		}
	};
}
function updateTailwindConfigPlugin(tailwindConfig) {
	return {
		postcssPlugin: "update-tailwind-config",
		Once(root) {
			if (!tailwindConfig?.plugins) return;
			const quote = getQuoteType(root) === "single" ? "'" : "\"";
			const pluginNodes = root.nodes.filter((node) => node.type === "atrule" && node.name === "plugin");
			const lastPluginNode = pluginNodes[pluginNodes.length - 1] || root.nodes[0];
			for (const plugin of tailwindConfig.plugins) {
				const pluginName = plugin.replace(/^require\(["']|["']\)$/g, "");
				if (pluginNodes.some((node) => {
					return node.params.replace(/["']/g, "") === pluginName;
				})) continue;
				const pluginNode = postcss.atRule({
					name: "plugin",
					params: `${quote}${pluginName}${quote}`,
					raws: {
						semicolon: true,
						before: "\n"
					}
				});
				root.insertAfter(lastPluginNode, pluginNode);
				root.insertBefore(pluginNode, postcss.comment({ text: "---break---" }));
			}
		}
	};
}
function updateTailwindConfigKeyframesPlugin(tailwindConfig) {
	return {
		postcssPlugin: "update-tailwind-config-keyframes",
		Once(root) {
			if (!tailwindConfig?.theme?.extend?.keyframes) return;
			const themeNode = upsertThemeNode(root);
			const existingKeyFrameNodes = themeNode.nodes?.filter((node) => node.type === "atrule" && node.name === "keyframes");
			const keyframeValueSchema = z.record(z.string(), z.record(z.string(), z.string()));
			for (const [keyframeName, keyframeValue] of Object.entries(tailwindConfig.theme.extend.keyframes)) {
				if (typeof keyframeName !== "string") continue;
				const parsedKeyframeValue = keyframeValueSchema.safeParse(keyframeValue);
				if (!parsedKeyframeValue.success) continue;
				if (existingKeyFrameNodes?.find((node) => node.type === "atrule" && node.name === "keyframes" && node.params === keyframeName)) continue;
				const keyframeNode = postcss.atRule({
					name: "keyframes",
					params: keyframeName,
					nodes: [],
					raws: {
						semicolon: true,
						between: " ",
						before: "\n  "
					}
				});
				for (const [key, values] of Object.entries(parsedKeyframeValue.data)) {
					const rule = postcss.rule({
						selector: key,
						nodes: Object.entries(values).map(([key$1, value]) => postcss.decl({
							prop: key$1,
							value,
							raws: {
								semicolon: true,
								before: "\n      ",
								between: ": "
							}
						})),
						raws: {
							semicolon: true,
							between: " ",
							before: "\n    "
						}
					});
					keyframeNode.append(rule);
				}
				themeNode.append(keyframeNode);
				themeNode.insertBefore(keyframeNode, postcss.comment({ text: "---break---" }));
			}
		}
	};
}
function updateTailwindConfigAnimationPlugin(tailwindConfig) {
	return {
		postcssPlugin: "update-tailwind-config-animation",
		Once(root) {
			if (!tailwindConfig?.theme?.extend?.animation) return;
			const themeNode = upsertThemeNode(root);
			const existingAnimationNodes = themeNode.nodes?.filter((node) => node.type === "decl" && node.prop.startsWith("--animate-"));
			const parsedAnimationValue = z.record(z.string(), z.string()).safeParse(tailwindConfig.theme.extend.animation);
			if (!parsedAnimationValue.success) return;
			for (const [key, value] of Object.entries(parsedAnimationValue.data)) {
				const prop = `--animate-${key}`;
				if (existingAnimationNodes?.find((node) => node.prop === prop)) continue;
				const animationNode = postcss.decl({
					prop,
					value,
					raws: {
						semicolon: true,
						between: ": ",
						before: "\n  "
					}
				});
				themeNode.append(animationNode);
			}
		}
	};
}
function getQuoteType(root) {
	if (root.nodes[0].toString().includes("'")) return "single";
	return "double";
}
function isLocalHSLValue(value) {
	if (value.startsWith("hsl") || value.startsWith("rgb") || value.startsWith("#") || value.startsWith("oklch")) return false;
	const chunks = value.split(" ");
	return chunks.length === 3 && chunks.slice(1, 3).every((chunk) => chunk.includes("%"));
}
function isColorValue(value) {
	return value.startsWith("hsl") || value.startsWith("rgb") || value.startsWith("#") || value.startsWith("oklch") || value.includes("var(--color-");
}

//#endregion
//#region src/utils/updaters/update-env-vars.ts
async function updateEnvVars(envVars, config, options) {
	if (!envVars || Object.keys(envVars).length === 0) return {
		envVarsAdded: [],
		envFileUpdated: null,
		envFileCreated: null
	};
	options = {
		silent: false,
		...options
	};
	const envSpinner = spinner(`Adding environment variables.`, { silent: options.silent })?.start();
	const projectRoot = config.resolvedPaths.cwd;
	let envFilePath = path.join(projectRoot, ".env.local");
	const existingEnvFile = findExistingEnvFile(projectRoot);
	if (existingEnvFile) envFilePath = existingEnvFile;
	const envFileExists = existsSync(envFilePath);
	const envFileName = path.basename(envFilePath);
	const newEnvContent = Object.entries(envVars).map(([key, value]) => `${key}=${value}`).join("\n");
	let envVarsAdded = [];
	let envFileUpdated = null;
	let envFileCreated = null;
	if (envFileExists) {
		const existingContent = await promises.readFile(envFilePath, "utf-8");
		const mergedContent = mergeEnvContent(existingContent, newEnvContent);
		envVarsAdded = getNewEnvKeys(existingContent, newEnvContent);
		if (envVarsAdded.length > 0) {
			await promises.writeFile(envFilePath, mergedContent, "utf-8");
			envFileUpdated = path.relative(projectRoot, envFilePath);
			envSpinner?.succeed(`Added the following variables to ${highlighter.info(envFileName)}:`);
			if (!options.silent) for (const key of envVarsAdded) logger.log(`  ${highlighter.success("+")} ${key}`);
		} else envSpinner?.stop();
	} else {
		await promises.writeFile(envFilePath, `${newEnvContent}\n`, "utf-8");
		envFileCreated = path.relative(projectRoot, envFilePath);
		envVarsAdded = Object.keys(envVars);
		envSpinner?.succeed(`Added the following variables to ${highlighter.info(envFileName)}:`);
		if (!options.silent) for (const key of envVarsAdded) logger.log(`  ${highlighter.success("+")} ${key}`);
	}
	if (!options.silent && envVarsAdded.length > 0) logger.break();
	return {
		envVarsAdded,
		envFileUpdated,
		envFileCreated
	};
}

//#endregion
//#region src/utils/add-components.ts
async function addComponents(components, config, options) {
	options = {
		overwrite: false,
		silent: false,
		isNewProject: false,
		baseStyle: true,
		...options
	};
	const workspaceConfig = await getWorkspaceConfig(config);
	if (workspaceConfig && workspaceConfig.ui && workspaceConfig.ui.resolvedPaths.cwd !== config.resolvedPaths.cwd) return await addWorkspaceComponents(components, config, workspaceConfig, {
		...options,
		isRemote: components?.length === 1 && !!components[0].match(/\/chat\/b\//)
	});
	return await addProjectComponents(components, config, options);
}
async function addProjectComponents(components, config, options) {
	if (!options.baseStyle && !components.length) return;
	const registrySpinner = spinner(`Checking registry.`, { silent: options.silent })?.start();
	const tree = await resolveRegistryTree(components, configWithDefaults(config));
	if (!tree) {
		registrySpinner?.fail();
		return handleError(/* @__PURE__ */ new Error("Failed to fetch components from registry."));
	}
	try {
		validateFilesTarget(tree.files ?? [], config.resolvedPaths.cwd);
	} catch (error) {
		registrySpinner?.fail();
		return handleError(error);
	}
	registrySpinner?.succeed();
	const tailwindVersion = await getProjectTailwindVersionFromConfig(config);
	await updateTailwindConfig(tree.tailwind?.config, config, {
		silent: options.silent,
		tailwindVersion
	});
	const overwriteCssVars = await shouldOverwriteCssVars(components, config);
	const fontImports = resolveFontImports(config);
	await updateCssVars(overlayFontHeadingVar(tree.cssVars, config), config, {
		cleanupDefaultNextStyles: options.isNewProject,
		silent: options.silent,
		tailwindVersion,
		tailwindConfig: tree.tailwind?.config,
		overwriteCssVars,
		initIndex: options.baseStyle,
		fontImports
	});
	await updateCss(tree.css, config, { silent: options.silent });
	await updateEnvVars(tree.envVars, config, { silent: options.silent });
	await updateDependencies(replaceIconLibraryInDependencies(tree.dependencies, config.iconLibrary), tree.devDependencies, config, { silent: options.silent });
	await updateFiles(tree.files, config, {
		overwrite: options.overwrite,
		silent: options.silent,
		path: options.path
	});
	if (tree.docs) logger.info(tree.docs);
}
async function addWorkspaceComponents(components, config, workspaceConfig, options) {
	if (!options.baseStyle && !components.length) return;
	const registrySpinner = spinner(`Checking registry.`, { silent: options.silent })?.start();
	const tree = await resolveRegistryTree(components, configWithDefaults(config));
	if (!tree) {
		registrySpinner?.fail();
		return handleError(/* @__PURE__ */ new Error("Failed to fetch components from registry."));
	}
	try {
		validateFilesTarget(tree.files ?? [], config.resolvedPaths.cwd);
	} catch (error) {
		registrySpinner?.fail();
		return handleError(error);
	}
	registrySpinner?.succeed();
	const filesCreated = [];
	const filesUpdated = [];
	const filesSkipped = [];
	const rootSpinner = spinner(`Installing components.`)?.start();
	const mainTargetConfig = workspaceConfig.ui;
	const tailwindVersion = await getProjectTailwindVersionFromConfig(mainTargetConfig);
	const workspaceRoot = findCommonRoot(config.resolvedPaths.cwd, mainTargetConfig.resolvedPaths.ui);
	if (tree.tailwind?.config) {
		await updateTailwindConfig(tree.tailwind?.config, mainTargetConfig, {
			silent: true,
			tailwindVersion
		});
		filesUpdated.push(path.relative(workspaceRoot, mainTargetConfig.resolvedPaths.tailwindConfig));
	}
	if (tree.cssVars) {
		const overwriteCssVars = await shouldOverwriteCssVars(components, config);
		const fontImports = resolveFontImports(mainTargetConfig);
		await updateCssVars(overlayFontHeadingVar(tree.cssVars, mainTargetConfig), mainTargetConfig, {
			silent: true,
			tailwindVersion,
			tailwindConfig: tree.tailwind?.config,
			overwriteCssVars,
			fontImports
		});
		filesUpdated.push(path.relative(workspaceRoot, mainTargetConfig.resolvedPaths.tailwindCss));
	}
	if (tree.css) {
		await updateCss(tree.css, mainTargetConfig, { silent: true });
		filesUpdated.push(path.relative(workspaceRoot, mainTargetConfig.resolvedPaths.tailwindCss));
	}
	if (tree.envVars) await updateEnvVars(tree.envVars, mainTargetConfig, { silent: true });
	await updateDependencies(replaceIconLibraryInDependencies(tree.dependencies, mainTargetConfig.iconLibrary), tree.devDependencies, mainTargetConfig, { silent: true });
	const filesByType = /* @__PURE__ */ new Map();
	for (const file of tree.files ?? []) {
		const type = file.type || "registry:ui";
		if (!filesByType.has(type)) filesByType.set(type, []);
		filesByType.get(type).push(file);
	}
	for (const type of Array.from(filesByType.keys())) {
		const typeFiles = filesByType.get(type);
		let targetConfig = type === "registry:ui" ? workspaceConfig.ui : config;
		const typeWorkspaceRoot = findCommonRoot(config.resolvedPaths.cwd, targetConfig.resolvedPaths.ui || targetConfig.resolvedPaths.cwd);
		const packageRoot = await findPackageRoot(typeWorkspaceRoot, targetConfig.resolvedPaths.cwd) ?? targetConfig.resolvedPaths.cwd;
		const files = await updateFiles(typeFiles, targetConfig, {
			overwrite: options.overwrite,
			silent: true,
			rootSpinner,
			isRemote: options.isRemote,
			isWorkspace: true,
			path: options.path
		});
		filesCreated.push(...files.filesCreated.map((file) => path.relative(typeWorkspaceRoot, path.join(packageRoot, file))));
		filesUpdated.push(...files.filesUpdated.map((file) => path.relative(typeWorkspaceRoot, path.join(packageRoot, file))));
		filesSkipped.push(...files.filesSkipped.map((file) => path.relative(typeWorkspaceRoot, path.join(packageRoot, file))));
	}
	rootSpinner?.succeed();
	filesCreated.sort();
	filesUpdated.sort();
	filesSkipped.sort();
	if (!(filesCreated.length || filesUpdated.length) && !filesSkipped.length) spinner(`No files updated.`, { silent: options.silent })?.info();
	if (filesCreated.length) {
		spinner(`Created ${filesCreated.length} ${filesCreated.length === 1 ? "file" : "files"}:`, { silent: options.silent })?.succeed();
		for (const file of filesCreated) logger.log(`  - ${file}`);
	}
	if (filesUpdated.length) {
		spinner(`Updated ${filesUpdated.length} ${filesUpdated.length === 1 ? "file" : "files"}:`, { silent: options.silent })?.info();
		for (const file of filesUpdated) logger.log(`  - ${file}`);
	}
	if (filesSkipped.length) {
		spinner(`Skipped ${filesSkipped.length} ${filesUpdated.length === 1 ? "file" : "files"}: (use --overwrite to overwrite)`, { silent: options.silent })?.info();
		for (const file of filesSkipped) logger.log(`  - ${file}`);
	}
	if (tree.docs) logger.info(tree.docs);
}
/**
* Collects the Google Fonts `@import` strings that should be present in the
* project's CSS file for the active config. Body font + (optional) heading
* font, de-duplicated. Empty fonts are skipped.
*/
function resolveFontImports(config) {
	const imports = [];
	const push = (name) => {
		if (!name) return;
		const imp = getFontImport(name);
		if (imp && !imports.includes(imp)) imports.push(imp);
	};
	push(config.font);
	if (config.fontHeading && config.fontHeading !== "inherit" && config.fontHeading !== config.font) push(config.fontHeading);
	return imports;
}
/**
* Builds the CSS `@theme inline` value for `--font-heading` from the active
* config, mirroring the server-side `buildRegistryBase` output so older
* deployments of shadcn-vue.com (that don't emit `--font-heading` yet) still
* produce a working theme var on the CLI side. Once the registry route is
* redeployed this becomes redundant — the server's authoritative value will
* overwrite it via `updateCssVarsPluginV4` with `overwriteCssVars: true`.
*
* Returns `undefined` when there's no font configured at all.
*/
function resolveFontHeadingVar(config) {
	if (!config.font) return;
	const fontHeading = config.fontHeading;
	if (!fontHeading || fontHeading === "inherit" || fontHeading === config.font) return `var(${getFontVariable(config.font)})`;
	const headingFont = getFont(fontHeading);
	if (!headingFont) return;
	const fallback = headingFont.variable === "--font-mono" ? "monospace" : "sans-serif";
	return `'${headingFont.family} Variable', ${fallback}`;
}
/**
* Merges a synthesized `--font-heading` entry into the theme vars we pass to
* `updateCssVars`. Only runs when the server response didn't already include
* one, so a redeployed server stays authoritative.
*/
function overlayFontHeadingVar(cssVars, config) {
	const value = resolveFontHeadingVar(config);
	if (!value) return cssVars;
	const theme = cssVars?.theme ?? {};
	if (theme["--font-heading"]) return cssVars;
	return {
		...cssVars ?? {},
		theme: {
			...theme,
			"--font-heading": value
		}
	};
}
async function shouldOverwriteCssVars(components, config) {
	const result = await getRegistryItems(components, { config });
	return z.array(registryItemSchema).parse(result).some((component) => component.type === "registry:theme" || component.type === "registry:style" || component.type === "registry:base" || component.type === "registry:font");
}
function validateFilesTarget(files, cwd) {
	for (const file of files) {
		if (!file?.target) continue;
		if (!isSafeTarget(file.target, cwd)) throw new Error(`We found an unsafe file path "${file.target} in the registry item. Installation aborted.`);
	}
}

//#endregion
//#region src/utils/create-project.ts
const TEMPLATES = {
	nuxt: "nuxt",
	vite: "vite",
	astro: "astro",
	laravel: "laravel"
};
const REPO = "github:unovue/shadcn-vue";
async function createProject(options) {
	let template = options.template && TEMPLATES[options.template] ? options.template : "nuxt";
	let projectName = options.name ?? "my-vue-app";
	if (!options.force) {
		const { type, name } = await prompts([{
			type: options.template ? null : "select",
			name: "type",
			message: `The path ${highlighter.info(options.cwd)} does not contain a package.json file.\n  Would you like to start a new project?`,
			choices: [
				{
					title: "Nuxt 4",
					value: "nuxt"
				},
				{
					title: "Vite",
					value: "vite"
				},
				{
					title: "Astro",
					value: "astro"
				},
				{
					title: "Laravel",
					value: "laravel"
				}
			],
			initial: 0
		}, {
			type: options.name ? null : "text",
			name: "name",
			message: "What is your project named?",
			initial: projectName,
			format: (value) => value.trim(),
			validate: (value) => value.length > 128 ? `Name should be less than 128 characters.` : true
		}]);
		template = type ?? template;
		projectName = name ?? projectName;
	}
	const packageManager = await getPackageManager(options.cwd, { withFallback: true });
	const projectPath = `${options.cwd}/${projectName}`;
	try {
		await fsExtra.access(options.cwd, fsExtra.constants.W_OK);
	} catch {
		logger.break();
		logger.error(`The path ${highlighter.info(options.cwd)} is not writable.`);
		logger.error(`It is likely you do not have write permissions for this folder or the path ${highlighter.info(options.cwd)} does not exist.`);
		logger.break();
		process.exit(1);
	}
	if (fsExtra.existsSync(path.resolve(options.cwd, projectName, "package.json"))) {
		logger.break();
		logger.error(`A project with the name ${highlighter.info(projectName)} already exists.`);
		logger.error(`Please choose a different name and try again.`);
		logger.break();
		process.exit(1);
	}
	const createSpinner = spinner(`Creating a new ${template} project. This may take a few minutes.`).start();
	try {
		const templateDir = process.env.SHADCN_TEMPLATE_DIR;
		if (templateDir) await fsExtra.copy(path.join(templateDir, template), projectPath);
		else await downloadTemplate(`${REPO}/templates/${template}#dev`, { dir: projectPath });
		const pkgPath = path.join(projectPath, "package.json");
		const pkg = await fsExtra.readJson(pkgPath);
		pkg.name = projectName;
		await fsExtra.writeJson(pkgPath, pkg, { spaces: 2 });
		await x("git", ["init"], { nodeOptions: { cwd: projectPath } });
		await installDependencies({
			cwd: projectPath,
			packageManager,
			silent: true
		});
		createSpinner?.succeed(`Created a new ${template} project.`);
	} catch (error) {
		createSpinner?.fail(`Something went wrong creating a new ${template} project.`);
		handleError(error);
	}
	return {
		projectPath,
		projectName,
		template
	};
}

//#endregion
//#region src/utils/env-loader.ts
async function loadEnvFiles(cwd = process.cwd()) {
	try {
		const { config } = await import("@dotenvx/dotenvx");
		for (const envFile of [
			".env.local",
			".env.development.local",
			".env.development",
			".env"
		]) {
			const envPath = join(cwd, envFile);
			if (existsSync(envPath)) config({
				path: envPath,
				overload: false,
				quiet: true
			});
		}
	} catch (error) {
		logger.warn("Failed to load env files:", error);
	}
}

//#endregion
//#region src/utils/file-helper.ts
const FILE_BACKUP_SUFFIX = ".bak";
function createFileBackup(filePath) {
	if (!fsExtra.existsSync(filePath)) return null;
	const backupPath = `${filePath}${FILE_BACKUP_SUFFIX}`;
	try {
		fsExtra.renameSync(filePath, backupPath);
		return backupPath;
	} catch (error) {
		console.error(`Failed to create backup of ${filePath}: ${error}`);
		return null;
	}
}
function restoreFileBackup(filePath) {
	const backupPath = `${filePath}${FILE_BACKUP_SUFFIX}`;
	if (!fsExtra.existsSync(backupPath)) return false;
	try {
		fsExtra.renameSync(backupPath, filePath);
		return true;
	} catch (error) {
		console.error(`Warning: Could not restore backup file ${backupPath}: ${error}`);
		return false;
	}
}
function deleteFileBackup(filePath) {
	const backupPath = `${filePath}${FILE_BACKUP_SUFFIX}`;
	if (!fsExtra.existsSync(backupPath)) return false;
	try {
		fsExtra.unlinkSync(backupPath);
		return true;
	} catch {
		return false;
	}
}
/**
* Runs `task` with a **copy** of `filePath` kept under `<filePath><suffix>`,
* restoring it if the task throws OR if the process exits before the task
* returns. The original file stays in place for the task to read.
*
* This is the rollback primitive `apply` wraps around `runInit`. A plain
* try/catch isn't enough because `runInit → addComponents → handleError`
* funnels failures into `process.exit(1)`, which bypasses any catch block.
* The process-exit listener fires before that, so rollback still runs.
*
* Unlike shadcn-ui's `withFileBackup` (which renames), this copies so the
* original file stays readable by code that re-reads it during the task —
* shadcn-vue's `runInit` reads `components.json` from disk via
* `getProjectConfig`, so the original must stay in place.
*/
async function withFileCopyBackup(filePath, task, options = {}) {
	if (!fsExtra.existsSync(filePath)) return task();
	const backupPath = `${filePath}${options.suffix ?? FILE_BACKUP_SUFFIX}`;
	try {
		fsExtra.copyFileSync(filePath, backupPath);
	} catch (error) {
		options.onBackupFailure?.(filePath);
		throw new Error(`Could not back up ${filePath}: ${error.message}`);
	}
	const restoreOnExit = () => {
		if (fsExtra.existsSync(backupPath)) try {
			fsExtra.copyFileSync(backupPath, filePath);
			fsExtra.unlinkSync(backupPath);
		} catch {}
	};
	process.on("exit", restoreOnExit);
	try {
		const result = await task();
		process.removeListener("exit", restoreOnExit);
		if (fsExtra.existsSync(backupPath)) fsExtra.unlinkSync(backupPath);
		return result;
	} catch (error) {
		process.removeListener("exit", restoreOnExit);
		restoreOnExit();
		throw error;
	}
}

//#endregion
//#region src/utils/updaters/update-tailwind-content.ts
async function updateTailwindContent(content, config, options) {
	if (!content) return;
	options = {
		silent: false,
		...options
	};
	const tailwindFileRelativePath = path.relative(config.resolvedPaths.cwd, config.resolvedPaths.tailwindConfig);
	const tailwindSpinner = spinner(`Updating ${highlighter.info(tailwindFileRelativePath)}`, { silent: options.silent }).start();
	const output = await transformTailwindContent(await promises.readFile(config.resolvedPaths.tailwindConfig, "utf8"), content, config);
	await promises.writeFile(config.resolvedPaths.tailwindConfig, output, "utf8");
	tailwindSpinner?.succeed();
}
async function transformTailwindContent(input, content, config) {
	const sourceFile = await _createSourceFile(input, config);
	const configObject = sourceFile.getDescendantsOfKind(SyntaxKind.ObjectLiteralExpression).find((node) => node.getProperties().some((property) => property.isKind(SyntaxKind.PropertyAssignment) && property.getName() === "content"));
	if (!configObject) return input;
	addTailwindConfigContent(configObject, content);
	return sourceFile.getFullText();
}
async function addTailwindConfigContent(configObject, content) {
	const quoteChar = _getQuoteChar(configObject);
	const existingProperty = configObject.getProperty("content");
	if (!existingProperty) {
		const newProperty = {
			name: "content",
			initializer: `[${quoteChar}${content.join(`${quoteChar}, ${quoteChar}`)}${quoteChar}]`
		};
		configObject.addPropertyAssignment(newProperty);
		return configObject;
	}
	if (existingProperty.isKind(SyntaxKind.PropertyAssignment)) {
		const initializer = existingProperty.getInitializer();
		if (initializer?.isKind(SyntaxKind.ArrayLiteralExpression)) for (const contentItem of content) {
			const newValue = `${quoteChar}${contentItem}${quoteChar}`;
			if (initializer.getElements().map((element) => element.getText()).includes(newValue)) continue;
			initializer.addElement(newValue);
		}
		return configObject;
	}
	return configObject;
}

//#endregion
//#region src/commands/init.ts
process.on("exit", (code) => {
	const filePath = path.resolve(process.cwd(), "components.json");
	if (code === 0) return deleteFileBackup(filePath);
	return restoreFileBackup(filePath);
});
const initOptionsSchema = z.object({
	cwd: z.string(),
	name: z.string().optional(),
	preset: z.union([z.boolean(), z.string()]).optional(),
	registryBaseConfig: rawConfigSchema.deepPartial().optional(),
	installStyleIndex: z.boolean().optional(),
	components: z.array(z.string()).optional(),
	yes: z.boolean(),
	defaults: z.boolean(),
	force: z.boolean(),
	silent: z.boolean(),
	isNewProject: z.boolean(),
	srcDir: z.boolean().optional(),
	cssVariables: z.boolean(),
	template: z.string().optional().refine((val) => {
		if (val) return TEMPLATES[val];
		return true;
	}, { message: "Invalid template. Please use 'nuxt', 'vite', 'astro', or 'laravel'." }),
	base: z.string().optional().refine((val) => {
		if (val) return BASES.find((base) => base.name === val);
		return true;
	}, { message: `Invalid base. Please use '${BASES.map((base) => base.name).join("', '")}'` }),
	style: z.string().optional().refine((val) => {
		if (val) return STYLES.find((style) => style.name === val);
		return true;
	}, { message: `Invalid style. Please use '${STYLES.map((style) => style.name).join("', '")}'` }),
	iconLibrary: z.string().optional().refine((val) => {
		if (val) return ICON_LIBRARIES$1.find((lib) => lib.name === val);
		return true;
	}, { message: `Invalid icon library. Please use '${ICON_LIBRARIES$1.map((lib) => lib.name).join("', '")}'` }),
	font: z.string().optional().refine((val) => {
		if (val) return FONTS.find((font) => font.name === val);
		return true;
	}, { message: `Invalid font. Please use '${FONTS.map((font) => font.name).join("', '")}'` }),
	baseColor: z.string().optional().refine((val) => {
		if (val) return BASE_COLORS.find((color) => color.name === val);
		return true;
	}, { message: `Invalid base color. Please use '${BASE_COLORS.map((color) => color.name).join("', '")}'` }),
	baseStyle: z.boolean(),
	monorepo: z.boolean().optional(),
	reinstall: z.boolean().optional(),
	rtl: z.boolean().optional(),
	pointer: z.boolean().optional()
});
function applyInitUrlOptions(url, options) {
	if (options.rtl) url.searchParams.set("rtl", "true");
	else if (options.rtl === false) url.searchParams.delete("rtl");
	if (options.pointer) url.searchParams.set("pointer", "true");
	else if (options.pointer === false) url.searchParams.delete("pointer");
	return url;
}
const init = new Command().name("init").alias("create").description("initialize your project and install dependencies").argument("[components...]", "names, url or local path to component").option("-p, --preset [preset]", `use a preset configuration, preset code, or URL. (${Object.keys(DEFAULT_PRESETS).join(", ")})`).option("-t, --template <template>", "the template to use. (nuxt, vite, astro, laravel)").option("--base <base>", "the component library base to use. (reka)", void 0).option("--style <style>", "the visual style to use. (vega, nova, maia, lyra, mira)", void 0).option("--icon-library <icon-library>", "the icon library to use. (lucide, tabler, hugeicons, phosphor, remixicon)", void 0).option("--font <font>", "the font to use. (inter, figtree, jetbrains-mono, geist, geist-mono)", void 0).option("-b, --base-color <base-color>", "the base color to use. (neutral, gray, zinc, stone, slate)", void 0).option("-y, --yes", "skip confirmation prompt.", true).option("-d, --defaults,", "use default configuration.", false).option("-f, --force", "force overwrite of existing configuration.", false).option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("-s, --silent", "mute output.", false).option("--src-dir", "use the src directory when creating a new project.", false).option("--no-src-dir", "do not use the src directory when creating a new project.").option("--css-variables", "use css variables for theming.", true).option("--no-css-variables", "do not use css variables for theming.").option("--no-base-style", "do not install the base shadcn style.").option("-n, --name <name>", "the name for the new project.").option("--reinstall", "re-install existing UI components.").option("--no-reinstall", "do not re-install existing UI components.").option("--rtl", "enable RTL support.").option("--no-rtl", "disable RTL support.").option("--pointer", "enable pointer cursor for buttons.").option("--no-pointer", "disable pointer cursor for buttons.").action(async (components, opts) => {
	if (opts.monorepo === true) {
		logger.break();
		logger.warn("The --monorepo flag is not yet supported in shadcn-vue.");
		process.exit(1);
	}
	if (opts.rtl === true) logger.info("RTL support enabled in config. Run `shadcn-vue migrate rtl` to transform installed components.");
	try {
		const presetsByName = DEFAULT_PRESETS;
		if (typeof opts.preset === "string" && !isUrl(opts.preset) && !isPresetCode(opts.preset) && !(opts.preset in presetsByName)) {
			logger.error(`Invalid preset: ${highlighter.info(opts.preset)}. Available presets: ${Object.keys(presetsByName).join(", ")}`);
			process.exit(1);
		}
		if (opts.defaults && opts.preset === void 0) opts.preset = "nova";
		if (opts.defaults) {
			opts.template = opts.template || "nuxt";
			opts.base = opts.base || "reka";
		}
		const options = initOptionsSchema.parse({
			cwd: path.resolve(opts.cwd),
			isNewProject: false,
			components,
			...opts
		});
		await loadEnvFiles(options.cwd);
		if (options.preset !== void 0) {
			const presetArg = options.preset === true ? true : options.preset;
			if (presetArg === true) {
				const result = await promptForPreset({
					rtl: options.rtl ?? false,
					pointer: options.pointer,
					template: options.template,
					base: options.base ?? await promptForBase()
				});
				if (result.kind === "cancelled") {
					logger.break();
					process.exit(1);
				}
				if (result.kind === "custom") {
					logger.break();
					process.exit(0);
				}
				components = [result.url, ...components];
			}
			if (typeof presetArg === "string") {
				let initUrl;
				if (isUrl(presetArg)) {
					const url = new URL(presetArg);
					applyInitUrlOptions(url, options);
					initUrl = url.toString();
				} else if (isPresetCode(presetArg)) {
					const decoded = decodePreset(presetArg);
					if (!decoded) {
						logger.error(`Invalid preset code: ${highlighter.info(presetArg)}`);
						logger.break();
						process.exit(1);
					}
					initUrl = resolveInitUrl({
						...decoded,
						base: options.base ?? "reka",
						rtl: options.rtl ?? false
					}, {
						template: options.template,
						preset: presetArg,
						pointer: options.pointer
					});
				} else {
					const preset = presetsByName[presetArg];
					if (!preset) throw new Error(`Unknown preset: ${presetArg}`);
					initUrl = resolveInitUrl({
						...preset,
						base: options.base ?? preset.base,
						rtl: options.rtl ?? preset.rtl
					}, {
						template: options.template,
						pointer: options.pointer
					});
				}
				components = [initUrl, ...components];
			}
			const { registryBaseConfig, installStyleIndex, url: cleanUrl } = await resolveRegistryBaseConfig(components[0], path.resolve(opts.cwd));
			components[0] = cleanUrl;
			if (registryBaseConfig) options.registryBaseConfig = registryBaseConfig;
			if (!installStyleIndex) {
				options.installStyleIndex = false;
				options.baseStyle = false;
			}
			options.components = components;
		}
		if (components.length > 0) {
			let shadowConfig = configWithDefaults(createConfig({ resolvedPaths: { cwd: options.cwd } }));
			const componentsJsonPath = path.resolve(options.cwd, "components.json");
			if (fsExtra.existsSync(componentsJsonPath)) {
				const existingConfig = await fsExtra.readJson(componentsJsonPath);
				const config = rawConfigSchema.partial().parse(existingConfig);
				const baseConfig = createConfig({ resolvedPaths: { cwd: options.cwd } });
				shadowConfig = configWithDefaults({
					...config,
					resolvedPaths: {
						...baseConfig.resolvedPaths,
						cwd: options.cwd
					}
				});
				createFileBackup(componentsJsonPath);
			}
			const { config: updatedConfig } = await ensureRegistriesInConfig(components, shadowConfig, {
				silent: true,
				writeFile: false
			});
			shadowConfig = updatedConfig;
			buildUrlAndHeadersForRegistryItem(components[0], shadowConfig);
			const [item] = await getRegistryItems([components[0]], { config: shadowConfig });
			if (item?.type === "registry:style") {
				options.baseColor = "neutral";
				options.baseStyle = item.extends === "none" ? false : options.baseStyle;
			}
		}
		if (!options.baseStyle) options.baseColor = "neutral";
		await runInit(options);
		logger.log(`${highlighter.success("Success!")} Project initialization completed.\nYou may now add components.`);
		deleteFileBackup(path.resolve(options.cwd, "components.json"));
		logger.break();
	} catch (error) {
		logger.break();
		handleError(error);
	} finally {
		clearRegistryContext();
	}
});
async function runInit(options) {
	let projectInfo;
	if (!options.skipPreflight) {
		const preflight = await preFlightInit(options);
		if (preflight.errors[MISSING_DIR_OR_EMPTY_PROJECT]) {
			const { projectPath } = await createProject(options);
			if (!projectPath) process.exit(1);
			options.cwd = projectPath;
			options.isNewProject = true;
		}
		projectInfo = preflight.projectInfo;
	} else projectInfo = await getProjectInfo(options.cwd);
	const projectConfig = await getProjectConfig(options.cwd, projectInfo);
	let config = projectConfig ? await promptForMinimalConfig(projectConfig, options) : await promptForConfig(await getConfig(options.cwd), options);
	if (!options.yes) {
		const { proceed } = await prompts({
			type: "confirm",
			name: "proceed",
			message: `Write configuration to ${highlighter.info("components.json")}. Proceed?`,
			initial: true
		});
		if (!proceed) process.exit(0);
	}
	const components = [...options.baseStyle ? ["index"] : [], ...options.components ?? []];
	const { config: configWithRegistries } = await ensureRegistriesInConfig(components, await resolveConfigPaths(options.cwd, config), { silent: true });
	if (configWithRegistries.registries) config.registries = configWithRegistries.registries;
	const componentSpinner = spinner(`Writing components.json.`).start();
	const targetPath = path.resolve(options.cwd, "components.json");
	const backupPath = `${targetPath}${FILE_BACKUP_SUFFIX}`;
	const mergeConfig = (base, override) => {
		const { registries, ...merged } = deepmerge(base, override);
		return {
			...merged,
			registries
		};
	};
	if (!options.force && fsExtra.existsSync(backupPath)) config = mergeConfig(await fsExtra.readJson(backupPath), config);
	if (options.registryBaseConfig) config = mergeConfig(config, options.registryBaseConfig);
	if (options.rtl !== void 0) config.rtl = options.rtl;
	if (options.pointer !== void 0) config.pointer = options.pointer;
	config.registries = Object.fromEntries(Object.entries(config.registries || {}).filter(([key]) => !Object.keys(BUILTIN_REGISTRIES).includes(key)));
	await promises.writeFile(targetPath, `${JSON.stringify(config, null, 2)}\n`, "utf8");
	componentSpinner.succeed();
	const fullConfig = await resolveConfigPaths(options.cwd, config);
	await addComponents(components, fullConfig, {
		overwrite: true,
		silent: options.silent,
		baseStyle: options.baseStyle,
		isNewProject: options.isNewProject || projectInfo?.framework.name === "nuxt4"
	});
	if (options.isNewProject && options.srcDir) await updateTailwindContent(["./src/**/*.{js,ts,jsx,tsx,mdx}"], fullConfig, { silent: options.silent });
	return fullConfig;
}
async function promptForConfig(defaultConfig = null, opts) {
	let base = opts?.base ?? "reka";
	let style = opts?.style ?? "vega";
	let font = opts?.font ?? "inter";
	let iconLibrary = opts?.iconLibrary ?? "lucide";
	let baseColor = opts?.baseColor ?? "neutral";
	let typescript = defaultConfig?.typescript ?? true;
	let tailwindCss = defaultConfig?.tailwind.css ?? DEFAULT_TAILWIND_CSS;
	let tailwindCssVariables = opts?.cssVariables ?? defaultConfig?.tailwind.cssVariables ?? true;
	let tailwindPrefix = defaultConfig?.tailwind.prefix ?? "";
	let tailwindConfig = defaultConfig?.tailwind.config ?? DEFAULT_TAILWIND_CONFIG;
	let componentsAlias = defaultConfig?.aliases.components ?? DEFAULT_COMPONENTS;
	let utilsAlias = defaultConfig?.aliases.utils ?? DEFAULT_UTILS;
	if (opts?.preset === void 0 && !opts?.defaults) {
		const [styles, baseColors] = await Promise.all([getRegistryStyles(), getRegistryBaseColors()]);
		logger.info("");
		const options = await prompts([
			{
				type: "toggle",
				name: "typescript",
				message: `Would you like to use ${highlighter.info("TypeScript")} (recommended)?`,
				initial: typescript,
				active: "yes",
				inactive: "no"
			},
			{
				type: "select",
				name: "base",
				message: `Which ${highlighter.info("component library")} would you like to use?`,
				choices: BASES.map((b) => ({
					title: b.label,
					value: b.name,
					description: b.description
				})),
				initial: 0
			},
			{
				type: "select",
				name: "style",
				message: `Which ${highlighter.info("visual style")} would you like to use?`,
				choices: [...STYLES.map((s) => ({
					title: s.label,
					value: s.name,
					description: s.description
				})), ...styles.filter((s) => !STYLES.find((st) => st.name === s.name)).map((s) => ({
					title: s.label,
					value: s.name
				}))],
				initial: 0
			},
			{
				type: "select",
				name: "iconLibrary",
				message: `Which ${highlighter.info("icon library")} would you like to use?`,
				choices: ICON_LIBRARIES$1.map((lib) => ({
					title: lib.label,
					value: lib.name
				})),
				initial: 0
			},
			{
				type: "select",
				name: "font",
				message: `Which ${highlighter.info("font")} would you like to use?`,
				choices: FONTS.map((f) => ({
					title: f.label,
					value: f.name
				})),
				initial: 0
			},
			{
				type: "select",
				name: "tailwindBaseColor",
				message: `Which color would you like to use as the ${highlighter.info("base color")}?`,
				choices: baseColors.map((color) => ({
					title: color.label,
					value: color.name
				}))
			},
			{
				type: "text",
				name: "tailwindCss",
				message: `Where is your ${highlighter.info("global CSS")} file?`,
				initial: tailwindCss
			},
			{
				type: "toggle",
				name: "tailwindCssVariables",
				message: `Would you like to use ${highlighter.info("CSS variables")} for theming?`,
				initial: tailwindCssVariables,
				active: "yes",
				inactive: "no"
			},
			{
				type: "text",
				name: "tailwindPrefix",
				message: `Are you using a custom ${highlighter.info("tailwind prefix eg. tw-")}? (Leave blank if not)`,
				initial: ""
			},
			{
				type: "text",
				name: "tailwindConfig",
				message: `Where is your ${highlighter.info("tailwind.config.js")} located?`,
				initial: tailwindConfig
			},
			{
				type: "text",
				name: "components",
				message: `Configure the import alias for ${highlighter.info("components")}:`,
				initial: componentsAlias
			},
			{
				type: "text",
				name: "utils",
				message: `Configure the import alias for ${highlighter.info("utils")}:`,
				initial: utilsAlias
			}
		]);
		base = options.base ?? base;
		style = options.style ?? style;
		font = options.font ?? font;
		iconLibrary = options.iconLibrary ?? iconLibrary;
		baseColor = options.tailwindBaseColor ?? baseColor;
		typescript = options.typescript ?? typescript;
		tailwindCss = options.tailwindCss ?? tailwindCss;
		tailwindCssVariables = options.tailwindCssVariables ?? tailwindCssVariables;
		tailwindPrefix = options.tailwindPrefix ?? tailwindPrefix;
		tailwindConfig = options.tailwindConfig ?? tailwindConfig;
		componentsAlias = options.components ?? componentsAlias;
		utilsAlias = options.utils ?? utilsAlias;
	}
	return rawConfigSchema.parse({
		$schema: "https://shadcn-vue.com/schema.json",
		style: composeStyleId(base, style),
		font,
		iconLibrary,
		rtl: opts?.rtl ?? false,
		pointer: opts?.pointer ?? false,
		tailwind: {
			config: tailwindConfig,
			css: tailwindCss,
			baseColor,
			cssVariables: tailwindCssVariables,
			prefix: tailwindPrefix
		},
		typescript,
		aliases: {
			utils: utilsAlias,
			components: componentsAlias,
			lib: componentsAlias.replace(/\/components$/, "/lib"),
			composables: componentsAlias.replace(/\/components$/, "/composables")
		}
	});
}
async function promptForMinimalConfig(defaultConfig, opts) {
	let base = opts.base ?? "reka";
	let style = opts.style ?? defaultConfig.style;
	let iconLibrary = opts.iconLibrary ?? defaultConfig.iconLibrary ?? "lucide";
	let font = opts.font ?? defaultConfig.font ?? "inter";
	let baseColor = opts.baseColor ?? defaultConfig.tailwind.baseColor;
	let cssVariables = defaultConfig.tailwind.cssVariables;
	if (opts.preset === void 0 && !opts.defaults) {
		const [styles, baseColors, tailwindVersion] = await Promise.all([
			getRegistryStyles(),
			getRegistryBaseColors(),
			getProjectTailwindVersionFromConfig(defaultConfig)
		]);
		const options = await prompts([
			{
				type: opts.base ? null : "select",
				name: "base",
				message: `Which ${highlighter.info("component library")} would you like to use?`,
				choices: BASES.map((b) => ({
					title: b.label,
					value: b.name,
					description: b.description
				})),
				initial: 0
			},
			{
				type: tailwindVersion === "v4" || opts.style ? null : "select",
				name: "style",
				message: `Which ${highlighter.info("visual style")} would you like to use?`,
				choices: [...STYLES.map((s) => ({
					title: s.name === "vega" ? "Vega (Recommended)" : s.label,
					value: s.name,
					description: s.description
				})), ...styles.filter((s) => !STYLES.find((st) => st.name === s.name)).map((s) => ({
					title: s.label,
					value: s.name
				}))],
				initial: 0
			},
			{
				type: opts.iconLibrary ? null : "select",
				name: "iconLibrary",
				message: `Which ${highlighter.info("icon library")} would you like to use?`,
				choices: ICON_LIBRARIES$1.map((lib) => ({
					title: lib.label,
					value: lib.name
				})),
				initial: 0
			},
			{
				type: opts.font ? null : "select",
				name: "font",
				message: `Which ${highlighter.info("font")} would you like to use?`,
				choices: FONTS.map((f) => ({
					title: f.label,
					value: f.name
				})),
				initial: 0
			},
			{
				type: opts.baseColor ? null : "select",
				name: "tailwindBaseColor",
				message: `Which color would you like to use as the ${highlighter.info("base color")}?`,
				choices: baseColors.map((color) => ({
					title: color.label,
					value: color.name
				}))
			}
		]);
		base = options.base ?? base;
		style = options.style ?? style ?? "vega";
		iconLibrary = options.iconLibrary ?? iconLibrary;
		font = options.font ?? font;
		baseColor = options.tailwindBaseColor ?? baseColor;
		cssVariables = opts.cssVariables;
	}
	return rawConfigSchema.parse({
		$schema: defaultConfig?.$schema,
		style: composeStyleId(base, style),
		font,
		...defaultConfig.fontHeading && { fontHeading: defaultConfig.fontHeading },
		iconLibrary,
		rtl: opts.rtl ?? defaultConfig.rtl ?? false,
		pointer: opts.pointer ?? defaultConfig.pointer ?? false,
		tailwind: {
			...defaultConfig?.tailwind,
			baseColor,
			cssVariables
		},
		typescript: defaultConfig.typescript,
		aliases: defaultConfig?.aliases
	});
}

//#endregion
//#region src/preflights/preflight-add.ts
async function preFlightAdd(options) {
	const errors = {};
	if (!fsExtra.existsSync(options.cwd) || !fsExtra.existsSync(path.resolve(options.cwd, "package.json"))) {
		errors[MISSING_DIR_OR_EMPTY_PROJECT] = true;
		return {
			errors,
			config: null
		};
	}
	if (!fsExtra.existsSync(path.resolve(options.cwd, "components.json"))) {
		errors[MISSING_CONFIG] = true;
		return {
			errors,
			config: null
		};
	}
	try {
		return {
			errors,
			config: await getConfig(options.cwd)
		};
	} catch (error) {
		logger.break();
		logger.error(`An invalid ${highlighter.info("components.json")} file was found at ${highlighter.info(options.cwd)}.\nBefore you can add components, you must create a valid ${highlighter.info("components.json")} file by running the ${highlighter.info("init")} command.`);
		logger.error(`Learn more at ${highlighter.info("https://shadcn-vue.com/docs/components-json")}.`);
		logger.break();
		process.exit(1);
	}
}

//#endregion
//#region src/commands/add.ts
const addOptionsSchema = z.object({
	components: z.array(z.string()).optional(),
	yes: z.boolean(),
	overwrite: z.boolean(),
	cwd: z.string(),
	all: z.boolean(),
	path: z.string().optional(),
	silent: z.boolean(),
	srcDir: z.boolean().optional(),
	cssVariables: z.boolean(),
	dryRun: z.boolean().optional(),
	diff: z.union([z.string(), z.literal(true)]).optional(),
	view: z.union([z.string(), z.literal(true)]).optional()
});
const add = new Command().name("add").description("add a component to your project").argument("[components...]", "names, url or local path to component").option("-y, --yes", "skip confirmation prompt.", false).option("-o, --overwrite", "overwrite existing files.", false).option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("-a, --all", "add all available components", false).option("-p, --path <path>", "the path to add the component to.").option("-s, --silent", "mute output.", false).option("--css-variables", "use css variables for theming.", true).option("--no-css-variables", "do not use css variables for theming.").option("--dry-run", "preview changes without writing files.", false).option("--diff [path]", "show diff for a file.").option("--view [path]", "show file contents.").action(async (components, opts) => {
	try {
		const options = addOptionsSchema.parse({
			components,
			cwd: path.resolve(opts.cwd),
			...opts
		});
		if (options.dryRun || options.diff || options.view) {
			logger.break();
			logger.warn("The --dry-run, --diff and --view options are not yet supported in shadcn-vue.");
			logger.info("Follow https://github.com/unovue/shadcn-vue for updates.");
			logger.break();
			process.exit(1);
		}
		await loadEnvFiles(options.cwd);
		let initialConfig = await getConfig(options.cwd);
		if (!initialConfig) initialConfig = createConfig({
			style: "new-york",
			resolvedPaths: { cwd: options.cwd }
		});
		let hasNewRegistries = false;
		if (components.length > 0) {
			const { config: updatedConfig$1, newRegistries } = await ensureRegistriesInConfig(components, initialConfig, {
				silent: options.silent,
				writeFile: false
			});
			initialConfig = updatedConfig$1;
			hasNewRegistries = newRegistries.length > 0;
		}
		let itemType;
		let shouldInstallBaseStyle = true;
		if (components.length > 0) {
			const [registryItem] = await getRegistryItems([components[0]], { config: initialConfig });
			itemType = registryItem?.type;
			shouldInstallBaseStyle = itemType !== "registry:theme" && itemType !== "registry:style";
			if (isUniversalRegistryItem(registryItem)) {
				await addComponents(components, initialConfig, {
					...options,
					baseStyle: shouldInstallBaseStyle
				});
				return;
			}
			if (!options.yes && (itemType === "registry:style" || itemType === "registry:theme")) {
				logger.break();
				const { confirm } = await prompts({
					type: "confirm",
					name: "confirm",
					message: highlighter.warn(`You are about to install a new ${itemType.replace("registry:", "")}. \nExisting CSS variables and components will be overwritten. Continue?`)
				});
				if (!confirm) {
					logger.break();
					logger.log(`Installation cancelled.`);
					logger.break();
					process.exit(1);
				}
			}
		}
		if (!options.components?.length) options.components = await promptForRegistryComponents(options);
		if ((await getProjectInfo(options.cwd))?.tailwindVersion === "v4") {
			const deprecatedComponents = DEPRECATED_COMPONENTS.filter((component) => options.components?.includes(component.name));
			if (deprecatedComponents?.length) {
				logger.break();
				deprecatedComponents.forEach((component) => {
					logger.warn(highlighter.warn(component.message));
				});
				logger.break();
				process.exit(1);
			}
		}
		let { errors, config } = await preFlightAdd(options);
		if (errors[MISSING_CONFIG]) {
			const { proceed } = await prompts({
				type: "confirm",
				name: "proceed",
				message: `You need to create a ${highlighter.info("components.json")} file to add components. Proceed?`,
				initial: true
			});
			if (!proceed) {
				logger.break();
				process.exit(1);
			}
			config = await runInit({
				cwd: options.cwd,
				yes: true,
				force: true,
				defaults: false,
				skipPreflight: false,
				silent: options.silent && !hasNewRegistries,
				isNewProject: false,
				srcDir: options.srcDir,
				cssVariables: options.cssVariables,
				baseStyle: shouldInstallBaseStyle,
				baseColor: shouldInstallBaseStyle ? void 0 : "neutral",
				components: options.components
			});
		}
		if (!config) throw new Error(`Failed to read config at ${highlighter.info(options.cwd)}.`);
		const { config: updatedConfig } = await ensureRegistriesInConfig(options.components, config, { silent: options.silent || hasNewRegistries });
		config = updatedConfig;
		await addComponents(options.components, config, {
			...options,
			baseStyle: shouldInstallBaseStyle
		});
	} catch (error) {
		logger.break();
		handleError(error);
	} finally {
		clearRegistryContext();
	}
});
async function promptForRegistryComponents(options) {
	const registryIndex = await getShadcnRegistryIndex();
	if (!registryIndex) {
		logger.break();
		handleError(/* @__PURE__ */ new Error("Failed to fetch registry index."));
		return [];
	}
	if (options.all) return registryIndex.map((entry) => entry.name).filter((component) => !DEPRECATED_COMPONENTS.some((c) => c.name === component));
	if (options.components?.length) return options.components;
	const { components } = await prompts({
		type: "multiselect",
		name: "components",
		message: "Which components would you like to add?",
		hint: "Space to select. A to toggle all. Enter to submit.",
		instructions: false,
		choices: registryIndex.filter((entry) => entry.type === "registry:ui" && !DEPRECATED_COMPONENTS.some((component) => component.name === entry.name)).map((entry) => ({
			title: entry.name,
			value: entry.name,
			selected: options.all ? true : options.components?.includes(entry.name)
		}))
	});
	if (!components?.length) {
		logger.warn("No components selected. Exiting.");
		logger.info("");
		process.exit(1);
	}
	const result = z.array(z.string()).safeParse(components);
	if (!result.success) {
		logger.error("");
		handleError(/* @__PURE__ */ new Error("Something went wrong. Please try again."));
		return [];
	}
	return result.data;
}

//#endregion
//#region src/preflights/preflight-apply.ts
async function preFlightApply(options) {
	const errors = {};
	if (!fsExtra.existsSync(options.cwd) || !fsExtra.existsSync(path.resolve(options.cwd, "package.json"))) {
		errors[MISSING_DIR_OR_EMPTY_PROJECT] = true;
		return {
			errors,
			config: null
		};
	}
	if (!fsExtra.existsSync(path.resolve(options.cwd, "components.json"))) {
		errors[MISSING_CONFIG] = true;
		return {
			errors,
			config: null
		};
	}
	try {
		const config = await getConfig(options.cwd);
		if (!config) reportInvalidConfig(options.cwd);
		return {
			errors,
			config
		};
	} catch {
		reportInvalidConfig(options.cwd);
	}
}
function reportInvalidConfig(cwd) {
	logger.break();
	logger.error(`An invalid or empty ${highlighter.info("components.json")} file was found at ${highlighter.info(cwd)}.\nBefore you can apply a preset, you must create a valid ${highlighter.info("components.json")} file by running the ${highlighter.info("init")} command.`);
	logger.break();
	process.exit(1);
}

//#endregion
//#region src/commands/apply.ts
const applyOptionsSchema = z.object({
	cwd: z.string(),
	positionalPreset: z.string().optional(),
	preset: z.string().optional(),
	yes: z.boolean(),
	silent: z.boolean()
});
const PRESET_NAMES = Object.keys(DEFAULT_PRESETS);
const apply = new Command().name("apply").description("apply a preset to an existing project").argument("[preset]", "the preset to apply").option("--preset <preset>", "preset configuration to apply").option("-y, --yes", "skip confirmation prompt.", false).option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("-s, --silent", "mute output.", false).action(async (positionalPreset, opts) => {
	try {
		const options = applyOptionsSchema.parse({
			...opts,
			cwd: path.resolve(opts.cwd),
			positionalPreset
		});
		const presetValue = resolveApplyPreset(options);
		const preflight = await preFlightApply(options);
		if (preflight.errors[MISSING_DIR_OR_EMPTY_PROJECT]) {
			logger.break();
			logger.error(`The ${highlighter.info("apply")} command only works in an existing project.`);
			logger.error(`Run ${highlighter.info(getInitCommand(presetValue))} first.`);
			logger.break();
			process.exit(1);
		}
		if (preflight.errors[MISSING_CONFIG]) {
			logger.break();
			logger.error(`No ${highlighter.info("components.json")} found at ${highlighter.info(options.cwd)}.`);
			logger.error(`Run ${highlighter.info(getInitCommand(presetValue))} first.`);
			logger.break();
			process.exit(1);
		}
		const existingConfig = preflight.config;
		if (!existingConfig) process.exit(1);
		if (!presetValue) {
			logger.break();
			logger.error(`Please provide a preset to apply.\nAvailable presets: ${PRESET_NAMES.join(", ")}\nYou can also pass an encoded preset code or a preset URL.`);
			logger.break();
			process.exit(1);
		}
		if (!isUrl(presetValue) && !isPresetCode(presetValue) && !(presetValue in DEFAULT_PRESETS)) {
			logger.break();
			logger.error(`Invalid preset: ${highlighter.info(presetValue)}.\nAvailable presets: ${PRESET_NAMES.join(", ")}`);
			logger.break();
			process.exit(1);
		}
		const currentBase = getBase(existingConfig.style);
		const currentRtl = existingConfig.rtl ?? false;
		const initUrl = resolveApplyInitUrl(presetValue, {
			base: currentBase,
			rtl: currentRtl
		});
		if (!initUrl) {
			logger.break();
			logger.error(`Invalid preset: ${highlighter.info(presetValue)}.\nAvailable presets: ${PRESET_NAMES.join(", ")}`);
			logger.break();
			process.exit(1);
		}
		const reinstallComponents = await getProjectComponents(options.cwd);
		if (!options.yes) {
			logger.break();
			logger.warn(highlighter.warn(`Applying a new preset will overwrite existing UI components, fonts, and CSS variables.`));
			logger.warn(`Commit or stash your changes before continuing so you can easily go back.`);
			logger.break();
			logger.log("  The following components will be re-installed:");
			if (reinstallComponents.length) for (let i = 0; i < reinstallComponents.length; i += 8) logger.log(`  - ${reinstallComponents.slice(i, i + 8).join(", ")}`);
			else logger.log("  - No installed UI components were detected.");
			logger.break();
			const { proceed } = await prompts({
				type: "confirm",
				name: "proceed",
				message: "Would you like to continue?",
				initial: false
			});
			if (!proceed) {
				logger.break();
				process.exit(1);
			}
		}
		await loadEnvFiles(options.cwd);
		await withFileCopyBackup(path.resolve(options.cwd, "components.json"), async () => {
			const { registryBaseConfig, installStyleIndex, url: cleanInitUrl } = await resolveRegistryBaseConfig(initUrl, options.cwd, { registries: existingConfig.registries });
			await runInit({
				cwd: options.cwd,
				preset: presetValue,
				base: currentBase,
				registryBaseConfig,
				installStyleIndex,
				components: [cleanInitUrl, ...reinstallComponents],
				yes: true,
				force: false,
				defaults: false,
				silent: options.silent,
				isNewProject: false,
				cssVariables: existingConfig.tailwind?.cssVariables ?? true,
				baseStyle: installStyleIndex !== false,
				rtl: currentRtl,
				reinstall: true,
				skipPreflight: true
			});
		}, {
			suffix: ".apply.bak",
			onBackupFailure: () => {
				logger.error(`Could not back up ${highlighter.info("components.json")}. Aborting.`);
			}
		});
		logger.break();
		logger.log(`${highlighter.success("Success!")} Preset applied.`);
		logger.break();
	} catch (error) {
		logger.break();
		handleError(error);
	} finally {
		clearRegistryContext();
	}
});
function resolveApplyPreset(options) {
	const positionalPreset = options.positionalPreset?.trim();
	const flagPreset = options.preset?.trim();
	if (positionalPreset && flagPreset && positionalPreset !== flagPreset) {
		logger.error(`Received two different preset values. Use either the positional preset or ${highlighter.info("--preset")}, or pass the same value to both.`);
		logger.break();
		process.exit(1);
	}
	return flagPreset ?? positionalPreset;
}
function quoteShellArg(value) {
	return /[^\w./:-]/.test(value) ? JSON.stringify(value) : value;
}
function getInitCommand(preset) {
	if (!preset) return "shadcn-vue init";
	return `shadcn-vue init --preset ${quoteShellArg(preset)}`;
}
/**
* Returns the component-library base for a composed style identifier
* (e.g. `reka-vega` → `reka`). Falls back to the first registered base
* (`reka`) when the style is empty or unrecognised.
*
* Mirrors shadcn-ui's `getBase` helper but adapted to shadcn-vue's
* `${base}-${visualStyle}` style id convention.
*/
function getBase(style) {
	const fallback = BASES[0]?.name ?? "reka";
	if (!style) return fallback;
	const [maybeBase] = style.split("-");
	if (maybeBase && BASES.find((b) => b.name === maybeBase)) return maybeBase;
	return fallback;
}
/**
* Resolves a preset input (named, encoded, or URL) into the final `/init`
* registry URL that `resolveRegistryBaseConfig` will fetch. The project's
* current `base` and `rtl` are forced onto the resulting URL so applying a
* preset never silently switches the user's component library or RTL setting.
*
* This is the shadcn-vue equivalent of shadcn-ui's `resolveApplyInitUrl`.
*/
function resolveApplyInitUrl(presetArg, options) {
	if (isUrl(presetArg)) {
		const url = new URL(presetArg);
		if (url.pathname === "/init" && presetArg.startsWith(SHADCN_VUE_URL)) url.searchParams.set("track", "1");
		url.searchParams.set("base", options.base);
		url.searchParams.set("rtl", String(options.rtl));
		return url.toString();
	}
	if (isPresetCode(presetArg)) {
		const decoded = decodePreset(presetArg);
		if (!decoded) {
			logger.error(`Invalid preset code: ${highlighter.info(presetArg)}`);
			logger.break();
			process.exit(1);
		}
		return resolveInitUrl({
			...decoded,
			base: options.base,
			rtl: options.rtl
		}, {
			template: options.template,
			preset: presetArg
		});
	}
	const preset = DEFAULT_PRESETS[presetArg];
	if (!preset) return null;
	return resolveInitUrl({
		...preset,
		base: options.base,
		rtl: options.rtl
	}, { template: options.template });
}

//#endregion
//#region src/preflights/preflight-build.ts
async function preFlightBuild(options) {
	const errors = {};
	const resolvePaths = {
		cwd: options.cwd,
		registryFile: path.resolve(options.cwd, options.registryFile),
		outputDir: path.resolve(options.cwd, options.outputDir)
	};
	if (!fsExtra.existsSync(resolvePaths.registryFile)) errors[BUILD_MISSING_REGISTRY_FILE] = true;
	await fsExtra.mkdir(resolvePaths.outputDir, { recursive: true });
	if (Object.keys(errors).length > 0) {
		if (errors[BUILD_MISSING_REGISTRY_FILE]) {
			logger.break();
			logger.error(`The path ${highlighter.info(resolvePaths.registryFile)} does not exist.`);
		}
		logger.break();
		process.exit(1);
	}
	return {
		errors,
		resolvePaths
	};
}

//#endregion
//#region src/commands/build.ts
const buildOptionsSchema = z.object({
	cwd: z.string(),
	registryFile: z.string(),
	outputDir: z.string()
});
const build = new Command().name("build").description("build components for a shadcn-vue registry").argument("[registry]", "path to registry.json file", "./registry.json").option("-o, --output <path>", "destination directory for json files", "./public/r").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).action(async (registry, opts) => {
	try {
		const { resolvePaths } = await preFlightBuild(buildOptionsSchema.parse({
			cwd: path$2.resolve(opts.cwd),
			registryFile: registry,
			outputDir: opts.output
		}));
		const content = await fs$1.readFile(resolvePaths.registryFile, "utf-8");
		const result = registrySchema.safeParse(JSON.parse(content));
		if (!result.success) {
			logger.error(`Invalid registry file found at ${highlighter.info(resolvePaths.registryFile)}.`);
			process.exit(1);
		}
		const buildSpinner = spinner("Building registry...");
		for (const registryItem of result.data.items) {
			buildSpinner.start(`Building ${registryItem.name}...`);
			registryItem.$schema = "https://shadcn-vue.com/schema/registry-item.json";
			for (const file of registryItem.files ?? []) file.content = (await fs$1.readFile(path$2.resolve(resolvePaths.cwd, file.path), "utf-8")).replace(/\r\n/g, "\n");
			const result$1 = registryItemSchema.safeParse(registryItem);
			if (!result$1.success) {
				logger.error(`Invalid registry item found for ${highlighter.info(registryItem.name)}.`);
				continue;
			}
			await fs$1.writeFile(path$2.resolve(resolvePaths.outputDir, `${result$1.data.name}.json`), JSON.stringify(result$1.data, null, 2));
		}
		await fs$1.copyFile(resolvePaths.registryFile, path$2.resolve(resolvePaths.outputDir, "registry.json"));
		buildSpinner.succeed("Building registry.");
	} catch (error) {
		logger.break();
		handleError(error);
	}
});

//#endregion
//#region src/commands/diff.ts
const updateOptionsSchema = z.object({
	component: z.string().optional(),
	yes: z.boolean(),
	cwd: z.string(),
	path: z.string().optional()
});
const diff = new Command().name("diff").description("check for updates against the registry").argument("[component]", "the component name").option("-y, --yes", "skip confirmation prompt.", false).option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).action(async (name, opts) => {
	try {
		const options = updateOptionsSchema.parse({
			component: name,
			...opts
		});
		const cwd = path.resolve(options.cwd);
		if (!existsSync(cwd)) {
			logger.error(`The path ${cwd} does not exist. Please try again.`);
			process.exit(1);
		}
		const config = await getConfig(cwd);
		if (!config) {
			logger.warn(`Configuration is missing. Please run ${highlighter.success(`init`)} to create a components.json file.`);
			process.exit(1);
		}
		const registryIndex = await getShadcnRegistryIndex();
		if (!registryIndex) {
			handleError(/* @__PURE__ */ new Error("Failed to fetch registry index."));
			process.exit(1);
		}
		if (!options.component) {
			const targetDir = config.resolvedPaths.components;
			const projectComponents = registryIndex.filter((item) => {
				for (const file of item.files ?? []) if (existsSync(path.resolve(targetDir, typeof file === "string" ? file : file.path))) return true;
				return false;
			});
			const componentsWithUpdates = [];
			for (const component$1 of projectComponents) {
				const changes$1 = await diffComponent(component$1, config);
				if (changes$1.length) componentsWithUpdates.push({
					name: component$1.name,
					changes: changes$1
				});
			}
			if (!componentsWithUpdates.length) {
				logger.info("No updates found.");
				process.exit(0);
			}
			logger.info("The following components have updates available:");
			for (const component$1 of componentsWithUpdates) {
				logger.info(`- ${component$1.name}`);
				for (const change of component$1.changes) logger.info(`  - ${change.filePath}`);
			}
			logger.break();
			logger.info(`Run ${highlighter.success(`diff <component>`)} to see the changes.`);
			process.exit(0);
		}
		const component = registryIndex.find((item) => item.name === options.component);
		if (!component) {
			logger.error(`The component ${highlighter.success(options.component)} does not exist.`);
			process.exit(1);
		}
		const changes = await diffComponent(component, config);
		if (!changes.length) {
			logger.info(`No updates found for ${options.component}.`);
			process.exit(0);
		}
		for (const change of changes) {
			logger.info(`- ${change.filePath}`);
			await printDiff(change.patch);
			logger.info("");
		}
	} catch (error) {
		handleError(error);
	}
});
async function diffComponent(component, config) {
	const payload = await fetchTree(config.style, [component]);
	const baseColor = await getRegistryBaseColor(config.tailwind.baseColor);
	if (!payload) return [];
	const changes = [];
	for (const item of payload) {
		const targetDir = await getItemTargetPath(config, item);
		if (!targetDir) continue;
		for (const file of item.files ?? []) {
			const filePath = path.resolve(targetDir, typeof file === "string" ? file : file.path);
			if (!existsSync(filePath)) continue;
			const fileContent = await promises.readFile(filePath, "utf8");
			if (typeof file === "string" || !file.content) continue;
			const patch = diffLines(await transform$1({
				filename: file.path,
				raw: file.content,
				config,
				baseColor
			}), fileContent);
			if (patch.length > 1) changes.push({
				filePath,
				patch
			});
		}
	}
	return changes;
}
async function printDiff(diff$1) {
	diff$1.forEach((part) => {
		if (part) {
			if (part.added) return process.stdout.write(highlighter.success(part.value));
			if (part.removed) return process.stdout.write(highlighter.error(part.value));
			return process.stdout.write(part.value);
		}
	});
}

//#endregion
//#region src/commands/docs.ts
const docs = new Command().name("docs").description("get docs, api references and usage examples for components").argument("<components...>", "component names").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("-b, --base <base>", "the base to use (reka). defaults to project base.").option("--json", "output as JSON.", false).action(async (components, opts) => {
	try {
		const baseFromStyle = (await getConfig(path.resolve(opts.cwd)))?.style?.split("/")?.[0];
		const base = opts.base ?? baseFromStyle ?? "reka";
		const index = await getShadcnRegistryIndex();
		if (!index) {
			logger.error("Failed to fetch the registry index.");
			process.exit(1);
		}
		const results = [];
		for (const component of components) {
			const item = index.find((entry) => entry.name === component);
			if (!item) {
				logger.error(`Component ${highlighter.info(component)} not found in the shadcn-vue registry.`);
				process.exit(1);
			}
			const metaLinks = (item.meta?.links)?.[base];
			let links;
			if (metaLinks && Object.keys(metaLinks).length > 0) links = metaLinks;
			else links = { docs: `${SHADCN_VUE_URL}/docs/components/${component}` };
			results.push({
				component,
				base,
				links
			});
		}
		if (opts.json) {
			consola.log(JSON.stringify({
				base,
				results
			}, null, 2));
			return;
		}
		if (results.length === 0) return;
		const maxKeyLength = Math.max(...results.flatMap((r) => Object.keys(r.links).map((k) => k.length)));
		for (const { component, links } of results) {
			logger.log(highlighter.info(component));
			for (const [key, value] of Object.entries(links)) logger.log(`  - ${key.padEnd(maxKeyLength + 2)}${value}`);
			logger.break();
		}
	} catch (error) {
		handleError(error);
	}
});

//#endregion
//#region src/commands/info.ts
const info = new Command().name("info").description("get information about your project").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("--json", "output as JSON.", false).action(async (opts) => {
	try {
		const projectInfo = await getProjectInfo(opts.cwd);
		const config = await getConfig(opts.cwd);
		if (opts.json) {
			consola.log(JSON.stringify({
				project: projectInfo ?? null,
				config: config ?? null
			}, null, 2));
			return;
		}
		logger.info("> project info");
		consola.log(projectInfo);
		logger.break();
		logger.info("> components.json");
		consola.log(config);
	} catch (error) {
		handleError(error);
	}
});

//#endregion
//#region src/commands/mcp.ts
const SHADCN_MCP_VERSION = "latest";
const CLIENTS = [
	{
		name: "claude",
		label: "Claude Code",
		configPath: ".mcp.json",
		config: { mcpServers: { shadcnVue: {
			command: "npx",
			args: [`shadcn-vue@${SHADCN_MCP_VERSION}`, "mcp"]
		} } }
	},
	{
		name: "cursor",
		label: "Cursor",
		configPath: ".cursor/mcp.json",
		config: { mcpServers: { shadcnVue: {
			command: "npx",
			args: [`shadcn-vue@${SHADCN_MCP_VERSION}`, "mcp"]
		} } }
	},
	{
		name: "vscode",
		label: "VS Code",
		configPath: ".vscode/mcp.json",
		config: { servers: { shadcnVue: {
			command: "npx",
			args: [`shadcn-vue@${SHADCN_MCP_VERSION}`, "mcp"]
		} } }
	},
	{
		name: "codex",
		label: "Codex",
		configPath: ".codex/config.toml",
		config: `[mcp_servers.shadcn_vue]
command = "npx"
args = ["shadcn-vue@${SHADCN_MCP_VERSION}", "mcp"]
`
	},
	{
		name: "opencode",
		label: "Opencode",
		configPath: "opencode.json",
		config: {
			$schema: "https://opencode.ai/config.json",
			mcp: { shadcnVue: {
				type: "local",
				enabled: true,
				command: [
					"npx",
					`shadcn-vue@${SHADCN_MCP_VERSION}`,
					"mcp"
				]
			} }
		}
	}
];
const DEPENDENCIES = [`shadcn-vue@${SHADCN_MCP_VERSION}`];
const mcp = new Command().name("mcp").description("MCP server and configuration commands").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).action(async (options) => {
	try {
		await loadEnvFiles(options.cwd);
		const transport = new StdioServerTransport();
		await server.connect(transport);
	} catch (error) {
		logger.break();
		handleError(error);
	}
});
const mcpInitOptionsSchema = z$1.object({
	client: z$1.enum([
		"claude",
		"cursor",
		"vscode",
		"codex",
		"opencode"
	]),
	cwd: z$1.string()
});
mcp.command("init").description("Initialize MCP configuration for your client").option("--client <client>", `MCP client (${CLIENTS.map((c) => c.name).join(", ")})`).action(async (opts, command) => {
	try {
		const cwd = (command.parent?.opts() || {}).cwd || process.cwd();
		let client = opts.client;
		if (!client) {
			const response = await prompts({
				type: "select",
				name: "client",
				message: "Which MCP client are you using?",
				choices: CLIENTS.map((c) => ({
					title: c.label,
					value: c.name
				}))
			});
			if (!response.client) {
				logger.break();
				process.exit(1);
			}
			client = response.client;
		}
		const options = mcpInitOptionsSchema.parse({
			client,
			cwd
		});
		const config = await getConfig(options.cwd);
		if (options.client === "codex") {
			if (config) await updateDependencies([], DEPENDENCIES, config, { silent: false });
			else {
				const installSpinner = spinner("Installing dependencies...").start();
				await installMcpDependencies(options.cwd);
				installSpinner.succeed("Installing dependencies.");
			}
			logger.break();
			logger.log("To configure the shadcn-vue MCP server in Codex:");
			logger.break();
			logger.log(`1. Open or create the file ${highlighter.info("~/.codex/config.toml")}`);
			logger.log("2. Add the following configuration:");
			logger.log();
			logger.info(`[mcp_servers.shadcn_vue]
command = "npx"
args = ["shadcn-vue@${SHADCN_MCP_VERSION}", "mcp"]`);
			logger.break();
			logger.info("3. Restart Codex to load the MCP server");
			logger.break();
			process.exit(0);
		}
		const configSpinner = spinner("Configuring MCP server...").start();
		const configPath = await runMcpInit(options);
		configSpinner.succeed("Configuring MCP server.");
		if (config) await updateDependencies([], DEPENDENCIES, config, { silent: false });
		else {
			const installSpinner = spinner("Installing dependencies...").start();
			await installMcpDependencies(options.cwd);
			installSpinner.succeed("Installing dependencies.");
		}
		logger.break();
		logger.success(`Configuration saved to ${configPath}.`);
		logger.break();
	} catch (error) {
		handleError(error);
	}
});
const overwriteMerge = (_, sourceArray) => sourceArray;
async function runMcpInit(options) {
	const { client, cwd } = options;
	const clientInfo = CLIENTS.find((c) => c.name === client);
	if (!clientInfo) throw new Error(`Unknown client: ${client}. Available clients: ${CLIENTS.map((c) => c.name).join(", ")}`);
	const configPath = path.join(cwd, clientInfo.configPath);
	const dir = path.dirname(configPath);
	await fsExtra.ensureDir(dir);
	let existingConfig = {};
	try {
		const content = await promises.readFile(configPath, "utf-8");
		existingConfig = JSON.parse(content);
	} catch {}
	const mergedConfig = deepmerge(existingConfig, clientInfo.config, { arrayMerge: overwriteMerge });
	await promises.writeFile(configPath, `${JSON.stringify(mergedConfig, null, 2)}\n`, "utf-8");
	return clientInfo.configPath;
}
async function installMcpDependencies(cwd) {
	await addDevDependency(DEPENDENCIES, {
		cwd,
		packageManager: await getPackageManager(cwd, { withFallback: true }),
		silent: true
	});
}

//#endregion
//#region src/migrations/migrate-icons.ts
async function migrateIcons(config) {
	if (!config.resolvedPaths.ui) throw new Error("We could not find a valid `ui` path in your `components.json` file. Please ensure you have a valid `ui` path in your `components.json` file.");
	const uiPath = config.resolvedPaths.ui;
	const [files, registryIcons] = await Promise.all([glob("**/*.{js,ts,jsx,tsx}", { cwd: uiPath }), getRegistryIcons()]);
	if (Object.keys(registryIcons).length === 0) throw new Error("Something went wrong fetching the registry icons.");
	const libraryChoices = Object.entries(ICON_LIBRARIES).map(([name, iconLibrary]) => ({
		title: iconLibrary.name,
		value: name
	}));
	const migrateOptions = await prompts([{
		type: "select",
		name: "sourceLibrary",
		message: `Which icon library would you like to ${highlighter.info("migrate from")}?`,
		choices: libraryChoices
	}, {
		type: "select",
		name: "targetLibrary",
		message: `Which icon library would you like to ${highlighter.info("migrate to")}?`,
		choices: libraryChoices
	}]);
	if (migrateOptions.sourceLibrary === migrateOptions.targetLibrary) throw new Error("You cannot migrate to the same icon library. Please choose a different icon library.");
	if (!(migrateOptions.sourceLibrary in ICON_LIBRARIES && migrateOptions.targetLibrary in ICON_LIBRARIES)) throw new Error("Invalid icon library. Please choose a valid icon library.");
	const sourceLibrary = ICON_LIBRARIES[migrateOptions.sourceLibrary];
	const targetLibrary = ICON_LIBRARIES[migrateOptions.targetLibrary];
	const { confirm } = await prompts({
		type: "confirm",
		name: "confirm",
		initial: true,
		message: `We will migrate ${highlighter.info(files.length)} files in ${highlighter.info(`./${path.relative(config.resolvedPaths.cwd, uiPath)}`)} from ${highlighter.info(sourceLibrary.name)} to ${highlighter.info(targetLibrary.name)}. Continue?`
	});
	if (!confirm) {
		logger.info("Migration cancelled.");
		process.exit(0);
	}
	if (targetLibrary.package) await updateDependencies([targetLibrary.package], [], config, { silent: false });
	const migrationSpinner = spinner(`Migrating icons...`)?.start();
	await Promise.all(files.map(async (file) => {
		migrationSpinner.text = `Migrating ${file}...`;
		const filePath = path.join(uiPath, file);
		const content = await migrateIconsFile(await promises.readFile(filePath, "utf-8"), migrateOptions.sourceLibrary, migrateOptions.targetLibrary, registryIcons);
		await promises.writeFile(filePath, content);
	}));
	migrationSpinner.succeed("Migration complete.");
}
async function migrateIconsFile(content, sourceLibrary, targetLibrary, iconsMapping) {
	const sourceLibraryImport = ICON_LIBRARIES[sourceLibrary]?.import;
	const targetLibraryImport = ICON_LIBRARIES[targetLibrary]?.import;
	const dir = await promises.mkdtemp(path.join(tmpdir(), "shadcn-"));
	const project = new Project({ compilerOptions: {} });
	const tempFile = path.join(dir, `shadcn-icons-${randomBytes(4).toString("hex")}.tsx`);
	const sourceFile = project.createSourceFile(tempFile, content, { scriptKind: ScriptKind.TSX });
	const targetedIcons = [];
	for (const importDeclaration of sourceFile.getImportDeclarations() ?? []) {
		if (importDeclaration.getModuleSpecifier()?.getText() !== `"${sourceLibraryImport}"`) continue;
		for (const specifier of importDeclaration.getNamedImports() ?? []) {
			const iconName = specifier.getName();
			const targetedIcon = Object.values(iconsMapping).find((icon) => icon[sourceLibrary] === iconName)?.[targetLibrary];
			if (!targetedIcon || targetedIcons.includes(targetedIcon)) continue;
			targetedIcons.push(targetedIcon);
			specifier.remove();
			sourceFile.getDescendantsOfKind(SyntaxKind.JsxSelfClosingElement).filter((node) => node.getTagNameNode()?.getText() === iconName).forEach((node) => node.getTagNameNode()?.replaceWithText(targetedIcon));
		}
		if (importDeclaration.getNamedImports()?.length === 0) importDeclaration.remove();
	}
	if (targetedIcons.length > 0) sourceFile.addImportDeclaration({
		moduleSpecifier: targetLibraryImport,
		namedImports: targetedIcons.map((icon) => ({ name: icon }))
	});
	return await sourceFile.getText();
}

//#endregion
//#region src/migrations/migrate-rtl.ts
const FILES_NEEDING_MANUAL_REVIEW = [
	"Sidebar.vue",
	"Pagination.vue",
	"Calendar.vue"
];
const RTL_DOCS_URL = "https://shadcn-vue.com/docs/rtl";
function migrateRtlFile(content, filename, config) {
	return transform(content, filename, [transformRtl({
		filename,
		raw: content,
		config: {
			...config,
			rtl: true
		}
	})]).code;
}
async function migrateRtl(config, options = {}) {
	let files;
	let basePath;
	if (options.path) {
		basePath = config.resolvedPaths.cwd;
		if (options.path.includes("*")) files = await glob(options.path, {
			cwd: basePath,
			onlyFiles: true,
			ignore: ["**/node_modules/**"]
		});
		else {
			const fullPath = path.resolve(basePath, options.path);
			const stat = await promises.stat(fullPath).catch(() => null);
			if (!stat) throw new Error(`File not found: ${options.path}`);
			if (stat.isDirectory()) {
				basePath = fullPath;
				files = await glob("**/*.{vue,js,ts,jsx,tsx}", {
					cwd: basePath,
					onlyFiles: true,
					ignore: ["**/node_modules/**"]
				});
			} else if (stat.isFile()) files = [options.path];
			else throw new Error(`Unsupported path type: ${options.path}`);
		}
		if (files.length === 0) throw new Error(`No files found matching: ${options.path}`);
	} else {
		if (!config.resolvedPaths.ui) throw new Error("Could not find a valid `ui` path in your `components.json`. Please provide a path or glob pattern.");
		basePath = config.resolvedPaths.ui;
		files = await glob("**/*.{vue,js,ts,jsx,tsx}", {
			cwd: basePath,
			onlyFiles: true
		});
	}
	if (!options.yes) {
		const relativePath = options.path ? options.path : `./${path.relative(config.resolvedPaths.cwd, basePath)}`;
		const { confirm } = await prompts({
			type: "confirm",
			name: "confirm",
			initial: true,
			message: `We will migrate ${highlighter.info(String(files.length))} file(s) in ${highlighter.info(relativePath)} to RTL. Continue?`
		});
		if (!confirm) {
			logger.info("Migration cancelled.");
			process.exit(0);
		}
	}
	const configSpinner = spinner("Updating components.json...").start();
	try {
		const configPath = path.resolve(config.resolvedPaths.cwd, "components.json");
		const existingConfig = JSON.parse(await promises.readFile(configPath, "utf-8"));
		existingConfig.rtl = true;
		await promises.writeFile(configPath, `${JSON.stringify(existingConfig, null, 2)}\n`);
		configSpinner.succeed("Updated components.json.");
	} catch {
		configSpinner.fail("Failed to update components.json.");
		throw new Error("Could not update components.json. Please manually set `rtl: true`.");
	}
	const migrationSpinner = spinner("Migrating files to RTL...").start();
	let transformedCount = 0;
	const filesNeedingReview = [];
	await Promise.all(files.map(async (file) => {
		migrationSpinner.text = `Migrating ${file}...`;
		const filePath = path.join(basePath, file);
		const content = await promises.readFile(filePath, "utf-8");
		const transformed = migrateRtlFile(content, filePath, config);
		if (transformed !== content) {
			await promises.writeFile(filePath, transformed);
			transformedCount++;
		}
		const fileName = path.basename(file);
		if (FILES_NEEDING_MANUAL_REVIEW.includes(fileName)) filesNeedingReview.push(file);
	}));
	migrationSpinner.succeed(`Migration complete. ${transformedCount} file(s) transformed.`);
	if (filesNeedingReview.length > 0) {
		logger.break();
		logger.warn("The following components may need manual RTL adjustments:");
		for (const file of filesNeedingReview) logger.info(`  - ${file}`);
		logger.break();
		logger.info(`See ${highlighter.info(RTL_DOCS_URL)} for more information.`);
	}
}

//#endregion
//#region src/preflights/preflight-migrate.ts
async function preFlightMigrate(options) {
	const errors = {};
	if (!fsExtra.existsSync(options.cwd) || !fsExtra.existsSync(path.resolve(options.cwd, "package.json"))) {
		errors[MISSING_DIR_OR_EMPTY_PROJECT] = true;
		return {
			errors,
			config: null
		};
	}
	if (!fsExtra.existsSync(path.resolve(options.cwd, "components.json"))) {
		errors[MISSING_CONFIG] = true;
		return {
			errors,
			config: null
		};
	}
	try {
		return {
			errors,
			config: await getConfig(options.cwd)
		};
	} catch (error) {
		logger.break();
		logger.error(`An invalid ${highlighter.info("components.json")} file was found at ${highlighter.info(options.cwd)}.\nBefore you can run a migration, you must create a valid ${highlighter.info("components.json")} file by running the ${highlighter.info("init")} command.`);
		logger.error(`Learn more at ${highlighter.info("https://shadcn-vue.com/docs/components-json")}.`);
		logger.break();
		process.exit(1);
	}
}

//#endregion
//#region src/commands/migrate.ts
const migrations = [{
	name: "icons",
	description: "migrate your ui components to a different icon library."
}, {
	name: "rtl",
	description: "migrate your components to support RTL (right-to-left)."
}];
const migrateOptionsSchema = z.object({
	cwd: z.string(),
	list: z.boolean(),
	yes: z.boolean(),
	migration: z.string().refine((value) => value && migrations.some((migration) => migration.name === value), { message: "You must specify a valid migration. Run `shadcn migrate --list` to see available migrations." }).optional(),
	path: z.string().optional()
});
const migrate = new Command().name("migrate").description("run a migration.").argument("[migration]", "the migration to run.").argument("[path]", "optional path or glob pattern to migrate.").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("-l, --list", "list all migrations.", false).option("-y, --yes", "skip confirmation prompt.", false).action(async (migration, migratePath, opts) => {
	try {
		const options = migrateOptionsSchema.parse({
			cwd: path.resolve(opts.cwd),
			migration,
			path: migratePath,
			list: opts.list,
			yes: opts.yes
		});
		if (options.list || !options.migration) {
			logger.info("Available migrations:");
			for (const migration$1 of migrations) logger.info(`- ${migration$1.name}: ${migration$1.description}`);
			return;
		}
		if (!options.migration) throw new Error("You must specify a migration. Run `shadcn migrate --list` to see available migrations.");
		const { errors, config } = await preFlightMigrate(options);
		if (errors[MISSING_DIR_OR_EMPTY_PROJECT] || errors[MISSING_CONFIG]) throw new Error("No `components.json` file found. Ensure you are at the root of your project.");
		if (!config) throw new Error("Something went wrong reading your `components.json` file. Please ensure you have a valid `components.json` file.");
		if (options.migration === "icons") await migrateIcons(config);
		if (options.migration === "rtl") await migrateRtl(config, {
			yes: options.yes,
			path: options.path
		});
	} catch (error) {
		logger.break();
		handleError(error);
	}
});

//#endregion
//#region src/commands/search.ts
const searchOptionsSchema = z.object({
	cwd: z.string(),
	query: z.string().optional(),
	limit: z.number().optional(),
	offset: z.number().optional()
});
const search = new Command().name("search").alias("list").description("search items from registries").argument("<registries...>", "the registry names or urls to search items from. Names must be prefixed with @.").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).option("-q, --query <query>", "query string").option("-l, --limit <number>", "maximum number of items to display per registry", "100").option("-o, --offset <number>", "number of items to skip", "0").action(async (registries, opts) => {
	try {
		const options = searchOptionsSchema.parse({
			cwd: path.resolve(opts.cwd),
			query: opts.query,
			limit: opts.limit ? Number.parseInt(opts.limit, 10) : void 0,
			offset: opts.offset ? Number.parseInt(opts.offset, 10) : void 0
		});
		await loadEnvFiles(options.cwd);
		const defaultConfig = createConfig({
			style: "new-york",
			resolvedPaths: { cwd: options.cwd }
		});
		let shadowConfig = configWithDefaults(defaultConfig);
		const componentsJsonPath = path.resolve(options.cwd, "components.json");
		if (fsExtra.existsSync(componentsJsonPath)) {
			const existingConfig = await fsExtra.readJson(componentsJsonPath);
			const partialConfig = rawConfigSchema.partial().parse(existingConfig);
			shadowConfig = configWithDefaults({
				...defaultConfig,
				...partialConfig
			});
		}
		let config = shadowConfig;
		try {
			const fullConfig = await getConfig(options.cwd);
			if (fullConfig) config = configWithDefaults(fullConfig);
		} catch {}
		const { config: updatedConfig, newRegistries } = await ensureRegistriesInConfig(registries.map((registry) => `${registry}/registry`), config, {
			silent: true,
			writeFile: false
		});
		if (newRegistries.length > 0) config.registries = updatedConfig.registries;
		validateRegistryConfigForItems(registries, config);
		const results = await searchRegistries(registries, {
			query: options.query,
			limit: options.limit,
			offset: options.offset,
			config
		});
		console.log(JSON.stringify(results, null, 2));
		process.exit(0);
	} catch (error) {
		handleError(error);
	} finally {
		clearRegistryContext();
	}
});

//#endregion
//#region src/commands/view.ts
const viewOptionsSchema = z.object({ cwd: z.string() });
const view = new Command().name("view").description("view items from the registry").argument("<items...>", "the item names or URLs to view").option("-c, --cwd <cwd>", "the working directory. defaults to the current directory.", process.cwd()).action(async (items, opts) => {
	try {
		const options = viewOptionsSchema.parse({ cwd: path.resolve(opts.cwd) });
		await loadEnvFiles(options.cwd);
		let shadowConfig = configWithDefaults({});
		const componentsJsonPath = path.resolve(options.cwd, "components.json");
		if (fsExtra.existsSync(componentsJsonPath)) {
			const existingConfig = await fsExtra.readJson(componentsJsonPath);
			shadowConfig = configWithDefaults(rawConfigSchema.partial().parse(existingConfig));
		}
		let config = shadowConfig;
		try {
			const fullConfig = await getConfig(options.cwd);
			if (fullConfig) config = configWithDefaults(fullConfig);
		} catch {}
		const { config: updatedConfig, newRegistries } = await ensureRegistriesInConfig(items, config, {
			silent: true,
			writeFile: false
		});
		if (newRegistries.length > 0) config.registries = updatedConfig.registries;
		validateRegistryConfigForItems(items, config);
		const payload = await getRegistryItems(items, { config });
		consola.log(JSON.stringify(payload, null, 2));
		process.exit(0);
	} catch (error) {
		handleError(error);
	} finally {
		clearRegistryContext();
	}
});

//#endregion
//#region package.json
var version = "2.7.4";

//#endregion
//#region src/index.ts
process.on("SIGINT", () => process.exit(0));
process.on("SIGTERM", () => process.exit(0));
async function main() {
	const program = new Command().name("shadcn-vue").description("add components and dependencies to your project").version(version || "1.0.0", "-v, --version", "display the version number");
	program.addCommand(init).addCommand(add).addCommand(apply).addCommand(diff).addCommand(docs).addCommand(view).addCommand(search).addCommand(migrate).addCommand(info).addCommand(build).addCommand(mcp);
	program.parse();
}
main();

//#endregion
export { fetchTree, getItemTargetPath, getRegistriesConfig, getRegistriesIndex, getRegistry, getRegistryBase, getRegistryBaseColor, getRegistryBaseColors, getRegistryBases, getRegistryFont, getRegistryFonts, getRegistryIconLibraries, getRegistryIconLibrary, getRegistryIcons, getRegistryItems, getRegistryPreset, getRegistryPresets, getRegistryStyles, getRegistryVisualStyle, getRegistryVisualStyles, getShadcnRegistryIndex, resolveRegistryItems, resolveTree };
