// shadcn-vue@2.7.4/dist/registry-BZ439r4u.js
import { a as BASES, d as ICON_LIBRARIES, f as REGISTRY_URL, l as FALLBACK_STYLE, m as STYLES, n as transformIcons, o as BASE_COLORS, p as SHADCN_VUE_URL, s as BUILTIN_REGISTRIES, t as transformMenu, u as FONTS } from "./transform-menu-ZTp9ez-j.js";
import { C as searchResultsSchema, T as workspaceConfigSchema, _ as registryItemSchema, a as rawConfigSchema, b as registryResolvedItemsTreeSchema, h as registryItemFileSchema, l as registryConfigSchema, n as configSchema, o as registriesIndexSchema, r as iconsSchema, s as registryBaseColorSchema, u as registryIndexSchema, w as stylesSchema, x as registrySchema } from "./schema-CjAXtlQ7.js";
import path, { basename } from "pathe";
import prompts from "prompts";
import { z } from "zod";
import fs, { existsSync, promises, statSync } from "fs";
import deepmerge from "deepmerge";
import fsExtra from "fs-extra";
import { createPathsMatcher, getTsconfig } from "get-tsconfig";
import { coerce } from "semver";
import { glob } from "tinyglobby";
import open from "open";
import { loadConfig } from "c12";
import { colors } from "consola/utils";
import { homedir, tmpdir } from "os";
import { Project, QuoteKind, ScriptKind, SyntaxKind } from "ts-morph";
import { transform } from "vue-metamorph";
import consola from "consola";
import ora from "ora";
import { transform as transform$1 } from "@unovue/detypes";
import { createHash } from "crypto";
import { ofetch } from "ofetch";
import { ProxyAgent } from "undici";
import objectToString from "stringify-object";
import path$1 from "path";
import { addDependency, addDevDependency, detectPackageManager } from "nypm";
import fuzzysort from "fuzzysort";

//#region src/utils/highlighter.ts
const highlighter = {
	error: colors.red,
	warn: colors.yellow,
	info: colors.cyan,
	success: colors.green
};

//#endregion
//#region src/registry/errors.ts
const RegistryErrorCode = {
	NETWORK_ERROR: "NETWORK_ERROR",
	NOT_FOUND: "NOT_FOUND",
	UNAUTHORIZED: "UNAUTHORIZED",
	FORBIDDEN: "FORBIDDEN",
	FETCH_ERROR: "FETCH_ERROR",
	NOT_CONFIGURED: "NOT_CONFIGURED",
	INVALID_CONFIG: "INVALID_CONFIG",
	MISSING_ENV_VARS: "MISSING_ENV_VARS",
	LOCAL_FILE_ERROR: "LOCAL_FILE_ERROR",
	PARSE_ERROR: "PARSE_ERROR",
	VALIDATION_ERROR: "VALIDATION_ERROR",
	UNKNOWN_ERROR: "UNKNOWN_ERROR"
};
var RegistryError = class extends Error {
	constructor(message, options = {}) {
		super(message);
		this.name = "RegistryError";
		this.code = options.code || RegistryErrorCode.UNKNOWN_ERROR;
		this.statusCode = options.statusCode;
		this.cause = options.cause;
		this.context = options.context;
		this.suggestion = options.suggestion;
		this.timestamp = /* @__PURE__ */ new Date();
		if (Error.captureStackTrace) Error.captureStackTrace(this, this.constructor);
	}
	toJSON() {
		return {
			name: this.name,
			message: this.message,
			code: this.code,
			statusCode: this.statusCode,
			context: this.context,
			suggestion: this.suggestion,
			timestamp: this.timestamp,
			stack: this.stack
		};
	}
};
var RegistryNotFoundError = class extends RegistryError {
	constructor(url, cause) {
		const message = `The item at ${url} was not found. It may not exist at the registry.`;
		super(message, {
			code: RegistryErrorCode.NOT_FOUND,
			statusCode: 404,
			cause,
			context: { url },
			suggestion: "Check if the item name is correct and the registry URL is accessible."
		});
		this.url = url;
		this.name = "RegistryNotFoundError";
	}
};
var RegistryUnauthorizedError = class extends RegistryError {
	constructor(url, cause) {
		const message = `You are not authorized to access the item at ${url}. If this is a remote registry, you may need to authenticate.`;
		super(message, {
			code: RegistryErrorCode.UNAUTHORIZED,
			statusCode: 401,
			cause,
			context: { url },
			suggestion: "Check your authentication credentials and environment variables."
		});
		this.url = url;
		this.name = "RegistryUnauthorizedError";
	}
};
var RegistryForbiddenError = class extends RegistryError {
	constructor(url, cause) {
		const message = `You are not authorized to access the item at ${url}. If this is a remote registry, you may need to authenticate.`;
		super(message, {
			code: RegistryErrorCode.FORBIDDEN,
			statusCode: 403,
			cause,
			context: { url },
			suggestion: "Check your authentication credentials and environment variables."
		});
		this.url = url;
		this.name = "RegistryForbiddenError";
	}
};
var RegistryFetchError = class extends RegistryError {
	constructor(url, statusCode, responseBody, cause) {
		const baseMessage = statusCode ? `Failed to fetch from registry (${statusCode}): ${url}` : `Failed to fetch from registry: ${url}`;
		const message = typeof cause === "string" && cause ? `${baseMessage} - ${cause}` : baseMessage;
		let suggestion = "Check your network connection and try again.";
		if (statusCode === 404) suggestion = "The requested resource was not found. Check the URL or item name.";
		else if (statusCode === 500) suggestion = "The registry server encountered an error. Try again later.";
		else if (statusCode && statusCode >= 400 && statusCode < 500) suggestion = "There was a client error. Check your request parameters.";
		super(message, {
			code: RegistryErrorCode.FETCH_ERROR,
			statusCode,
			cause,
			context: {
				url,
				responseBody
			},
			suggestion
		});
		this.url = url;
		this.responseBody = responseBody;
		this.name = "RegistryFetchError";
	}
};
var RegistryNotConfiguredError = class extends RegistryError {
	constructor(registryName) {
		const message = registryName ? `Unknown registry "${registryName}". Make sure it is defined in components.json as follows:
{
  "registries": {
    "${registryName}": "[URL_TO_REGISTRY]"
  }
}` : "Unknown registry. Make sure it is defined in components.json under \"registries\".";
		super(message, {
			code: RegistryErrorCode.NOT_CONFIGURED,
			context: { registryName },
			suggestion: "Add the registry configuration to your components.json file. Consult the registry documentation for the correct format."
		});
		this.registryName = registryName;
		this.name = "RegistryNotConfiguredError";
	}
};
var RegistryLocalFileError = class extends RegistryError {
	constructor(filePath, cause) {
		super(`Failed to read local registry file: ${filePath}`, {
			code: RegistryErrorCode.LOCAL_FILE_ERROR,
			cause,
			context: { filePath },
			suggestion: "Check if the file exists and you have read permissions."
		});
		this.filePath = filePath;
		this.name = "RegistryLocalFileError";
	}
};
var RegistryParseError = class extends RegistryError {
	constructor(item, parseError) {
		let message = `Failed to parse registry item: ${item}`;
		if (parseError instanceof z.ZodError) message = `Failed to parse registry item: ${item}\n${parseError.errors.map((e) => `  - ${e.path.join(".")}: ${e.message}`).join("\n")}`;
		super(message, {
			code: RegistryErrorCode.PARSE_ERROR,
			cause: parseError,
			context: { item },
			suggestion: "The registry item may be corrupted or have an invalid format. Please make sure it returns a valid JSON object. See https://shadcn-vue.com/schema/registry-item.json."
		});
		this.item = item;
		this.parseError = parseError;
		this.name = "RegistryParseError";
	}
};
var RegistryMissingEnvironmentVariablesError = class extends RegistryError {
	constructor(registryName, missingVars) {
		const message = `Registry "${registryName}" requires the following environment variables:\n\n${missingVars.map((v) => `  • ${v}`).join("\n")}`;
		super(message, {
			code: RegistryErrorCode.MISSING_ENV_VARS,
			context: {
				registryName,
				missingVars
			},
			suggestion: "Set the required environment variables to your .env or .env.local file."
		});
		this.registryName = registryName;
		this.missingVars = missingVars;
		this.name = "RegistryMissingEnvironmentVariablesError";
	}
};
var RegistryInvalidNamespaceError = class extends RegistryError {
	constructor(name) {
		const message = `Invalid registry namespace: "${name}". Registry names must start with @ (e.g., @shadcn, @v0).`;
		super(message, {
			code: RegistryErrorCode.VALIDATION_ERROR,
			context: { name },
			suggestion: "Use a valid registry name starting with @ or provide a direct URL to the registry."
		});
		this.name = name;
		this.name = "RegistryInvalidNamespaceError";
	}
};
var ConfigParseError = class extends RegistryError {
	constructor(cwd, parseError) {
		let message = `Invalid components.json configuration in ${cwd}.`;
		if (parseError instanceof Error && parseError.message.includes("built-in registry and cannot be overridden")) message = `Invalid components.json configuration in ${highlighter.info(`${cwd}/components.json`)}:\n  - ${parseError.message}`;
		if (parseError instanceof SyntaxError) message = `Invalid components.json configuration in ${highlighter.info(`${cwd}/components.json`)}:\n  - Syntax error: ${parseError.message.replace(`${cwd}/components.json`, "")}`;
		if (parseError instanceof z.ZodError) message = `Invalid components.json configuration in ${highlighter.info(`${cwd}/components.json`)}:\n${parseError.errors.map((e) => `  - ${e.path.join(".")}: ${e.message}`).join("\n")}`;
		super(message, {
			code: RegistryErrorCode.INVALID_CONFIG,
			cause: parseError,
			context: { cwd },
			suggestion: "Check your components.json file for syntax errors or invalid configuration. Run 'npx shadcn@latest init' to regenerate a valid configuration."
		});
		this.cwd = cwd;
		this.name = "ConfigParseError";
	}
};
var RegistriesIndexParseError = class extends RegistryError {
	constructor(parseError) {
		let message = "Failed to parse registries index";
		if (parseError instanceof z.ZodError) {
			const invalidNamespaces = parseError.errors.filter((e) => e.path.length > 0).map((e) => `"${e.path[0]}"`).filter((v, i, arr) => arr.indexOf(v) === i);
			if (invalidNamespaces.length > 0) message = `Failed to parse registries index. Invalid registry namespace(s): ${invalidNamespaces.join(", ")}\n${parseError.errors.map((e) => `  - ${e.path.join(".")}: ${e.message}`).join("\n")}`;
			else message = `Failed to parse registries index:\n${parseError.errors.map((e) => `  - ${e.path.join(".")}: ${e.message}`).join("\n")}`;
		}
		super(message, {
			code: RegistryErrorCode.PARSE_ERROR,
			cause: parseError,
			context: { parseError },
			suggestion: "The registries index may be corrupted or have invalid registry namespace format. Registry names must start with @ (e.g., @shadcn, @example)."
		});
		this.parseError = parseError;
		this.name = "RegistriesIndexParseError";
	}
};

//#endregion
//#region src/utils/resolve-import.ts
function resolveImport(importPath, config) {
	const matcher = createPathsMatcher(config);
	if (matcher === null) return;
	const paths = matcher(importPath);
	if (paths[0]) return paths[0];
	const tsconfigPaths = config.config.compilerOptions?.paths ?? {};
	const baseUrl = config.config.compilerOptions?.baseUrl ?? ".";
	const configDir = path.dirname(config.path);
	const resolvedBaseUrl = path.resolve(configDir, baseUrl);
	for (const [pattern, mappings] of Object.entries(tsconfigPaths)) if (pattern.replace(/\/\*$/, "") === importPath && Array.isArray(mappings) && mappings.length > 0) {
		const mapping = String(mappings[0]).replace(/\/\*$/, "");
		return path.resolve(resolvedBaseUrl, mapping);
	}
}

//#endregion
//#region src/utils/get-config.ts
const DEFAULT_COMPONENTS = "@/components";
const DEFAULT_UTILS = "@/lib/utils";
const DEFAULT_TAILWIND_CSS = "assets/css/tailwind.css";
const DEFAULT_TAILWIND_CONFIG = "tailwind.config.js";
async function getConfig(cwd) {
	const config = await getRawConfig(cwd);
	if (!config) return null;
	if (!config.iconLibrary) config.iconLibrary = config.style === "new-york" ? "radix" : "lucide";
	return await resolveConfigPaths(cwd, config);
}
async function resolveConfigPaths(cwd, config) {
	config.registries = {
		...BUILTIN_REGISTRIES,
		...config.registries || {}
	};
	const detectedFramework = await detectFrameworkConfigFiles(cwd);
	const isTypeScript = await isTypeScriptProject(cwd);
	const tsConfig = await getTsconfig(path.resolve(cwd, detectedFramework?.name === "nuxt4" ? "./.nuxt/tsconfig.app.json" : detectedFramework?.name === "nuxt3" ? "./.nuxt/tsconfig.json" : detectedFramework?.name === "inertia" ? "./inertia/tsconfig.json" : isTypeScript ? "./tsconfig.json" : "./jsconfig.json"), isTypeScript ? void 0 : "jsconfig.json");
	if (tsConfig === null) throw new Error(`Failed to load ${config.typescript ? "tsconfig" : "jsconfig"}.json.`.trim());
	return configSchema.parse({
		...config,
		resolvedPaths: {
			cwd,
			tailwindConfig: config.tailwind.config ? path.resolve(cwd, config.tailwind.config) : "",
			tailwindCss: path.resolve(cwd, config.tailwind.css),
			utils: await resolveImport(config.aliases.utils, tsConfig),
			components: await resolveImport(config.aliases.components, tsConfig),
			ui: config.aliases.ui ? await resolveImport(config.aliases.ui, tsConfig) : path.resolve(await resolveImport(config.aliases.components, tsConfig) ?? cwd, "ui"),
			lib: config.aliases.lib ? await resolveImport(config.aliases.lib, tsConfig) : path.resolve(await resolveImport(config.aliases.utils, tsConfig) ?? cwd, ".."),
			hooks: config.aliases.hooks ? await resolveImport(config.aliases.hooks, tsConfig) : path.resolve(await resolveImport(config.aliases.components, tsConfig) ?? cwd, "..", "hooks"),
			composables: config.aliases.composables ? await resolveImport(config.aliases.composables, tsConfig) : path.resolve(await resolveImport(config.aliases.components, tsConfig) ?? cwd, "..", "composables")
		}
	});
}
async function getRawConfig(cwd) {
	try {
		const configResult = await loadConfig({
			name: "components",
			configFile: "components",
			cwd,
			dotenv: false,
			packageJson: false,
			rcFile: false,
			jitiOptions: {
				rebuildFsCache: true,
				moduleCache: true
			}
		});
		if (!configResult.config || Object.keys(configResult.config).length === 0) return null;
		const config = rawConfigSchema.parse(configResult.config);
		if (config.registries) {
			for (const registryName of Object.keys(config.registries)) if (registryName in BUILTIN_REGISTRIES) throw new Error(`"${registryName}" is a built-in registry and cannot be overridden.`);
		}
		return config;
	} catch (error) {
		throw new ConfigParseError(cwd, error);
	}
}
async function getWorkspaceConfig(config) {
	let resolvedAliases = {};
	for (const key of Object.keys(config.aliases)) {
		if (!isAliasKey(key, config)) continue;
		const resolvedPath = config.resolvedPaths[key];
		const packageRoot = await findPackageRoot(config.resolvedPaths.cwd, resolvedPath);
		if (!packageRoot) {
			resolvedAliases[key] = config;
			continue;
		}
		resolvedAliases[key] = await getConfig(packageRoot);
	}
	const result = workspaceConfigSchema.safeParse(resolvedAliases);
	if (!result.success) return null;
	return result.data;
}
async function findPackageRoot(cwd, resolvedPath) {
	const commonRoot = findCommonRoot$1(cwd, resolvedPath);
	const relativePath = path.relative(commonRoot, resolvedPath);
	const matchingPackageRoot = (await glob("**/package.json", {
		cwd: commonRoot,
		deep: 3,
		ignore: [
			"**/node_modules/**",
			"**/dist/**",
			"**/build/**",
			"**/public/**"
		]
	})).map((pkgPath) => path.dirname(pkgPath)).find((pkgDir) => relativePath.startsWith(pkgDir));
	return matchingPackageRoot ? path.join(commonRoot, matchingPackageRoot) : null;
}
function isAliasKey(key, config) {
	return Object.keys(config.resolvedPaths).filter((key$1) => key$1 !== "utils").includes(key);
}
function findCommonRoot$1(cwd, resolvedPath) {
	const parts1 = cwd.split(path.sep);
	const parts2 = resolvedPath.split(path.sep);
	const commonParts = [];
	for (let i = 0; i < Math.min(parts1.length, parts2.length); i++) {
		if (parts1[i] !== parts2[i]) break;
		commonParts.push(parts1[i]);
	}
	return commonParts.join(path.sep);
}
async function getTargetStyleFromConfig(cwd, fallback) {
	if ((await getProjectInfo(cwd))?.tailwindVersion === "v4" && fallback === "new-york") return "new-york-v4";
	return fallback;
}
/**
* Creates a config object with sensible defaults.
* Useful for universal registry items that bypass framework detection.
*
* @param partial - Partial config values to override defaults
* @returns A complete Config object
*/
function createConfig(partial) {
	const defaultConfig = {
		typescript: true,
		resolvedPaths: {
			cwd: process.cwd(),
			tailwindConfig: "",
			tailwindCss: "",
			utils: "",
			components: "",
			ui: "",
			lib: "",
			hooks: "",
			composables: ""
		},
		style: "",
		font: "inter",
		tailwind: {
			config: "",
			css: "",
			baseColor: "",
			cssVariables: false
		},
		aliases: {
			components: "",
			utils: ""
		},
		registries: { ...BUILTIN_REGISTRIES }
	};
	if (partial) return {
		...defaultConfig,
		...partial,
		resolvedPaths: {
			...defaultConfig.resolvedPaths,
			...partial.resolvedPaths || {}
		},
		tailwind: {
			...defaultConfig.tailwind,
			...partial.tailwind || {}
		},
		aliases: {
			...defaultConfig.aliases,
			...partial.aliases || {}
		},
		registries: {
			...defaultConfig.registries,
			...partial.registries || {}
		}
	};
	return defaultConfig;
}

//#endregion
//#region src/registry/config.ts
const VISUAL_STYLES = new Set([
	"vega",
	"nova",
	"maia",
	"lyra",
	"mira",
	"luma",
	"sera"
]);
/**
* Composes a base + visual style choice into the canonical style identifier
* stored in `components.json`. For visual styles, returns `${base}-${style}`
* (e.g. `"reka-luma"`). For non-visual styles like `"new-york-v4"`, returns
* the style unchanged.
*/
function composeStyleId(base, style) {
	if (!style) return FALLBACK_STYLE;
	if (style.includes("-") || !VISUAL_STYLES.has(style)) return style;
	return `${base || "reka"}-${style}`;
}
/**
* Resolves the registry style segment used in fetch URLs like
* `styles/{registryStyle}/<comp>.json`.
*
* The `style` field in `components.json` is the **full** style identifier
* (e.g. `"reka-luma"`, `"new-york-v4"`), matching shadcn-ui's model. We just
* pass it through, falling back to `FALLBACK_STYLE` when unset.
*/
function resolveRegistryStyle(style) {
	return style || FALLBACK_STYLE;
}
function resolveStyleFromConfig(config) {
	if (!config.style) return FALLBACK_STYLE;
	if (config.style === "new-york" && config.tailwind?.config === "") return FALLBACK_STYLE;
	return config.style;
}
function configWithDefaults(config) {
	const baseConfig = createConfig({
		style: FALLBACK_STYLE,
		registries: BUILTIN_REGISTRIES
	});
	if (!config) return baseConfig;
	return configSchema.parse(deepmerge(baseConfig, {
		...config,
		style: resolveStyleFromConfig(config),
		registries: {
			...BUILTIN_REGISTRIES,
			...config.registries
		}
	}));
}

//#endregion
//#region src/registry/env.ts
function expandEnvVars(value) {
	return value.replace(/\$\{(\w+)\}/g, (_match, key) => process.env[key] || "");
}
function extractEnvVars(value) {
	const vars = [];
	const regex = /\$\{(\w+)\}/g;
	let match;
	while ((match = regex.exec(value)) !== null) vars.push(match[1]);
	return vars;
}

//#endregion
//#region src/registry/parser.ts
const REGISTRY_PATTERN = /^(@[a-z0-9](?:[\w-]*[a-z0-9])?)\/(.+)$/i;
function parseRegistryAndItemFromString(name) {
	if (!name.startsWith("@")) return {
		registry: null,
		item: name
	};
	const match = name.match(REGISTRY_PATTERN);
	if (match) return {
		registry: match[1],
		item: match[2]
	};
	return {
		registry: null,
		item: name
	};
}

//#endregion
//#region src/utils/compare.ts
function isContentSame(existingContent, newContent, options = {}) {
	const { ignoreImports = false } = options;
	const normalizedExisting = existingContent.replace(/\r\n/g, "\n").trim();
	const normalizedNew = newContent.replace(/\r\n/g, "\n").trim();
	if (normalizedExisting === normalizedNew) return true;
	if (!ignoreImports) return false;
	const importRegex = /^(import\s+(?:type\s+)?(?:\*\s+as\s+\w+|\{[^}]*\}|\w+)?(?:\s*,\s*(?:\{[^}]*\}|\w+))?\s+from\s+["'])([^"']+)(["'])/gm;
	const normalizeImports = (content) => {
		return content.replace(importRegex, (_match, prefix, importPath, suffix) => {
			if (importPath.startsWith(".")) return `${prefix}${importPath}${suffix}`;
			const parts = importPath.split("/");
			return `${prefix}@normalized/${parts[parts.length - 1]}${suffix}`;
		});
	};
	return normalizeImports(normalizedExisting) === normalizeImports(normalizedNew);
}

//#endregion
//#region src/utils/env-helpers.ts
function isEnvFile(filePath) {
	const fileName = path.basename(filePath);
	return /^\.env(?:\.|$)/.test(fileName);
}
/**
* Finds a file variant in the project.
* TODO: abstract this to a more generic function.
*/
function findExistingEnvFile(targetDir) {
	for (const variant of [
		".env.local",
		".env",
		".env.development.local",
		".env.development"
	]) {
		const filePath = path.join(targetDir, variant);
		if (existsSync(filePath)) return filePath;
	}
	return null;
}
/**
* Parse .env content into key-value pairs.
*/
function parseEnvContent(content) {
	const lines = content.split("\n");
	const env = {};
	for (const line of lines) {
		const trimmed = line.trim();
		if (!trimmed || trimmed.startsWith("#")) continue;
		const equalIndex = trimmed.indexOf("=");
		if (equalIndex === -1) continue;
		const key = trimmed.substring(0, equalIndex).trim();
		const value = trimmed.substring(equalIndex + 1).trim();
		if (key) env[key] = value.replace(/^["']|["']$/g, "");
	}
	return env;
}
/**
* Get the list of new keys that would be added when merging env content.
*/
function getNewEnvKeys(existingContent, newContent) {
	const existingEnv = parseEnvContent(existingContent);
	const newEnv = parseEnvContent(newContent);
	const newKeys = [];
	for (const key of Object.keys(newEnv)) if (!(key in existingEnv)) newKeys.push(key);
	return newKeys;
}
/**
* Merge env content by appending ONLY new keys that don't exist in the existing content.
* Existing keys are preserved with their original values.
*/
function mergeEnvContent(existingContent, newContent) {
	const existingEnv = parseEnvContent(existingContent);
	const newEnv = parseEnvContent(newContent);
	let result = existingContent.trimEnd();
	if (result && !result.endsWith("\n")) result += "\n";
	const newKeys = [];
	for (const [key, value] of Object.entries(newEnv)) if (!(key in existingEnv)) newKeys.push(`${key}=${value}`);
	if (newKeys.length > 0) {
		if (result) result += "\n";
		result += newKeys.join("\n");
		return `${result}\n`;
	}
	if (result && !result.endsWith("\n")) return `${result}\n`;
	return result;
}

//#endregion
//#region src/utils/logger.ts
const logger = {
	error(...args) {
		consola.log(highlighter.error(args.join(" ")));
	},
	warn(...args) {
		consola.log(highlighter.warn(args.join(" ")));
	},
	info(...args) {
		consola.log(highlighter.info(args.join(" ")));
	},
	success(...args) {
		consola.log(highlighter.success(args.join(" ")));
	},
	log(...args) {
		consola.log(args.join(" "));
	},
	break() {
		consola.log("");
	}
};

//#endregion
//#region src/utils/spinner.ts
function spinner(text, options) {
	return ora({
		text,
		isSilent: options?.silent
	});
}

//#endregion
//#region src/utils/transformers/transform-css-vars.ts
function transformCssVars(opts) {
	return {
		type: "codemod",
		name: "add prefix to tailwind classes",
		transform({ scriptASTs, sfcAST, utils: { traverseScriptAST, traverseTemplateAST } }) {
			let transformCount = 0;
			const { baseColor, config } = opts;
			if (config.tailwind?.cssVariables || !baseColor?.inlineColors) return transformCount;
			for (const scriptAST of scriptASTs) traverseScriptAST(scriptAST, { visitLiteral(path$2) {
				if (path$2.parent.value.type !== "ImportDeclaration" && typeof path$2.node.value === "string") {
					const raw = path$2.node.value;
					const mapped = applyColorMapping(raw, baseColor.inlineColors).trim();
					if (mapped !== raw) {
						path$2.node.value = mapped;
						transformCount++;
					}
				}
				return this.traverse(path$2);
			} });
			if (sfcAST) traverseTemplateAST(sfcAST, {
				enterNode(node) {
					if (node.type === "Literal" && typeof node.value === "string") {
						if (!["BinaryExpression", "Property"].includes(node.parent?.type ?? "")) {
							const raw = node.value;
							const mapped = applyColorMapping(raw, baseColor.inlineColors).trim();
							if (mapped !== raw) {
								node.value = mapped;
								transformCount++;
							}
						}
					} else if (node.type === "VLiteral" && typeof node.value === "string") {
						if (node.parent.key.name === "class") {
							const raw = node.value;
							const mapped = applyColorMapping(raw, baseColor.inlineColors).trim();
							if (mapped !== raw) {
								node.value = mapped;
								transformCount++;
							}
						}
					}
				},
				leaveNode() {}
			});
			return transformCount;
		}
	};
}
function splitClassName(className) {
	if (!className.includes("/") && !className.includes(":")) return [
		null,
		className,
		null
	];
	const parts = [];
	const [rest, alpha] = className.split("/");
	if (!rest.includes(":")) return [
		null,
		rest,
		alpha
	];
	const split = rest.split(":");
	const name = split.pop();
	const variant = split.join(":");
	parts.push(variant ?? null, name ?? null, alpha ?? null);
	return parts;
}
const PREFIXES = [
	"bg-",
	"text-",
	"border-",
	"ring-offset-",
	"ring-"
];
function applyColorMapping(input, mapping) {
	if (input.includes(" border ")) input = input.replace(" border ", " border border-border ");
	const classNames = input.split(" ");
	const lightMode = /* @__PURE__ */ new Set();
	const darkMode = /* @__PURE__ */ new Set();
	for (const className of classNames) {
		const [variant, value, modifier] = splitClassName(className);
		const prefix = PREFIXES.find((prefix$1) => value?.startsWith(prefix$1));
		if (!prefix) {
			if (!lightMode.has(className)) lightMode.add(className);
			continue;
		}
		const needle = value?.replace(prefix, "");
		if (needle && needle in mapping.light) {
			lightMode.add([variant, `${prefix}${mapping.light[needle]}`].filter(Boolean).join(":") + (modifier ? `/${modifier}` : ""));
			darkMode.add([
				"dark",
				variant,
				`${prefix}${mapping.dark[needle]}`
			].filter(Boolean).join(":") + (modifier ? `/${modifier}` : ""));
			continue;
		}
		if (!lightMode.has(className)) lightMode.add(className);
	}
	return [...Array.from(lightMode), ...Array.from(darkMode)].join(" ").trim();
}

//#endregion
//#region src/utils/transformers/transform-import.ts
function transformImport(opts) {
	return {
		type: "codemod",
		name: "modify import based on user config",
		transform({ scriptASTs, utils: { traverseScriptAST } }) {
			let transformCount = 0;
			const { config, isRemote } = opts;
			const utilsAlias = config.aliases?.utils;
			const utilsImport = `${typeof utilsAlias === "string" && utilsAlias.includes("/") ? utilsAlias.split("/")[0] : "@"}/lib/utils`;
			for (const scriptAST of scriptASTs) traverseScriptAST(scriptAST, { visitLiteral(path$2) {
				if (typeof path$2.node.value === "string") {
					const parent = path$2.parent.value;
					if (parent.type === "ImportDeclaration" || parent.type === "CallExpression" && parent.callee?.name === "import") {
						const sourcePath = path$2.node.value;
						const updatedImport = updateImportAliases(sourcePath, config, isRemote);
						if (updatedImport !== sourcePath) {
							path$2.node.value = updatedImport;
							transformCount++;
						}
						if (utilsImport === updatedImport || updatedImport === "@/lib/utils") {
							if (parent.type === "ImportDeclaration") {
								if ((parent.specifiers?.map((node) => node.local?.name ?? "") ?? []).find((i) => i === "cn") && config.aliases.utils) {
									path$2.node.value = utilsImport === updatedImport ? updatedImport.replace(utilsImport, config.aliases.utils) : config.aliases.utils;
									transformCount++;
								}
							} else if (parent.type === "CallExpression") {
								const grandParent = path$2.parent.parent?.value;
								if (grandParent?.type === "VariableDeclarator" && grandParent.id?.type === "ObjectPattern") {
									if (grandParent.id.properties?.some((prop) => prop.key?.name === "cn") && config.aliases.utils) {
										path$2.node.value = utilsImport === updatedImport ? updatedImport.replace(utilsImport, config.aliases.utils) : config.aliases.utils;
										transformCount++;
									}
								}
							}
						}
					}
				}
				return this.traverse(path$2);
			} });
			return transformCount;
		}
	};
}
function updateImportAliases(moduleSpecifier, config, isRemote = false) {
	if (!moduleSpecifier.startsWith("@/") && !isRemote) return moduleSpecifier;
	if (isRemote && moduleSpecifier.startsWith("@/")) moduleSpecifier = moduleSpecifier.replace(/^@\//, `@/registry/new-york/`);
	if (moduleSpecifier.match(/^@\/styles\//)) moduleSpecifier = moduleSpecifier.replace(/^@\/styles\//, "@/registry/");
	if (!moduleSpecifier.startsWith("@/registry/")) {
		const alias = config.aliases.components.split("/")[0];
		return moduleSpecifier.replace(/^@\//, `${alias}/`);
	}
	if (moduleSpecifier.match(/^@\/registry\/(.+)\/ui/)) return moduleSpecifier.replace(/^@\/registry\/(.+)\/ui/, config.aliases.ui ?? `${config.aliases.components}/ui`);
	if (config.aliases.components && moduleSpecifier.match(/^@\/registry\/(.+)\/components/)) return moduleSpecifier.replace(/^@\/registry\/(.+)\/components/, config.aliases.components);
	if (config.aliases.lib && moduleSpecifier.match(/^@\/registry\/(.+)\/lib/)) return moduleSpecifier.replace(/^@\/registry\/(.+)\/lib/, config.aliases.lib);
	if (config.aliases.composables && moduleSpecifier.match(/^@\/registry\/(.+)\/composables/)) return moduleSpecifier.replace(/^@\/registry\/(.+)\/composables/, config.aliases.composables);
	return moduleSpecifier.replace(/^@\/registry\/[^/]+/, config.aliases.components);
}

//#endregion
//#region src/utils/transformers/transform-rtl.ts
const RTL_MAPPINGS = [
	["-ml-", "-ms-"],
	["-mr-", "-me-"],
	["ml-", "ms-"],
	["mr-", "me-"],
	["pl-", "ps-"],
	["pr-", "pe-"],
	["-left-", "-start-"],
	["-right-", "-end-"],
	["left-", "start-"],
	["right-", "end-"],
	["inset-l-", "inset-inline-start-"],
	["inset-r-", "inset-inline-end-"],
	["rounded-tl-", "rounded-ss-"],
	["rounded-tr-", "rounded-se-"],
	["rounded-bl-", "rounded-es-"],
	["rounded-br-", "rounded-ee-"],
	["rounded-l-", "rounded-s-"],
	["rounded-r-", "rounded-e-"],
	["border-l-", "border-s-"],
	["border-r-", "border-e-"],
	["border-l", "border-s"],
	["border-r", "border-e"],
	["text-left", "text-start"],
	["text-right", "text-end"],
	["scroll-ml-", "scroll-ms-"],
	["scroll-mr-", "scroll-me-"],
	["scroll-pl-", "scroll-ps-"],
	["scroll-pr-", "scroll-pe-"],
	["float-left", "float-start"],
	["float-right", "float-end"],
	["clear-left", "clear-start"],
	["clear-right", "clear-end"],
	["origin-top-left", "origin-top-start"],
	["origin-top-right", "origin-top-end"],
	["origin-bottom-left", "origin-bottom-start"],
	["origin-bottom-right", "origin-bottom-end"],
	["origin-left", "origin-start"],
	["origin-right", "origin-end"]
];
const RTL_TRANSLATE_X_MAPPINGS = [["-translate-x-", "translate-x-"], ["translate-x-", "-translate-x-"]];
const RTL_REVERSE_MAPPINGS = [["space-x-", "space-x-reverse"], ["divide-x-", "divide-x-reverse"]];
const RTL_SWAP_MAPPINGS = [["cursor-w-resize", "cursor-e-resize"], ["cursor-e-resize", "cursor-w-resize"]];
const RTL_LOGICAL_SIDE_SLIDE_MAPPINGS = [
	[
		"data-[side=inline-start]",
		"slide-in-from-right",
		"slide-in-from-end"
	],
	[
		"data-[side=inline-start]",
		"slide-out-to-right",
		"slide-out-to-end"
	],
	[
		"data-[side=inline-end]",
		"slide-in-from-left",
		"slide-in-from-start"
	],
	[
		"data-[side=inline-end]",
		"slide-out-to-left",
		"slide-out-to-start"
	]
];
const RTL_FLIP_MARKER = "cn-rtl-flip";
const POSITIONING_PREFIXES = [
	"-left-",
	"-right-",
	"left-",
	"right-"
];
function applyRtlMapping(input) {
	return input.split(" ").flatMap((className) => {
		if (className.startsWith("rtl:") || className.startsWith("ltr:")) return [className];
		if (className === RTL_FLIP_MARKER) return ["rtl:rotate-180"];
		const [variant, value, modifier] = splitClassName(className);
		if (!value) return [className];
		for (const [physical, rtlPhysical] of RTL_TRANSLATE_X_MAPPINGS) if (value.startsWith(physical)) {
			const rtlValue = value.replace(physical, rtlPhysical);
			return [className, variant ? `rtl:${variant}:${rtlValue}${modifier ? `/${modifier}` : ""}` : `rtl:${rtlValue}${modifier ? `/${modifier}` : ""}`];
		}
		for (const [prefix, reverseClass] of RTL_REVERSE_MAPPINGS) if (value.startsWith(prefix)) return [className, variant ? `rtl:${variant}:${reverseClass}` : `rtl:${reverseClass}`];
		for (const [physical, swapped] of RTL_SWAP_MAPPINGS) if (value === physical) return [className, variant ? `rtl:${variant}:${swapped}` : `rtl:${swapped}`];
		for (const [variantPattern, physical, logical] of RTL_LOGICAL_SIDE_SLIDE_MAPPINGS) if (variant?.includes(variantPattern) && value.startsWith(physical)) {
			const mappedValue$1 = value.replace(physical, logical);
			return [modifier ? `${variant}:${mappedValue$1}/${modifier}` : `${variant}:${mappedValue$1}`];
		}
		const isPhysicalSideVariant = variant?.includes("data-[side=left]") || variant?.includes("data-[side=right]");
		let mappedValue = value;
		for (const [physical, logical] of RTL_MAPPINGS) {
			if (isPhysicalSideVariant && POSITIONING_PREFIXES.some((p) => physical.startsWith(p))) continue;
			if (value.startsWith(physical)) {
				if (!physical.endsWith("-") && value !== physical) continue;
				mappedValue = value.replace(physical, logical);
				break;
			}
		}
		let result;
		if (variant) result = modifier ? `${variant}:${mappedValue}/${modifier}` : `${variant}:${mappedValue}`;
		else result = modifier ? `${mappedValue}/${modifier}` : mappedValue;
		return [result];
	}).join(" ");
}
const CLASS_ATTR_NAMES = new Set(["class", "className"]);
function transformRtl(opts) {
	return {
		type: "codemod",
		name: "transform physical tailwind classes to logical rtl classes",
		transform({ scriptASTs, sfcAST, utils: { traverseScriptAST, traverseTemplateAST } }) {
			let transformCount = 0;
			const { config } = opts;
			if (!config.rtl) return transformCount;
			for (const scriptAST of scriptASTs) traverseScriptAST(scriptAST, { visitLiteral(path$2) {
				if (path$2.parent.value.type !== "ImportDeclaration" && typeof path$2.node.value === "string") {
					const raw = path$2.node.value;
					const mapped = applyRtlMapping(raw);
					if (mapped !== raw) {
						path$2.node.value = mapped;
						transformCount++;
					}
				}
				return this.traverse(path$2);
			} });
			if (sfcAST) {
				const isInsideClassAttribute = (node) => {
					let current = node?.parent;
					while (current) {
						if (current.type === "VAttribute") {
							if (current.key?.type === "VIdentifier" && typeof current.key.name === "string" && CLASS_ATTR_NAMES.has(current.key.name)) return true;
							if (current.key?.type === "VDirectiveKey" && current.key.argument?.type === "VIdentifier" && CLASS_ATTR_NAMES.has(current.key.argument.name)) return true;
							return false;
						}
						current = current.parent;
					}
					return false;
				};
				traverseTemplateAST(sfcAST, {
					enterNode(node) {
						if (node.type === "VLiteral" && typeof node.value === "string") {
							if (node.parent?.type === "VAttribute" && node.parent.key?.type === "VIdentifier" && CLASS_ATTR_NAMES.has(node.parent.key.name)) {
								const cleanValue = node.value.replace(/"/g, "");
								const mapped = applyRtlMapping(cleanValue);
								if (mapped !== cleanValue) {
									node.value = mapped;
									transformCount++;
								}
							}
						} else if (node.type === "Literal" && typeof node.value === "string") {
							if (isInsideClassAttribute(node)) {
								const raw = node.value;
								const mapped = applyRtlMapping(raw);
								if (mapped !== raw) {
									node.value = mapped;
									transformCount++;
								}
							}
						}
					},
					leaveNode() {}
				});
			}
			return transformCount;
		}
	};
}

//#endregion
//#region src/utils/transformers/transform-sfc.ts
async function transformSFC(opts) {
	if (opts.config?.typescript) return opts.raw;
	return await transformByDetype(opts.raw, opts.filename).then((res) => res);
}
async function transformByDetype(content, filename) {
	return await transform$1(content, filename, {
		removeTsComments: true,
		prettierOptions: { proseWrap: "never" }
	});
}

//#endregion
//#region src/utils/transformers/transform-style.ts
const STYLE_CLASS_MAPPINGS = {
	nova: {
		"p-6": "p-4",
		"p-8": "p-6",
		"px-6": "px-4",
		"px-8": "px-6",
		"py-6": "py-4",
		"py-8": "py-6",
		"m-6": "m-4",
		"m-8": "m-6",
		"gap-6": "gap-4",
		"gap-8": "gap-6",
		"space-y-6": "space-y-4",
		"space-y-8": "space-y-6",
		"space-x-6": "space-x-4",
		"space-x-8": "space-x-6"
	},
	maia: {
		"rounded-md": "rounded-xl",
		"rounded-lg": "rounded-2xl",
		"rounded-sm": "rounded-md",
		"p-4": "p-6",
		"p-3": "p-4",
		"px-4": "px-6",
		"py-4": "py-6",
		"gap-4": "gap-6",
		"gap-3": "gap-4"
	},
	lyra: {
		"rounded-md": "rounded-none",
		"rounded-lg": "rounded-none",
		"rounded-xl": "rounded-none",
		"rounded-2xl": "rounded-none",
		"rounded-sm": "rounded-none",
		"rounded-full": "rounded-none"
	},
	mira: {
		"p-4": "p-2",
		"p-6": "p-4",
		"p-8": "p-5",
		"px-4": "px-2",
		"px-6": "px-4",
		"py-4": "py-2",
		"py-6": "py-4",
		"m-4": "m-2",
		"m-6": "m-4",
		"gap-4": "gap-2",
		"gap-6": "gap-4",
		"space-y-4": "space-y-2",
		"space-y-6": "space-y-4",
		"space-x-4": "space-x-2",
		"space-x-6": "space-x-4",
		"text-base": "text-sm",
		"text-lg": "text-base",
		"text-xl": "text-lg",
		"h-10": "h-8",
		"h-12": "h-10",
		"w-10": "w-8",
		"w-12": "w-10"
	},
	luma: {
		"rounded-sm": "rounded-xl",
		"rounded-md": "rounded-2xl",
		"rounded-lg": "rounded-3xl",
		"rounded-xl": "rounded-3xl"
	},
	sera: {
		"rounded-sm": "rounded-none",
		"rounded-md": "rounded-none",
		"rounded-lg": "rounded-none",
		"rounded-xl": "rounded-none",
		"rounded-2xl": "rounded-none",
		"rounded-3xl": "rounded-none"
	}
};
/**
* Apply class mappings to a string of CSS classes.
* Uses word boundary matching to avoid partial replacements.
*/
function applyClassMappings(value, classMapping) {
	let result = value;
	for (const [from, to] of Object.entries(classMapping)) {
		const regex = new RegExp(`\\b${from}\\b`, "g");
		result = result.replace(regex, to);
	}
	return result;
}
/**
* Transform component classes based on the selected visual style.
* Handles:
* - Static class attributes in templates (class="...")
* - Dynamic class bindings in templates (:class="cn(...)")
* - String literals in script (CVA variants, inline classes)
*
* Vega is the default style, so no transformations are applied.
*/
function transformStyle(opts) {
	return {
		type: "codemod",
		name: "transform-style",
		transform({ scriptASTs, sfcAST, utils: { traverseScriptAST, traverseTemplateAST } }) {
			let transformCount = 0;
			const { config } = opts;
			const style = config.style?.split("-")[0] || "vega";
			if (style === "vega" || style === "new") return transformCount;
			const classMapping = STYLE_CLASS_MAPPINGS[style];
			if (!classMapping) return transformCount;
			for (const scriptAST of scriptASTs) traverseScriptAST(scriptAST, {
				visitLiteral(path$2) {
					if (path$2.parent.value.type === "ImportDeclaration") return this.traverse(path$2);
					if (typeof path$2.node.value === "string") {
						const originalValue = path$2.node.value;
						const newValue = applyClassMappings(originalValue, classMapping);
						if (newValue !== originalValue) {
							path$2.node.value = newValue;
							transformCount++;
						}
					}
					return this.traverse(path$2);
				},
				visitTemplateLiteral(path$2) {
					for (const quasi of path$2.node.quasis) if (quasi.value.raw) {
						const originalValue = quasi.value.raw;
						const newValue = applyClassMappings(originalValue, classMapping);
						if (newValue !== originalValue) {
							quasi.value.raw = newValue;
							quasi.value.cooked = newValue;
							transformCount++;
						}
					}
					return this.traverse(path$2);
				}
			});
			if (sfcAST) traverseTemplateAST(sfcAST, { enterNode(node) {
				if (node.type !== "VElement" || !node.startTag?.attributes) return;
				for (const attr of node.startTag.attributes) {
					if (attr.type === "VAttribute" && attr.key.type === "VIdentifier" && attr.key.name === "class" && attr.value && "value" in attr.value && typeof attr.value.value === "string") {
						const originalValue = attr.value.value;
						const newValue = applyClassMappings(originalValue, classMapping);
						if (newValue !== originalValue) {
							attr.value.value = newValue;
							transformCount++;
						}
					}
					if (attr.type === "VAttribute" && attr.key.type === "VDirectiveKey" && attr.key.argument?.type === "VIdentifier" && attr.key.argument.name === "class" && attr.value?.type === "VExpressionContainer" && attr.value.expression) transformExpression(attr.value.expression, classMapping, (count) => {
						transformCount += count;
					});
				}
			} });
			return transformCount;
		}
	};
}
/**
* Recursively traverse a Vue expression AST node to find and transform
* string literals that contain CSS classes.
*/
function transformExpression(node, classMapping, onTransform) {
	if (!node) return;
	if (node.type === "Literal" && typeof node.value === "string") {
		const originalValue = node.value;
		const newValue = applyClassMappings(originalValue, classMapping);
		if (newValue !== originalValue) {
			node.value = newValue;
			if (node.raw) {
				const quote = node.raw[0];
				node.raw = `${quote}${newValue}${quote}`;
			}
			onTransform(1);
		}
		return;
	}
	if (node.type === "TemplateLiteral" && node.quasis) {
		for (const quasi of node.quasis) if (quasi.value?.raw) {
			const originalValue = quasi.value.raw;
			const newValue = applyClassMappings(originalValue, classMapping);
			if (newValue !== originalValue) {
				quasi.value.raw = newValue;
				quasi.value.cooked = newValue;
				onTransform(1);
			}
		}
		if (node.expressions) for (const expr of node.expressions) transformExpression(expr, classMapping, onTransform);
		return;
	}
	if (node.type === "CallExpression" && node.arguments) {
		for (const arg of node.arguments) transformExpression(arg, classMapping, onTransform);
		return;
	}
	if (node.type === "ArrayExpression" && node.elements) {
		for (const element of node.elements) if (element) transformExpression(element, classMapping, onTransform);
		return;
	}
	if (node.type === "ConditionalExpression") {
		transformExpression(node.consequent, classMapping, onTransform);
		transformExpression(node.alternate, classMapping, onTransform);
		return;
	}
	if (node.type === "LogicalExpression") {
		transformExpression(node.left, classMapping, onTransform);
		transformExpression(node.right, classMapping, onTransform);
		return;
	}
	if (node.type === "ObjectExpression" && node.properties) {
		for (const prop of node.properties) if (prop.key) transformExpression(prop.key, classMapping, onTransform);
	}
}

//#endregion
//#region src/utils/transformers/transform-tw-prefix.ts
async function transformTwPrefix(opts) {
	const tailwindVersion = await getProjectTailwindVersionFromConfig(opts.config);
	return {
		type: "codemod",
		name: "add prefix to tailwind classes",
		transform({ scriptASTs, sfcAST, utils: { traverseScriptAST, traverseTemplateAST, astHelpers } }) {
			let transformCount = 0;
			const { config } = opts;
			if (!config.tailwind?.prefix) return transformCount;
			const addPrefix = (input) => {
				const result = applyPrefix(input, config.tailwind.prefix, tailwindVersion);
				transformCount++;
				return result;
			};
			function isVariantProperty(node) {
				if (node.type === "Property") {
					if (node.key?.type === "Identifier") {
						const keyName = node.key.name;
						return [
							"variant",
							"size",
							"color",
							"type",
							"state"
						].includes(keyName);
					}
					if (node.key?.type === "Literal" && typeof node.key.value === "string") {
						const keyName = node.key.value;
						return [
							"variant",
							"size",
							"color",
							"type",
							"state"
						].includes(keyName);
					}
				}
				return false;
			}
			function traverseExpression(expression) {
				if (expression.type === "CallExpression" && expression.callee?.type === "Identifier" && expression.callee.name === "cn") expression.arguments.forEach((arg) => {
					if (arg.type === "Literal" && typeof arg.value === "string") arg.value = addPrefix(arg.value);
					else if (arg.type === "ConditionalExpression") {
						if (arg.consequent?.type === "Literal" && typeof arg.consequent.value === "string") arg.consequent.value = addPrefix(arg.consequent.value);
						if (arg.alternate?.type === "Literal" && typeof arg.alternate.value === "string") arg.alternate.value = addPrefix(arg.alternate.value);
					} else if (arg.type === "BinaryExpression") {
						if (arg.right?.type === "Literal" && typeof arg.right.value === "string") arg.right.value = addPrefix(arg.right.value);
					} else if (arg.type === "ObjectExpression") arg.properties.forEach((prop) => {
						if (prop.type === "Property" && prop.value?.type === "Literal" && typeof prop.value.value === "string") {
							if (!isVariantProperty(prop)) prop.value.value = addPrefix(prop.value.value);
						}
					});
					else astHelpers.findAll(arg, { type: "Literal" }).forEach((literal) => {
						if (typeof literal.value === "string") {
							let shouldTransform = true;
							let parent = literal.parent;
							while (parent) {
								if (isVariantProperty(parent)) {
									shouldTransform = false;
									break;
								}
								parent = parent.parent;
							}
							if (shouldTransform) literal.value = addPrefix(literal.value);
						}
					});
				});
				else if (expression.type === "ConditionalExpression") {
					if (expression.consequent) traverseExpression(expression.consequent);
					if (expression.alternate) traverseExpression(expression.alternate);
				} else if (expression.type === "BinaryExpression") {
					if (expression.left) traverseExpression(expression.left);
					if (expression.right) traverseExpression(expression.right);
				}
			}
			for (const scriptAST of scriptASTs) traverseScriptAST(scriptAST, { visitCallExpression(path$2) {
				if (path$2.node.callee.type === "Identifier" && path$2.node.callee.name === "cva") {
					const args = path$2.node.arguments;
					if (args[0]?.type === "Literal" && typeof args[0].value === "string") args[0].value = addPrefix(args[0].value);
					if (args[1]?.type === "ObjectExpression") {
						const variantsProperty = args[1].properties.find((prop) => prop.type === "Property" && prop.key.type === "Identifier" && prop.key.name === "variants");
						if (variantsProperty && variantsProperty.type === "Property" && variantsProperty.value.type === "ObjectExpression") astHelpers.findAll(variantsProperty.value, { type: "Property" }).forEach((prop) => {
							if (prop.value?.type === "Literal" && typeof prop.value.value === "string") prop.value.value = addPrefix(prop.value.value);
							else if (prop.value?.type === "ArrayExpression") prop.value.elements.forEach((element) => {
								if (element?.type === "Literal" && typeof element.value === "string") element.value = addPrefix(element.value);
							});
						});
					}
				}
				if (path$2.node.callee.type === "Identifier" && path$2.node.callee.name === "cn") path$2.node.arguments.forEach((arg) => {
					if (arg.type === "Literal" && typeof arg.value === "string") arg.value = addPrefix(arg.value);
					else if (arg.type === "ConditionalExpression") {
						if (arg.consequent?.type === "Literal" && typeof arg.consequent.value === "string") arg.consequent.value = addPrefix(arg.consequent.value);
						if (arg.alternate?.type === "Literal" && typeof arg.alternate.value === "string") arg.alternate.value = addPrefix(arg.alternate.value);
					} else if (arg.type === "BinaryExpression") {
						if (arg.right?.type === "Literal" && typeof arg.right.value === "string") arg.right.value = addPrefix(arg.right.value);
					} else if (arg.type === "ObjectExpression") arg.properties.forEach((prop) => {
						if (prop.type === "Property" && prop.value?.type === "Literal" && typeof prop.value.value === "string") {
							if (!isVariantProperty(prop)) prop.value.value = addPrefix(prop.value.value);
						}
					});
					else astHelpers.findAll(arg, { type: "Literal" }).forEach((literal) => {
						if (typeof literal.value === "string") {
							let shouldTransform = true;
							let parent = literal.parent;
							while (parent) {
								if (isVariantProperty(parent)) {
									shouldTransform = false;
									break;
								}
								parent = parent.parent;
							}
							if (shouldTransform) literal.value = addPrefix(literal.value);
						}
					});
				});
				return this.traverse(path$2);
			} });
			if (sfcAST) traverseTemplateAST(sfcAST, {
				enterNode(node) {
					if (node.type === "VAttribute" && node.key.type === "VDirectiveKey") {
						if (node.key.argument?.type === "VIdentifier") {
							const argName = node.key.argument.name;
							if ([
								"class",
								"className",
								"classes",
								"classNames"
							].includes(argName)) {
								if (node.value?.type === "VExpressionContainer" && node.value.expression) traverseExpression(node.value.expression);
							}
						}
					} else if (node.type === "VLiteral" && typeof node.value === "string") {
						if (node.parent?.type === "VAttribute" && node.parent.key?.type === "VIdentifier" && [
							"class",
							"className",
							"classes",
							"classNames"
						].includes(node.parent.key.name)) node.value = addPrefix(node.value.replace(/"/g, ""));
					}
				},
				leaveNode() {}
			});
			return transformCount;
		}
	};
}
function applyPrefix(input, prefix = "", tailwindVersion) {
	if (tailwindVersion === "v3") return input.split(" ").map((className) => {
		const [variant, value, modifier] = splitClassName(className);
		if (variant) return modifier ? `${variant}:${prefix}${value}/${modifier}` : `${variant}:${prefix}${value}`;
		else return modifier ? `${prefix}${value}/${modifier}` : `${prefix}${value}`;
	}).join(" ");
	return input.split(" ").map((className) => className.indexOf(`${prefix}:`) === 0 ? className : `${prefix}:${className.trim()}`).join(" ");
}

//#endregion
//#region src/utils/transformers/index.ts
async function transform$2(opts) {
	if ([
		".md",
		".mdx",
		".txt"
	].some((ext) => opts.filename.endsWith(ext))) return opts.raw;
	const source = await transformSFC(opts);
	const registryIcons = await getRegistryIcons();
	return transform(source, opts.filename, [
		transformImport(opts),
		transformCssVars(opts),
		transformRtl(opts),
		await transformTwPrefix(opts),
		transformIcons(opts, registryIcons),
		transformMenu(opts),
		transformStyle(opts)
	]).code;
}

//#endregion
//#region src/utils/updaters/update-files.ts
async function updateFiles(files, config, options) {
	if (!files?.length) return {
		filesCreated: [],
		filesUpdated: [],
		filesSkipped: []
	};
	options = {
		overwrite: false,
		force: false,
		silent: false,
		isRemote: false,
		isWorkspace: false,
		...options
	};
	const filesCreatedSpinner = spinner(`Updating files.`, { silent: options.silent })?.start();
	const [projectInfo, baseColor] = await Promise.all([getProjectInfo(config.resolvedPaths.cwd), config.tailwind.baseColor ? getRegistryBaseColor(config.tailwind.baseColor) : Promise.resolve(void 0)]);
	let filesCreated = [];
	let filesUpdated = [];
	let filesSkipped = [];
	let envVarsAdded = [];
	let envFile = null;
	for (let index = 0; index < files.length; index++) {
		const file = files[index];
		if (!file.content) continue;
		let filePath = resolveFilePath(file, config, {
			framework: projectInfo?.framework.name,
			commonRoot: findCommonRoot(files.map((f) => f.path), file.path),
			path: options.path,
			fileIndex: index
		});
		if (!filePath) continue;
		basename(file.path);
		const targetDir = path.dirname(filePath);
		if (!config.typescript) filePath = filePath.replace(/\.ts?$/, (match) => ".js");
		if (isEnvFile(filePath) && !existsSync(filePath)) {
			const alternativeEnvFile = findExistingEnvFile(targetDir);
			if (alternativeEnvFile) filePath = alternativeEnvFile;
		}
		const existingFile = existsSync(filePath);
		if (existingFile && statSync(filePath).isDirectory()) throw new Error(`Cannot write to ${filePath}: path exists and is a directory. Please provide a file path instead.`);
		const content = isEnvFile(filePath) ? file.content : await transform$2({
			filename: file.path,
			raw: file.content,
			config,
			baseColor,
			isRemote: options.isRemote
		});
		if (existingFile && !isEnvFile(filePath)) {
			if (isContentSame(await promises.readFile(filePath, "utf-8"), content, { ignoreImports: options.isWorkspace })) {
				filesSkipped.push(path.relative(config.resolvedPaths.cwd, filePath));
				continue;
			}
		}
		if (existingFile && !options.overwrite && !isEnvFile(filePath)) {
			filesCreatedSpinner.stop();
			if (options.rootSpinner) options.rootSpinner.stop();
			const { overwrite } = await prompts({
				type: "confirm",
				name: "overwrite",
				message: `The file ${highlighter.info(path.relative(config.resolvedPaths.ui, filePath))} already exists. Would you like to overwrite?`,
				initial: false
			});
			if (!overwrite) {
				filesSkipped.push(path.relative(config.resolvedPaths.cwd, filePath));
				if (options.rootSpinner) options.rootSpinner.start();
				continue;
			}
			filesCreatedSpinner?.start();
			if (options.rootSpinner) options.rootSpinner.start();
		}
		if (!existsSync(targetDir)) await promises.mkdir(targetDir, { recursive: true });
		if (isEnvFile(filePath) && existingFile) {
			const existingFileContent = await promises.readFile(filePath, "utf-8");
			const mergedContent = mergeEnvContent(existingFileContent, content);
			envVarsAdded = getNewEnvKeys(existingFileContent, content);
			envFile = path.relative(config.resolvedPaths.cwd, filePath);
			if (!envVarsAdded.length) {
				filesSkipped.push(path.relative(config.resolvedPaths.cwd, filePath));
				continue;
			}
			await promises.writeFile(filePath, mergedContent, "utf-8");
			filesUpdated.push(path.relative(config.resolvedPaths.cwd, filePath));
			continue;
		}
		await promises.writeFile(filePath, content, "utf-8");
		if (!existingFile) {
			filesCreated.push(path.relative(config.resolvedPaths.cwd, filePath));
			if (isEnvFile(filePath)) {
				envVarsAdded = Object.keys(parseEnvContent(content));
				envFile = path.relative(config.resolvedPaths.cwd, filePath);
			}
		} else filesUpdated.push(path.relative(config.resolvedPaths.cwd, filePath));
	}
	const updatedFiles = await resolveImports([
		...filesCreated,
		...filesUpdated,
		...filesSkipped
	], config);
	filesUpdated.push(...updatedFiles);
	filesUpdated = filesUpdated.filter((file) => !filesCreated.includes(file));
	if (!(filesCreated.length || filesUpdated.length) && !filesSkipped.length) filesCreatedSpinner?.info("No files updated.");
	filesCreated = Array.from(new Set(filesCreated));
	filesUpdated = Array.from(new Set(filesUpdated));
	filesSkipped = Array.from(new Set(filesSkipped));
	if (filesCreated.length) {
		filesCreatedSpinner?.succeed(`Created ${filesCreated.length} ${filesCreated.length === 1 ? "file" : "files"}:`);
		if (!options.silent) for (const file of filesCreated) logger.log(`  - ${file}`);
	} else filesCreatedSpinner?.stop();
	if (filesUpdated.length) {
		spinner(`Updated ${filesUpdated.length} ${filesUpdated.length === 1 ? "file" : "files"}:`, { silent: options.silent })?.info();
		if (!options.silent) for (const file of filesUpdated) logger.log(`  - ${file}`);
	}
	if (filesSkipped.length) {
		spinner(`Skipped ${filesSkipped.length} ${filesUpdated.length === 1 ? "file" : "files"}: (files might be identical, use --overwrite to overwrite)`, { silent: options.silent })?.info();
		if (!options.silent) for (const file of filesSkipped) logger.log(`  - ${file}`);
	}
	if (envVarsAdded.length && envFile) {
		spinner(`Added the following variables to ${highlighter.info(envFile)}:`)?.info();
		if (!options.silent) for (const key of envVarsAdded) logger.log(`  ${highlighter.success("+")} ${key}`);
	}
	if (!options.silent) logger.break();
	return {
		filesCreated,
		filesUpdated,
		filesSkipped
	};
}
function resolveFilePath(file, config, options) {
	if (options.path) {
		const resolvedPath = path.isAbsolute(options.path) ? options.path : path.join(config.resolvedPaths.cwd, options.path);
		if (/\.[^/\\]+$/.test(resolvedPath)) {
			if (options.fileIndex === 0) return resolvedPath;
		} else {
			const fileName = path.basename(file.path);
			return path.join(resolvedPath, fileName);
		}
	}
	if (file.target) {
		if (file.target.startsWith("~/")) return path.join(config.resolvedPaths.cwd, file.target.replace("~/", ""));
		let target = file.target;
		if (file.type === "registry:page") {
			target = resolvePageTarget(target, options.framework);
			if (!target) return "";
		}
		return path.join(config.resolvedPaths.cwd, target.replace("src/", ""));
	}
	const targetDir = resolveFileTargetDirectory(file, config);
	const relativePath = resolveNestedFilePath(file.path, options.commonRoot, config);
	return path.join(targetDir, relativePath);
}
function resolveFileTargetDirectory(file, config) {
	if (file.type === "registry:ui") return config.resolvedPaths.ui;
	if (file.type === "registry:lib") return config.resolvedPaths.lib;
	if (file.type === "registry:block" || file.type === "registry:component") return config.resolvedPaths.components;
	if (file.type === "registry:hook" || file.type === "registry:composable") return config.resolvedPaths.composables;
	return config.resolvedPaths.components;
}
function findCommonRoot(paths, needle) {
	const normalizedPaths = paths.map((p) => p.replace(/^\//, ""));
	const normalizedNeedle = needle.replace(/^\//, "");
	const needleDir = normalizedNeedle.split("/").slice(0, -1).join("/");
	if (!needleDir) return "";
	const needleSegments = needleDir.split("/");
	for (let i = needleSegments.length; i > 0; i--) {
		const testPath = needleSegments.slice(0, i).join("/");
		if (normalizedPaths.some((path$2) => path$2 !== normalizedNeedle && path$2.startsWith(`${testPath}/`))) return `/${testPath}`;
	}
	return `/${needleDir}`;
}
function resolveNestedFilePath(filePath, commonRoot, config) {
	const normalizedFilePath = filePath.replace(/^\/|\/$/g, "");
	const normalizedCommonRoot = commonRoot.replace(/^\/|\/$/g, "");
	const aliases = Object.values(config.aliases).map((alias) => alias.replace(/^@\//, "").replace(/^\/|\/$/g, "")).sort((a, b) => b.length - a.length);
	for (const alias of aliases) if (normalizedCommonRoot.includes(alias)) {
		const aliasEndIndex = normalizedFilePath.indexOf(alias) + alias.length;
		return normalizedFilePath.substring(aliasEndIndex).replace(/^\//, "");
	}
	return normalizedCommonRoot.split("/").pop() + normalizedFilePath.replace(normalizedCommonRoot, "");
}
function resolvePageTarget(target, framework) {
	if (!framework) return "";
	if (framework === "nuxt3" || framework === "nuxt4") return target;
	if (framework === "laravel") {
		let result = target.replace(/^app\//, "resources/js/pages/");
		result = result.replace(/\/page(\.[jt]sx?)$/, "$1");
		return result;
	}
	return "";
}
async function resolveImports(filePaths, config) {
	const projectInfo = await getProjectInfo(config.resolvedPaths.cwd);
	const tsConfig = getTsconfig(config.resolvedPaths.cwd);
	const updatedFiles = [];
	if (!projectInfo || tsConfig === null) return [];
	for (const filepath of filePaths) {
		const resolvedPath = path.resolve(config.resolvedPaths.cwd, filepath);
		if (!existsSync(resolvedPath)) continue;
		const content = await promises.readFile(resolvedPath, "utf-8");
		try {
			const importResolverTransformer = (opts) => ({
				name: "import-resolver",
				transform(node) {
					if (node.type === "ImportDeclaration" && node.source?.value) {
						const moduleSpecifier = node.source.value;
						if (projectInfo?.aliasPrefix && !moduleSpecifier.startsWith(`${projectInfo.aliasPrefix}/`)) return;
						const probableImportFilePath = resolveImport(moduleSpecifier, tsConfig);
						if (!probableImportFilePath) return;
						const resolvedImportFilePath = resolveModuleByProbablePath(probableImportFilePath, filePaths, config);
						if (!resolvedImportFilePath) return;
						const newImport = toAliasedImport(resolvedImportFilePath, config, projectInfo);
						if (!newImport || newImport === moduleSpecifier) return;
						node.source.value = newImport;
						node.source.raw = `'${newImport}'`;
					}
				}
			});
			const result = transform(content, resolvedPath, [importResolverTransformer({})]);
			if (result.code !== content) {
				await promises.writeFile(resolvedPath, result.code, "utf-8");
				updatedFiles.push(filepath);
			}
		} catch (error) {
			console.warn(`Failed to transform imports in ${filepath}:`, error);
		}
	}
	return updatedFiles;
}
/**
* Given an absolute "probable" import path (no ext),
* plus an array of absolute file paths you already know about,
* return 0–N matches (best match first), and also check disk for any missing ones.
*/
function resolveModuleByProbablePath(probableImportFilePath, files, config, extensions = [
	".vue",
	".ts",
	".js",
	".tsx",
	".jsx",
	".css"
]) {
	const cwd = path.normalize(config.resolvedPaths.cwd);
	const relativeFiles = files.map((f) => f.split(path.sep).join(path.posix.sep));
	const fileSet = new Set(relativeFiles);
	const extInPath = path.extname(probableImportFilePath);
	const hasExt = extInPath !== "";
	const absBase = hasExt ? probableImportFilePath.slice(0, -extInPath.length) : probableImportFilePath;
	const relBase = path.relative(cwd, absBase).split(path.sep).join(path.posix.sep);
	const tryExts = hasExt ? [extInPath] : extensions;
	const candidates = /* @__PURE__ */ new Set();
	for (const e of tryExts) {
		const absCand = absBase + e;
		const relCand = path.posix.normalize(path.relative(cwd, absCand));
		if (fileSet.has(relCand) || existsSync(absCand)) candidates.add(relCand);
		const absIdx = path.join(absBase, `index${e}`);
		const relIdx = path.posix.normalize(path.relative(cwd, absIdx));
		if (fileSet.has(relIdx) || existsSync(absIdx)) candidates.add(relIdx);
	}
	const name = path.basename(absBase);
	for (const f of relativeFiles) if (tryExts.some((e) => f.endsWith(`/${name}${e}`))) candidates.add(f);
	if (candidates.size === 0) return null;
	return Array.from(candidates).sort((a, b) => {
		const aExt = path.posix.extname(a);
		const bExt = path.posix.extname(b);
		const ord = tryExts.indexOf(aExt) - tryExts.indexOf(bExt);
		if (ord !== 0) return ord;
		return (relBase && a.startsWith(relBase) ? -1 : 1) - (relBase && b.startsWith(relBase) ? -1 : 1);
	})[0];
}
function toAliasedImport(filePath, config, projectInfo) {
	const abs = path.normalize(path.join(config.resolvedPaths.cwd, filePath));
	const matches = Object.entries(config.resolvedPaths).filter(([, root]) => root && abs.startsWith(path.normalize(root + path.sep))).sort((a, b) => b[1].length - a[1].length);
	if (matches.length === 0) return null;
	const [aliasKey, rootDir] = matches[0];
	let rel = path.relative(rootDir, abs);
	rel = rel.split(path.sep).join("/");
	const ext = path.posix.extname(rel);
	const keepExt = [
		".vue",
		".ts",
		".tsx",
		".js",
		".jsx"
	].includes(ext) ? "" : ext;
	let noExt = rel.slice(0, rel.length - ext.length);
	if (noExt.endsWith("/index")) noExt = noExt.slice(0, -6);
	const aliasBase = aliasKey === "cwd" ? projectInfo.aliasPrefix : config.aliases[aliasKey];
	if (!aliasBase) return null;
	let suffix = noExt === "" ? "" : `/${noExt}`;
	suffix = suffix.replace("/src", "");
	return `${aliasBase}${suffix}${keepExt}`;
}

//#endregion
//#region src/registry/utils.ts
const project = new Project({ compilerOptions: {} });
function isUrl$1(path$2) {
	try {
		new URL(path$2);
		return true;
	} catch (error) {
		return false;
	}
}
function isLocalFile(path$2) {
	return path$2.endsWith(".json") && !isUrl$1(path$2);
}
/**
* Check if a registry item is universal (framework-agnostic).
* A universal registry item must:
* 1. Have type "registry:item" or "registry:file"
* 2. If it has files, all files must have explicit targets and be type "registry:file" or "registry:item"
* It can be installed without framework detection or components.json.
*/
function isUniversalRegistryItem(registryItem) {
	if (!registryItem) return false;
	if (registryItem.type !== "registry:item" && registryItem.type !== "registry:file") return false;
	return (registryItem.files ?? []).every((file) => !!file.target && (file.type === "registry:file" || file.type === "registry:item"));
}
async function deduplicateFilesByTarget(filesArrays, config) {
	if (!canDeduplicateFiles(config)) return z.array(registryItemFileSchema).parse(filesArrays.flat().filter(Boolean));
	const projectInfo = await getProjectInfo(config.resolvedPaths.cwd);
	const targetMap = /* @__PURE__ */ new Map();
	const allFiles = z.array(registryItemFileSchema).parse(filesArrays.flat().filter(Boolean));
	allFiles.forEach((file) => {
		const resolvedPath = resolveFilePath(file, config, {
			framework: projectInfo?.framework.name,
			commonRoot: findCommonRoot(allFiles.map((f) => f.path), file.path)
		});
		if (resolvedPath) targetMap.set(resolvedPath, file);
	});
	return Array.from(targetMap.values());
}
function canDeduplicateFiles(config) {
	return !!(config?.resolvedPaths?.cwd && (config?.resolvedPaths?.ui || config?.resolvedPaths?.lib || config?.resolvedPaths?.components || config?.resolvedPaths?.composables));
}

//#endregion
//#region src/registry/context.ts
const context = { headers: {} };
function setRegistryHeaders(headers) {
	context.headers = {
		...context.headers,
		...headers
	};
}
function getRegistryHeadersFromContext(url) {
	return context.headers[url] || {};
}
function clearRegistryContext() {
	context.headers = {};
}

//#endregion
//#region src/registry/validator.ts
function extractEnvVarsFromRegistryConfig(config) {
	const vars = /* @__PURE__ */ new Set();
	if (typeof config === "string") extractEnvVars(config).forEach((v) => vars.add(v));
	else {
		extractEnvVars(config.url).forEach((v) => vars.add(v));
		if (config.params) Object.values(config.params).forEach((value) => {
			extractEnvVars(value).forEach((v) => vars.add(v));
		});
		if (config.headers) Object.values(config.headers).forEach((value) => {
			extractEnvVars(value).forEach((v) => vars.add(v));
		});
	}
	return Array.from(vars);
}
function validateRegistryConfig(registryName, config) {
	const missing = extractEnvVarsFromRegistryConfig(config).filter((v) => !process.env[v]);
	if (missing.length > 0) throw new RegistryMissingEnvironmentVariablesError(registryName, missing);
}
function validateRegistryConfigForItems(items, config) {
	for (const item of items) buildUrlAndHeadersForRegistryItem(item, configWithDefaults(config));
	clearRegistryContext();
}

//#endregion
//#region src/registry/builder.ts
const NAME_PLACEHOLDER = "{name}";
const STYLE_PLACEHOLDER = "{style}";
const ENV_VAR_PATTERN = /\$\{(\w+)\}/g;
const QUERY_PARAM_SEPARATOR = "?";
const QUERY_PARAM_DELIMITER = "&";
function buildUrlAndHeadersForRegistryItem(name, config) {
	const { registry, item } = parseRegistryAndItemFromString(name);
	if (!registry) return null;
	const registryConfig = (config?.registries || {})[registry];
	if (!registryConfig) throw new RegistryNotConfiguredError(registry);
	validateRegistryConfig(registry, registryConfig);
	return {
		url: buildUrlFromRegistryConfig(item, registryConfig, config),
		headers: buildHeadersFromRegistryConfig(registryConfig)
	};
}
function buildUrlFromRegistryConfig(item, registryConfig, config) {
	const registryStyle = resolveRegistryStyle(config?.style);
	if (typeof registryConfig === "string") {
		let url = registryConfig.replace(NAME_PLACEHOLDER, item);
		if (registryStyle && url.includes(STYLE_PLACEHOLDER)) url = url.replace(STYLE_PLACEHOLDER, registryStyle);
		return expandEnvVars(url);
	}
	let baseUrl = registryConfig.url.replace(NAME_PLACEHOLDER, item);
	if (registryStyle && baseUrl.includes(STYLE_PLACEHOLDER)) baseUrl = baseUrl.replace(STYLE_PLACEHOLDER, registryStyle);
	baseUrl = expandEnvVars(baseUrl);
	if (!registryConfig.params) return baseUrl;
	return appendQueryParams(baseUrl, registryConfig.params);
}
function buildHeadersFromRegistryConfig(config) {
	if (typeof config === "string" || !config.headers) return {};
	const headers = {};
	for (const [key, value] of Object.entries(config.headers)) {
		const expandedValue = expandEnvVars(value);
		if (shouldIncludeHeader(value, expandedValue)) headers[key] = expandedValue;
	}
	return headers;
}
function appendQueryParams(baseUrl, params) {
	const urlParams = new URLSearchParams();
	for (const [key, value] of Object.entries(params)) {
		const expandedValue = expandEnvVars(value);
		if (expandedValue) urlParams.append(key, expandedValue);
	}
	const queryString = urlParams.toString();
	if (!queryString) return baseUrl;
	return `${baseUrl}${baseUrl.includes(QUERY_PARAM_SEPARATOR) ? QUERY_PARAM_DELIMITER : QUERY_PARAM_SEPARATOR}${queryString}`;
}
function shouldIncludeHeader(originalValue, expandedValue) {
	const trimmedExpanded = expandedValue.trim();
	if (!trimmedExpanded) return false;
	if (originalValue.includes("${")) {
		if (originalValue.match(ENV_VAR_PATTERN)) return trimmedExpanded !== originalValue.replace(ENV_VAR_PATTERN, "").trim();
	}
	return true;
}
/**
* Resolves a registry URL from a path or URL string.
* Handles special cases like v0 registry URLs that need /json suffix.
*
* @param pathOrUrl - Either a relative path or a full URL
* @returns The resolved registry URL
*/
function resolveRegistryUrl(pathOrUrl) {
	if (isUrl$1(pathOrUrl)) {
		const url = new URL(pathOrUrl);
		if (url.pathname.match(/\/chat\/b\//) && !url.pathname.endsWith("/json")) url.pathname = `${url.pathname}/json`;
		return url.toString();
	}
	return `${REGISTRY_URL}/${pathOrUrl}`;
}

//#endregion
//#region src/registry/fetcher.ts
const agent = process.env.https_proxy ? new ProxyAgent(process.env.https_proxy) : void 0;
const registryCache = /* @__PURE__ */ new Map();
async function fetchRegistry(paths, options = {}) {
	options = {
		useCache: true,
		...options
	};
	try {
		return await Promise.all(paths.map(async (path$2) => {
			const url = resolveRegistryUrl(path$2);
			if (options.useCache && registryCache.has(url)) return registryCache.get(url);
			const fetchPromise = (async () => {
				const headers = getRegistryHeadersFromContext(url);
				return (await ofetch.raw(url, {
					agent,
					dispatcher: agent,
					parseResponse: JSON.parse,
					headers: { ...headers }
				}).catch(async (error) => {
					if (!error.response) throw new RegistryFetchError(url, void 0, error.message);
					const response = error.response;
					let messageFromServer;
					if (response.headers.get("content-type")?.includes("application/json")) {
						const json = await response._data;
						const parsed = z.object({
							detail: z.string().optional(),
							title: z.string().optional(),
							message: z.string().optional(),
							error: z.string().optional()
						}).safeParse(json);
						if (parsed.success) {
							messageFromServer = parsed.data.detail || parsed.data.message;
							if (parsed.data.error) messageFromServer = `[${parsed.data.error}] ${messageFromServer}`;
						}
					}
					if (response.status === 401) throw new RegistryUnauthorizedError(url, messageFromServer);
					if (response.status === 404) throw new RegistryNotFoundError(url, messageFromServer);
					if (response.status === 403) throw new RegistryForbiddenError(url, messageFromServer);
					throw new RegistryFetchError(url, response.status, messageFromServer);
				}))._data;
			})();
			if (options.useCache) registryCache.set(url, fetchPromise);
			return fetchPromise;
		}));
	} catch (error) {
		throw error;
	}
}
async function fetchRegistryLocal(filePath) {
	try {
		let expandedPath = filePath;
		if (filePath.startsWith("~/")) expandedPath = path.join(homedir(), filePath.slice(2));
		const resolvedPath = path.resolve(expandedPath);
		const content = await promises.readFile(resolvedPath, "utf8");
		const parsed = JSON.parse(content);
		try {
			return registryItemSchema.parse(parsed);
		} catch (error) {
			throw new RegistryParseError(filePath, error);
		}
	} catch (error) {
		if (error instanceof Error && (error.message.includes("ENOENT") || error.message.includes("no such file"))) throw new RegistryLocalFileError(filePath, error);
		if (error instanceof RegistryParseError) throw error;
		throw new RegistryLocalFileError(filePath, error);
	}
}

//#endregion
//#region src/utils/handle-error.ts
function handleError(error) {
	logger.break();
	logger.error(`Something went wrong. Please check the error below for more details.`);
	logger.error(`If the problem persists, please open an issue on GitHub.`);
	logger.error("");
	if (typeof error === "string") {
		logger.error(error);
		logger.break();
		process.exit(1);
	}
	if (error instanceof RegistryError) {
		if (error.message) {
			logger.error(error.cause ? "Error:" : "Message:");
			logger.error(error.message);
		}
		if (error.cause) {
			logger.error("\nMessage:");
			logger.error(error.cause);
		}
		if (error.suggestion) {
			logger.error("\nSuggestion:");
			logger.error(error.suggestion);
		}
		logger.break();
		process.exit(1);
	}
	if (error instanceof z.ZodError) {
		logger.error("Validation failed:");
		for (const [key, value] of Object.entries(error.flatten().fieldErrors)) logger.error(`- ${highlighter.info(key)}: ${value}`);
		logger.break();
		process.exit(1);
	}
	if (error instanceof Error) {
		logger.error(error.message);
		logger.break();
		process.exit(1);
	}
	logger.break();
	process.exit(1);
}

//#endregion
//#region src/utils/updaters/update-tailwind-config.ts
async function updateTailwindConfig(tailwindConfig, config, options) {
	if (!tailwindConfig) return;
	options = {
		silent: false,
		tailwindVersion: "v3",
		...options
	};
	if (options.tailwindVersion === "v4") return;
	const tailwindFileRelativePath = path.relative(config.resolvedPaths.cwd, config.resolvedPaths.tailwindConfig);
	const tailwindSpinner = spinner(`Updating ${highlighter.info(tailwindFileRelativePath)}`, { silent: options.silent }).start();
	const output = await transformTailwindConfig(await promises.readFile(config.resolvedPaths.tailwindConfig, "utf8"), tailwindConfig, config);
	await promises.writeFile(config.resolvedPaths.tailwindConfig, output, "utf8");
	tailwindSpinner?.succeed();
}
async function transformTailwindConfig(input, tailwindConfig, config) {
	const sourceFile = await _createSourceFile(input, config);
	const configObject = sourceFile.getDescendantsOfKind(SyntaxKind.ObjectLiteralExpression).find((node) => node.getProperties().some((property) => property.isKind(SyntaxKind.PropertyAssignment) && property.getName() === "content"));
	if (!configObject) return input;
	addTailwindConfigProperty(configObject, {
		name: "darkMode",
		value: "class"
	}, { quoteChar: _getQuoteChar(configObject) });
	tailwindConfig.plugins?.forEach((plugin) => {
		addTailwindConfigPlugin(configObject, plugin);
	});
	if (tailwindConfig.theme) await addTailwindConfigTheme(configObject, tailwindConfig.theme);
	return sourceFile.getFullText();
}
function addTailwindConfigProperty(configObject, property, { quoteChar }) {
	const existingProperty = configObject.getProperty("darkMode");
	if (!existingProperty) {
		const newProperty = {
			name: property.name,
			initializer: `[${quoteChar}${property.value}${quoteChar}]`
		};
		if (property.name === "darkMode") {
			configObject.insertPropertyAssignment(0, newProperty);
			return configObject;
		}
		configObject.addPropertyAssignment(newProperty);
		return configObject;
	}
	if (existingProperty.isKind(SyntaxKind.PropertyAssignment)) {
		const initializer = existingProperty.getInitializer();
		const newValue = `${quoteChar}${property.value}${quoteChar}`;
		if (initializer?.isKind(SyntaxKind.StringLiteral)) {
			const initializerText = initializer.getText();
			initializer.replaceWithText(`[${initializerText}, ${newValue}]`);
			return configObject;
		}
		if (initializer?.isKind(SyntaxKind.ArrayLiteralExpression)) {
			if (initializer.getElements().map((element) => element.getText()).includes(newValue)) return configObject;
			initializer.addElement(newValue);
		}
		return configObject;
	}
	return configObject;
}
async function addTailwindConfigTheme(configObject, theme) {
	if (!configObject.getProperty("theme")) configObject.addPropertyAssignment({
		name: "theme",
		initializer: "{}"
	});
	nestSpreadProperties(configObject);
	const themeInitializer = (configObject.getPropertyOrThrow("theme")?.asKindOrThrow(SyntaxKind.PropertyAssignment)).getInitializer();
	if (themeInitializer?.isKind(SyntaxKind.ObjectLiteralExpression)) {
		const resultString = objectToString(deepmerge(await parseObjectLiteral(themeInitializer.getText()), theme, { arrayMerge: (dst, src) => src })).replace(/'\.\.\.(.*)'/g, "...$1").replace(/'"/g, "'").replace(/"'/g, "'").replace(/'\[/g, "[").replace(/\]'/g, "]").replace(/'\\'/g, "'").replace(/\\'/g, "'").replace(/\\''/g, "'").replace(/''/g, "'");
		themeInitializer.replaceWithText(resultString);
	}
	unnestSpreadProperties(configObject);
}
function addTailwindConfigPlugin(configObject, plugin) {
	const existingPlugins = configObject.getProperty("plugins");
	if (!existingPlugins) {
		configObject.addPropertyAssignment({
			name: "plugins",
			initializer: `[${plugin}]`
		});
		return configObject;
	}
	if (existingPlugins.isKind(SyntaxKind.PropertyAssignment)) {
		const initializer = existingPlugins.getInitializer();
		if (initializer?.isKind(SyntaxKind.ArrayLiteralExpression)) {
			if (initializer.getElements().map((element) => {
				return element.getText().replace(/["']/g, "");
			}).includes(plugin.replace(/["']/g, ""))) return configObject;
			initializer.addElement(plugin);
		}
		return configObject;
	}
	return configObject;
}
async function _createSourceFile(input, config) {
	const dir = await promises.mkdtemp(path.join(tmpdir(), "shadcn-"));
	const resolvedPath = config?.resolvedPaths?.tailwindConfig || "tailwind.config.ts";
	const tempFile = path.join(dir, `shadcn-${path.basename(resolvedPath)}`);
	return new Project({ compilerOptions: {} }).createSourceFile(tempFile, input, { scriptKind: path.extname(resolvedPath) === ".ts" ? ScriptKind.TS : ScriptKind.JS });
}
function _getQuoteChar(configObject) {
	return configObject.getFirstDescendantByKind(SyntaxKind.StringLiteral)?.getQuoteKind() === QuoteKind.Single ? "'" : "\"";
}
function nestSpreadProperties(obj) {
	const properties = obj.getProperties();
	for (let i = 0; i < properties.length; i++) {
		const prop = properties[i];
		if (prop.isKind(SyntaxKind.SpreadAssignment)) {
			const spreadAssignment = prop.asKindOrThrow(SyntaxKind.SpreadAssignment);
			const spreadText = spreadAssignment.getExpression().getText();
			obj.insertPropertyAssignment(i, {
				name: `"___${spreadText.replace(/^\.\.\./, "")}"`,
				initializer: `"...${spreadText.replace(/^\.\.\./, "")}"`
			});
			spreadAssignment.remove();
		} else if (prop.isKind(SyntaxKind.PropertyAssignment)) {
			const initializer = prop.asKindOrThrow(SyntaxKind.PropertyAssignment).getInitializer();
			if (initializer && initializer.isKind(SyntaxKind.ObjectLiteralExpression)) nestSpreadProperties(initializer.asKindOrThrow(SyntaxKind.ObjectLiteralExpression));
			else if (initializer && initializer.isKind(SyntaxKind.ArrayLiteralExpression)) nestSpreadElements(initializer.asKindOrThrow(SyntaxKind.ArrayLiteralExpression));
		}
	}
}
function nestSpreadElements(arr) {
	const elements = arr.getElements();
	for (let j = 0; j < elements.length; j++) {
		const element = elements[j];
		if (element.isKind(SyntaxKind.ObjectLiteralExpression)) nestSpreadProperties(element.asKindOrThrow(SyntaxKind.ObjectLiteralExpression));
		else if (element.isKind(SyntaxKind.ArrayLiteralExpression)) nestSpreadElements(element.asKindOrThrow(SyntaxKind.ArrayLiteralExpression));
		else if (element.isKind(SyntaxKind.SpreadElement)) {
			const spreadText = element.getText();
			arr.removeElement(j);
			arr.insertElement(j, `"${spreadText}"`);
		}
	}
}
function unnestSpreadProperties(obj) {
	const properties = obj.getProperties();
	for (let i = 0; i < properties.length; i++) {
		const prop = properties[i];
		if (prop.isKind(SyntaxKind.PropertyAssignment)) {
			const propAssignment = prop;
			const initializer = propAssignment.getInitializer();
			if (initializer && initializer.isKind(SyntaxKind.StringLiteral)) {
				const value = initializer.asKindOrThrow(SyntaxKind.StringLiteral).getLiteralValue();
				if (value.startsWith("...")) {
					obj.insertSpreadAssignment(i, { expression: value.slice(3) });
					propAssignment.remove();
				}
			} else if (initializer?.isKind(SyntaxKind.ObjectLiteralExpression)) unnestSpreadProperties(initializer);
			else if (initializer && initializer.isKind(SyntaxKind.ArrayLiteralExpression)) unsetSpreadElements(initializer.asKindOrThrow(SyntaxKind.ArrayLiteralExpression));
		}
	}
}
function unsetSpreadElements(arr) {
	const elements = arr.getElements();
	for (let j = 0; j < elements.length; j++) {
		const element = elements[j];
		if (element.isKind(SyntaxKind.ObjectLiteralExpression)) unnestSpreadProperties(element.asKindOrThrow(SyntaxKind.ObjectLiteralExpression));
		else if (element.isKind(SyntaxKind.ArrayLiteralExpression)) unsetSpreadElements(element.asKindOrThrow(SyntaxKind.ArrayLiteralExpression));
		else if (element.isKind(SyntaxKind.StringLiteral)) {
			const spreadText = element.getText();
			const spreadTest = /^['"](\.\.\..*)['"]$/g;
			if (spreadTest.test(spreadText)) {
				arr.removeElement(j);
				arr.insertElement(j, spreadText.replace(spreadTest, "$1"));
			}
		}
	}
}
async function parseObjectLiteral(objectLiteralString) {
	const statement = (await _createSourceFile(`const theme = ${objectLiteralString}`, null)).getStatements()[0];
	if (statement?.getKind() === SyntaxKind.VariableStatement) {
		const initializer = (statement.getDeclarationList()?.getDeclarations()[0]).getInitializer();
		if (initializer?.isKind(SyntaxKind.ObjectLiteralExpression)) return await parseObjectLiteralExpression(initializer);
	}
	throw new Error("Invalid input: not an object literal");
}
function parseObjectLiteralExpression(node) {
	const result = {};
	for (const property of node.getProperties()) if (property.isKind(SyntaxKind.PropertyAssignment)) {
		const name = property.getName().replace(/'/g, "");
		if (property.getInitializer()?.isKind(SyntaxKind.ObjectLiteralExpression)) result[name] = parseObjectLiteralExpression(property.getInitializer());
		else if (property.getInitializer()?.isKind(SyntaxKind.ArrayLiteralExpression)) result[name] = parseArrayLiteralExpression(property.getInitializer());
		else result[name] = parseValue(property.getInitializer());
	}
	return result;
}
function parseArrayLiteralExpression(node) {
	const result = [];
	for (const element of node.getElements()) if (element.isKind(SyntaxKind.ObjectLiteralExpression)) result.push(parseObjectLiteralExpression(element.asKindOrThrow(SyntaxKind.ObjectLiteralExpression)));
	else if (element.isKind(SyntaxKind.ArrayLiteralExpression)) result.push(parseArrayLiteralExpression(element.asKindOrThrow(SyntaxKind.ArrayLiteralExpression)));
	else result.push(parseValue(element));
	return result;
}
function parseValue(node) {
	switch (node.getKind()) {
		case SyntaxKind.StringLiteral: return node.getText();
		case SyntaxKind.NumericLiteral: return Number(node.getText());
		case SyntaxKind.TrueKeyword: return true;
		case SyntaxKind.FalseKeyword: return false;
		case SyntaxKind.NullKeyword: return null;
		case SyntaxKind.ArrayLiteralExpression: return node.getElements().map(parseValue);
		case SyntaxKind.ObjectLiteralExpression: return parseObjectLiteralExpression(node);
		default: return node.getText();
	}
}
function buildTailwindThemeColorsFromCssVars(cssVars) {
	const result = {};
	for (const key of Object.keys(cssVars)) {
		const parts = key.split("-");
		const colorName = parts[0];
		const subType = parts.slice(1).join("-");
		if (subType === "") if (typeof result[colorName] === "object") result[colorName].DEFAULT = `hsl(var(--${key}))`;
		else result[colorName] = `hsl(var(--${key}))`;
		else {
			if (typeof result[colorName] !== "object") result[colorName] = { DEFAULT: `hsl(var(--${colorName}))` };
			result[colorName][subType] = `hsl(var(--${key}))`;
		}
	}
	for (const [colorName, value] of Object.entries(result)) if (typeof value === "object" && value.DEFAULT === `hsl(var(--${colorName}))` && !(colorName in cssVars)) delete value.DEFAULT;
	return result;
}

//#endregion
//#region src/registry/resolver.ts
function resolveRegistryItemsFromRegistries(items, config) {
	const registryHeaders = {};
	const resolvedItems = [...items];
	if (!config?.registries) {
		setRegistryHeaders({});
		return resolvedItems;
	}
	for (let i = 0; i < resolvedItems.length; i++) {
		const resolved = buildUrlAndHeadersForRegistryItem(resolvedItems[i], config);
		if (resolved) {
			resolvedItems[i] = resolved.url;
			if (Object.keys(resolved.headers).length > 0) registryHeaders[resolved.url] = resolved.headers;
		}
	}
	setRegistryHeaders(registryHeaders);
	return resolvedItems;
}
async function fetchRegistryItems(items, config, options = {}) {
	return await Promise.all(items.map(async (item) => {
		if (isLocalFile(item)) return fetchRegistryLocal(item);
		if (isUrl$1(item)) {
			const [result$1] = await fetchRegistry([item], options);
			try {
				return registryItemSchema.parse(result$1);
			} catch (error) {
				throw new RegistryParseError(item, error);
			}
		}
		if (item.startsWith("@") && config?.registries) {
			const [result$1] = await fetchRegistry(resolveRegistryItemsFromRegistries([item], config), options);
			try {
				return registryItemSchema.parse(result$1);
			} catch (error) {
				throw new RegistryParseError(item, error);
			}
		}
		const [result] = await fetchRegistry([`styles/${resolveRegistryStyle(config?.style)}/${item}.json`], options);
		try {
			return registryItemSchema.parse(result);
		} catch (error) {
			throw new RegistryParseError(item, error);
		}
	}));
}
z.intersection(registryItemSchema, z.object({ _source: z.string().optional() }));
async function resolveRegistryTree(names, config, options = {}) {
	try {
		options = {
			useCache: true,
			...options
		};
		let payload = [];
		const allDependencyItems = [];
		const allDependencyRegistryNames = [];
		const uniqueNames = Array.from(new Set(names));
		const results = await fetchRegistryItems(uniqueNames, config, options);
		const resultMap = /* @__PURE__ */ new Map();
		for (let i = 0; i < results.length; i++) if (results[i]) resultMap.set(uniqueNames[i], results[i]);
		for (const [sourceName, item] of Array.from(resultMap.entries())) {
			const itemWithSource = {
				...item,
				_source: sourceName
			};
			payload.push(itemWithSource);
			if (item.registryDependencies) {
				let resolvedDependencies = item.registryDependencies;
				if (!config?.registries) {
					const namespacedDeps = item.registryDependencies.filter((dep) => dep.startsWith("@"));
					if (namespacedDeps.length > 0) {
						const { registry } = parseRegistryAndItemFromString(namespacedDeps[0]);
						throw new RegistryNotConfiguredError(registry);
					}
				} else resolvedDependencies = resolveRegistryItemsFromRegistries(item.registryDependencies, config);
				const { items, registryNames } = await resolveDependenciesRecursively(resolvedDependencies, config, options, new Set(uniqueNames));
				allDependencyItems.push(...items);
				allDependencyRegistryNames.push(...registryNames);
			}
		}
		payload.push(...allDependencyItems);
		if (allDependencyRegistryNames.length > 0) {
			const uniqueRegistryNames = Array.from(new Set(allDependencyRegistryNames));
			const nonNamespacedItems = uniqueRegistryNames.filter((name) => !name.startsWith("@"));
			const namespacedDepItems = uniqueRegistryNames.filter((name) => name.startsWith("@"));
			if (namespacedDepItems.length > 0) {
				const depResults = await fetchRegistryItems(namespacedDepItems, config, options);
				for (let i = 0; i < depResults.length; i++) {
					const itemWithSource = {
						...depResults[i],
						_source: namespacedDepItems[i]
					};
					payload.push(itemWithSource);
				}
			}
			if (nonNamespacedItems.length > 0) {
				const index = await getShadcnRegistryIndex();
				if (!index && payload.length === 0) return null;
				if (index) {
					if (nonNamespacedItems.includes("index")) nonNamespacedItems.unshift("index");
					const registryUrls = [];
					for (const name of nonNamespacedItems) {
						const itemDependencies = await resolveRegistryDependencies(name, config, options);
						registryUrls.push(...itemDependencies);
					}
					const result = await fetchRegistry(Array.from(new Set(registryUrls)), options);
					const registryPayload = z.array(registryItemSchema).parse(result);
					payload.push(...registryPayload);
				}
			}
		}
		if (!payload.length) return null;
		if (uniqueNames.includes("index") || allDependencyRegistryNames.includes("index")) {
			if (config.tailwind.baseColor) {
				const theme = await registryGetTheme(config.tailwind.baseColor, config);
				if (theme) payload.unshift(theme);
			}
		}
		const sourceMap = /* @__PURE__ */ new Map();
		payload.forEach((item) => {
			const source = item._source || item.name;
			sourceMap.set(item, source);
		});
		payload = topologicalSortRegistryItems(payload, sourceMap);
		payload.sort((a, b) => {
			if (a.type === "registry:theme" && b.type !== "registry:theme") return -1;
			if (a.type !== "registry:theme" && b.type === "registry:theme") return 1;
			return 0;
		});
		let tailwind = {};
		payload.forEach((item) => {
			tailwind = deepmerge(tailwind, item.tailwind ?? {});
		});
		let cssVars = {};
		payload.forEach((item) => {
			cssVars = deepmerge(cssVars, item.cssVars ?? {});
		});
		let css = {};
		payload.forEach((item) => {
			css = deepmerge(css, item.css ?? {});
		});
		let docs = "";
		payload.forEach((item) => {
			if (item.docs) docs += `${item.docs}\n`;
		});
		let envVars = {};
		payload.forEach((item) => {
			envVars = deepmerge(envVars, item.envVars ?? {});
		});
		const deduplicatedFiles = await deduplicateFilesByTarget(payload.map((item) => item.files ?? []), config);
		const parsed = registryResolvedItemsTreeSchema.parse({
			dependencies: deepmerge.all(payload.map((item) => item.dependencies ?? [])),
			devDependencies: deepmerge.all(payload.map((item) => item.devDependencies ?? [])),
			files: deduplicatedFiles,
			tailwind,
			cssVars,
			css,
			docs
		});
		if (Object.keys(envVars).length > 0) parsed.envVars = envVars;
		return parsed;
	} catch (error) {
		handleError(error);
		return null;
	}
}
async function resolveDependenciesRecursively(dependencies, config, options = {}, visited = /* @__PURE__ */ new Set()) {
	const items = [];
	const registryNames = [];
	for (const dep of dependencies) {
		if (visited.has(dep)) continue;
		visited.add(dep);
		if (isUrl$1(dep) || isLocalFile(dep)) {
			const [item] = await fetchRegistryItems([dep], config, options);
			if (item) {
				items.push(item);
				if (item.registryDependencies) {
					const nested = await resolveDependenciesRecursively(config?.registries ? resolveRegistryItemsFromRegistries(item.registryDependencies, config) : item.registryDependencies, config, options, visited);
					items.push(...nested.items);
					registryNames.push(...nested.registryNames);
				}
			}
		} else if (dep.startsWith("@") && config?.registries) {
			const { registry } = parseRegistryAndItemFromString(dep);
			if (registry && !(registry in config.registries)) throw new RegistryNotConfiguredError(registry);
			const [item] = await fetchRegistryItems([dep], config, options);
			if (item) {
				items.push(item);
				if (item.registryDependencies) {
					const nested = await resolveDependenciesRecursively(config?.registries ? resolveRegistryItemsFromRegistries(item.registryDependencies, config) : item.registryDependencies, config, options, visited);
					items.push(...nested.items);
					registryNames.push(...nested.registryNames);
				}
			}
		} else {
			registryNames.push(dep);
			if (config) try {
				const [item] = await fetchRegistryItems([dep], config, options);
				if (item && item.registryDependencies) {
					const nested = await resolveDependenciesRecursively(config?.registries ? resolveRegistryItemsFromRegistries(item.registryDependencies, config) : item.registryDependencies, config, options, visited);
					items.push(...nested.items);
					registryNames.push(...nested.registryNames);
				}
			} catch (error) {}
		}
	}
	return {
		items,
		registryNames
	};
}
async function resolveRegistryDependencies(url, config, options = {}) {
	if (isUrl$1(url)) return [url];
	const { registryNames } = await resolveDependenciesRecursively([url], config, options, /* @__PURE__ */ new Set());
	const style = config.resolvedPaths?.cwd ? resolveRegistryStyle(await getTargetStyleFromConfig(config.resolvedPaths.cwd, config.style)) : resolveRegistryStyle(config.style);
	const urls = registryNames.map((name) => resolveRegistryUrl(isUrl$1(name) ? name : `styles/${style}/${name}.json`));
	return Array.from(new Set(urls));
}
async function registryGetTheme(name, config) {
	const [baseColor, tailwindVersion] = await Promise.all([getRegistryBaseColor(name), getProjectTailwindVersionFromConfig(config)]);
	if (!baseColor) return null;
	const theme = {
		name,
		type: "registry:theme",
		tailwind: { config: { theme: { extend: {
			borderRadius: {
				lg: "var(--radius)",
				md: "calc(var(--radius) - 2px)",
				sm: "calc(var(--radius) - 4px)"
			},
			colors: {}
		} } } },
		cssVars: {
			theme: {},
			light: { radius: "0.5rem" },
			dark: {}
		}
	};
	if (config.tailwind.cssVariables) {
		theme.tailwind.config.theme.extend.colors = {
			...theme.tailwind.config.theme.extend.colors,
			...buildTailwindThemeColorsFromCssVars(baseColor.cssVars.dark ?? {})
		};
		theme.cssVars = {
			theme: {
				...baseColor.cssVars.theme,
				...theme.cssVars.theme
			},
			light: {
				...baseColor.cssVars.light,
				...theme.cssVars.light
			},
			dark: {
				...baseColor.cssVars.dark,
				...theme.cssVars.dark
			}
		};
		if (tailwindVersion === "v4" && baseColor.cssVarsV4) theme.cssVars = {
			theme: {
				...baseColor.cssVarsV4.theme,
				...theme.cssVars.theme
			},
			light: {
				radius: "0.625rem",
				...baseColor.cssVarsV4.light
			},
			dark: { ...baseColor.cssVarsV4.dark }
		};
	}
	return theme;
}
function computeItemHash(item, source) {
	const identifier = source || item.name;
	const hash = createHash("sha256").update(identifier).digest("hex").substring(0, 8);
	return `${item.name}::${hash}`;
}
function extractItemIdentifierFromDependency(dependency) {
	if (isUrl$1(dependency)) {
		const pathname = new URL(dependency).pathname;
		const match = pathname.match(/\/([^/]+)\.json$/);
		const name = match ? match[1] : path.basename(pathname, ".json");
		return {
			name,
			hash: computeItemHash({ name }, dependency)
		};
	}
	if (isLocalFile(dependency)) {
		const match = dependency.match(/\/([^/]+)\.json$/);
		const name = match ? match[1] : path.basename(dependency, ".json");
		return {
			name,
			hash: computeItemHash({ name }, dependency)
		};
	}
	const { item } = parseRegistryAndItemFromString(dependency);
	return {
		name: item,
		hash: computeItemHash({ name: item }, dependency)
	};
}
function topologicalSortRegistryItems(items, sourceMap) {
	const itemMap = /* @__PURE__ */ new Map();
	const hashToItem = /* @__PURE__ */ new Map();
	const inDegree = /* @__PURE__ */ new Map();
	const adjacencyList = /* @__PURE__ */ new Map();
	items.forEach((item) => {
		const hash = computeItemHash(item, sourceMap.get(item) || item.name);
		itemMap.set(hash, item);
		hashToItem.set(hash, item);
		inDegree.set(hash, 0);
		adjacencyList.set(hash, []);
	});
	const depToHashes = /* @__PURE__ */ new Map();
	items.forEach((item) => {
		const source = sourceMap.get(item) || item.name;
		const hash = computeItemHash(item, source);
		if (!depToHashes.has(item.name)) depToHashes.set(item.name, []);
		depToHashes.get(item.name).push(hash);
		if (source !== item.name) {
			if (!depToHashes.has(source)) depToHashes.set(source, []);
			depToHashes.get(source).push(hash);
		}
	});
	items.forEach((item) => {
		const itemHash = computeItemHash(item, sourceMap.get(item) || item.name);
		if (item.registryDependencies) item.registryDependencies.forEach((dep) => {
			let depHash;
			const exactMatches = depToHashes.get(dep) || [];
			if (exactMatches.length === 1) depHash = exactMatches[0];
			else if (exactMatches.length > 1) depHash = exactMatches[0];
			else {
				const { name } = extractItemIdentifierFromDependency(dep);
				const nameMatches = depToHashes.get(name) || [];
				if (nameMatches.length > 0) depHash = nameMatches[0];
			}
			if (depHash && itemMap.has(depHash)) {
				adjacencyList.get(depHash).push(itemHash);
				inDegree.set(itemHash, inDegree.get(itemHash) + 1);
			}
		});
	});
	const queue = [];
	const sorted = [];
	inDegree.forEach((degree, hash) => {
		if (degree === 0) queue.push(hash);
	});
	while (queue.length > 0) {
		const currentHash = queue.shift();
		const item = itemMap.get(currentHash);
		sorted.push(item);
		adjacencyList.get(currentHash).forEach((dependentHash) => {
			const newDegree = inDegree.get(dependentHash) - 1;
			inDegree.set(dependentHash, newDegree);
			if (newDegree === 0) queue.push(dependentHash);
		});
	}
	if (sorted.length !== items.length) {
		console.warn("Circular dependency detected in registry items");
		const sortedHashes = new Set(sorted.map((item) => {
			return computeItemHash(item, sourceMap.get(item) || item.name);
		}));
		items.forEach((item) => {
			const hash = computeItemHash(item, sourceMap.get(item) || item.name);
			if (!sortedHashes.has(hash)) sorted.push(item);
		});
	}
	return sorted;
}

//#endregion
//#region src/registry/namespaces.ts
async function resolveRegistryNamespaces(components, config) {
	const discoveredNamespaces = /* @__PURE__ */ new Set();
	const visitedItems = /* @__PURE__ */ new Set();
	const itemsToProcess = [...components];
	while (itemsToProcess.length > 0) {
		const currentItem = itemsToProcess.shift();
		if (visitedItems.has(currentItem)) continue;
		visitedItems.add(currentItem);
		const { registry } = parseRegistryAndItemFromString(currentItem);
		if (registry && !BUILTIN_REGISTRIES[registry]) discoveredNamespaces.add(registry);
		try {
			const [item] = await fetchRegistryItems([currentItem], config, { useCache: true });
			if (item?.registryDependencies) for (const dep of item.registryDependencies) {
				const { registry: depRegistry } = parseRegistryAndItemFromString(dep);
				if (depRegistry && !BUILTIN_REGISTRIES[depRegistry]) discoveredNamespaces.add(depRegistry);
				if (!visitedItems.has(dep)) itemsToProcess.push(dep);
			}
		} catch (error) {
			if (error instanceof RegistryNotConfiguredError) {
				const { registry: registry$1 } = parseRegistryAndItemFromString(currentItem);
				if (registry$1 && !BUILTIN_REGISTRIES[registry$1]) discoveredNamespaces.add(registry$1);
				continue;
			}
			continue;
		}
	}
	return Array.from(discoveredNamespaces);
}

//#endregion
//#region src/utils/registries.ts
async function ensureRegistriesInConfig(components, config, options = {}) {
	options = {
		silent: false,
		writeFile: true,
		...options
	};
	const missingRegistries = (await resolveRegistryNamespaces(components, config)).filter((registry) => !config.registries?.[registry] && !Object.keys(BUILTIN_REGISTRIES).includes(registry));
	if (missingRegistries.length === 0) return {
		config,
		newRegistries: []
	};
	const registryIndex = await getRegistriesIndex({ useCache: process.env.NODE_ENV !== "development" });
	if (!registryIndex) return {
		config,
		newRegistries: []
	};
	const foundRegistries = {};
	for (const registry of missingRegistries) if (registryIndex[registry]) foundRegistries[registry] = registryIndex[registry];
	if (Object.keys(foundRegistries).length === 0) return {
		config,
		newRegistries: []
	};
	const existingRegistries = Object.fromEntries(Object.entries(config.registries || {}).filter(([key]) => !Object.keys(BUILTIN_REGISTRIES).includes(key)));
	const newConfigWithRegistries = {
		...config,
		registries: {
			...existingRegistries,
			...foundRegistries
		}
	};
	if (options.writeFile) {
		const { resolvedPaths, ...configWithoutResolvedPaths } = newConfigWithRegistries;
		const configSpinner = spinner("Updating components.json.", { silent: options.silent }).start();
		const updatedConfig = rawConfigSchema.parse(configWithoutResolvedPaths);
		await fsExtra.writeFile(path.resolve(config.resolvedPaths.cwd, "components.json"), `${JSON.stringify(updatedConfig, null, 2)}\n`, "utf-8");
		configSpinner.succeed();
	}
	return {
		config: newConfigWithRegistries,
		newRegistries: Object.keys(foundRegistries)
	};
}

//#endregion
//#region src/preset/presets.ts
const DEFAULT_PRESETS = {
	vega: {
		title: "Vega",
		description: "Lucide / Inter",
		base: "reka",
		style: "vega",
		baseColor: "neutral",
		theme: "neutral",
		iconLibrary: "lucide",
		font: "inter",
		fontHeading: "inherit",
		menuAccent: "subtle",
		menuColor: "default",
		radius: "default",
		rtl: false
	},
	nova: {
		title: "Nova",
		description: "Lucide / Geist",
		base: "reka",
		style: "nova",
		baseColor: "neutral",
		theme: "neutral",
		iconLibrary: "lucide",
		font: "geist-sans",
		fontHeading: "inherit",
		menuAccent: "subtle",
		menuColor: "default",
		radius: "default",
		rtl: false
	},
	maia: {
		title: "Maia",
		description: "Hugeicons / Figtree",
		base: "reka",
		style: "maia",
		baseColor: "neutral",
		theme: "neutral",
		iconLibrary: "hugeicons",
		font: "figtree",
		fontHeading: "inherit",
		menuAccent: "subtle",
		menuColor: "default",
		radius: "default",
		rtl: false
	},
	lyra: {
		title: "Lyra",
		description: "Phosphor / JetBrains Mono",
		base: "reka",
		style: "lyra",
		baseColor: "neutral",
		theme: "neutral",
		iconLibrary: "phosphor",
		font: "jetbrains-mono",
		fontHeading: "inherit",
		menuAccent: "subtle",
		menuColor: "default",
		radius: "default",
		rtl: false
	},
	mira: {
		title: "Mira",
		description: "Hugeicons / Inter",
		base: "reka",
		style: "mira",
		baseColor: "neutral",
		theme: "neutral",
		iconLibrary: "hugeicons",
		font: "inter",
		fontHeading: "inherit",
		menuAccent: "subtle",
		menuColor: "default",
		radius: "default",
		rtl: false
	},
	luma: {
		title: "Luma",
		description: "Lucide / Inter",
		base: "reka",
		style: "luma",
		baseColor: "neutral",
		theme: "neutral",
		iconLibrary: "lucide",
		font: "inter",
		fontHeading: "inherit",
		menuAccent: "subtle",
		menuColor: "default",
		radius: "default",
		rtl: false
	},
	sera: {
		title: "Sera",
		description: "Lucide / Noto Sans + Playfair Display",
		base: "reka",
		style: "sera",
		baseColor: "taupe",
		theme: "taupe",
		iconLibrary: "lucide",
		font: "noto-sans",
		fontHeading: "playfair-display",
		menuAccent: "subtle",
		menuColor: "default",
		radius: "default",
		rtl: false
	}
};
function resolveCreateUrl(searchParams) {
	const url = new URL(`${SHADCN_VUE_URL}/create`);
	const { rtl, pointer, ...params } = searchParams ?? {};
	for (const [key, value] of Object.entries(params)) if (value !== void 0) url.searchParams.set(key, String(value));
	if (rtl) url.searchParams.set("rtl", "true");
	if (pointer) url.searchParams.set("pointer", "true");
	return url.toString();
}
async function promptToOpenPresetBuilder(options) {
	logger.break();
	logger.log(`  Build your custom preset on ${highlighter.info(options.createUrl)}`);
	logger.log(`  ${options.followUp}`);
	logger.break();
	if (options.prompt === false) return;
	const { proceed } = await prompts({
		type: "confirm",
		name: "proceed",
		message: "Open in browser?",
		initial: true
	});
	if (proceed) await open(options.createUrl);
}
function resolveInitUrl(preset, options) {
	const params = new URLSearchParams({
		base: preset.base,
		style: preset.style,
		baseColor: preset.baseColor,
		theme: preset.theme,
		iconLibrary: preset.iconLibrary,
		font: preset.font,
		rtl: String(preset.rtl ?? false),
		menuAccent: preset.menuAccent,
		menuColor: preset.menuColor,
		radius: preset.radius
	});
	if (preset.fontHeading && preset.fontHeading !== "inherit") params.set("fontHeading", preset.fontHeading);
	if (options?.preset) params.set("preset", options.preset);
	if (options?.template) params.set("template", options.template);
	if (options?.pointer) params.set("pointer", "true");
	params.set("track", "1");
	return `${SHADCN_VUE_URL}/init?${params.toString()}`;
}
async function promptForBase() {
	return "reka";
}
async function promptForPreset(options) {
	const presets = Object.entries(DEFAULT_PRESETS);
	const { selectedPreset } = await prompts({
		type: "select",
		name: "selectedPreset",
		message: `Which ${highlighter.info("preset")} would you like to use?`,
		choices: [...presets.map(([name, preset$1]) => ({
			title: preset$1.title,
			description: preset$1.description,
			value: name
		})), {
			title: "Custom",
			description: `Build your own at ${highlighter.info(`${SHADCN_VUE_URL}/create`)}`,
			value: "custom"
		}]
	});
	if (!selectedPreset) return { kind: "cancelled" };
	if (selectedPreset === "custom") {
		await promptToOpenPresetBuilder({
			createUrl: resolveCreateUrl({
				command: "init",
				rtl: options.rtl,
				pointer: options.pointer,
				base: options.base,
				...options.template && { template: options.template }
			}),
			followUp: `Then ${highlighter.info("copy and run the command")} from shadcn-vue.com.`
		});
		return { kind: "custom" };
	}
	const preset = DEFAULT_PRESETS[selectedPreset];
	if (!preset) return { kind: "cancelled" };
	return {
		kind: "preset",
		url: resolveInitUrl({
			...preset,
			base: options.base,
			rtl: options.rtl
		}, {
			template: options.template,
			pointer: options.pointer
		}),
		base: options.base
	};
}
async function resolveRegistryBaseConfig(initUrl, cwd, options) {
	let shadowConfig = configWithDefaults(createConfig({
		resolvedPaths: { cwd },
		...options?.registries && { registries: options.registries }
	}));
	const { config: updatedConfig } = await ensureRegistriesInConfig([initUrl], shadowConfig, {
		silent: true,
		writeFile: false
	});
	shadowConfig = updatedConfig;
	buildUrlAndHeadersForRegistryItem(initUrl, shadowConfig);
	const [item] = await getRegistryItems([initUrl], {
		config: shadowConfig,
		useCache: true
	});
	if (item?.type !== "registry:base" || !item.config) throw new Error(`Expected a registry:base item with a config from ${initUrl}, got ${item?.type ?? "nothing"}.`);
	const registryBaseConfig = item.config;
	if (isUrl$1(initUrl)) {
		const params = new URL(initUrl).searchParams;
		const fontHeading = params.get("fontHeading");
		if (fontHeading && fontHeading !== "inherit" && fontHeading !== params.get("font") && !registryBaseConfig.fontHeading) registryBaseConfig.fontHeading = fontHeading;
	}
	let cleanUrl = initUrl;
	if (isShadcnVueInitUrl(initUrl)) {
		const url = new URL(initUrl);
		url.searchParams.delete("track");
		cleanUrl = url.toString();
	}
	return {
		registryBaseConfig,
		installStyleIndex: item?.extends !== "none",
		url: cleanUrl
	};
}
function isShadcnVueInitUrl(url) {
	if (!isUrl$1(url)) return false;
	const parsed = new URL(url);
	const trusted = new URL(SHADCN_VUE_URL);
	return parsed.origin === trusted.origin && parsed.pathname === "/init";
}

//#endregion
//#region src/registry/api.ts
async function getRegistry(name, options) {
	const { config, useCache } = options || {};
	if (isUrl$1(name)) {
		const [result$1] = await fetchRegistry([name], { useCache });
		try {
			return registrySchema.parse(result$1);
		} catch (error) {
			throw new RegistryParseError(name, error);
		}
	}
	if (!name.startsWith("@")) throw new RegistryInvalidNamespaceError(name);
	let registryName = name;
	if (!registryName.endsWith("/registry")) registryName = `${registryName}/registry`;
	const urlAndHeaders = buildUrlAndHeadersForRegistryItem(registryName, configWithDefaults(config));
	if (!urlAndHeaders?.url) throw new RegistryNotFoundError(registryName);
	if (urlAndHeaders.headers && Object.keys(urlAndHeaders.headers).length > 0) setRegistryHeaders({ [urlAndHeaders.url]: urlAndHeaders.headers });
	const [result] = await fetchRegistry([urlAndHeaders.url], { useCache });
	try {
		return registrySchema.parse(result);
	} catch (error) {
		throw new RegistryParseError(registryName, error);
	}
}
async function getRegistryItems(items, options) {
	const { config, useCache = false } = options || {};
	clearRegistryContext();
	return fetchRegistryItems(items, configWithDefaults(config), { useCache });
}
async function resolveRegistryItems(items, options) {
	const { config, useCache = false } = options || {};
	clearRegistryContext();
	return resolveRegistryTree(items, configWithDefaults(config), { useCache });
}
async function getRegistriesConfig(cwd) {
	const configResult = await getRawConfig(cwd);
	if (!configResult) return { registries: BUILTIN_REGISTRIES };
	const registriesConfig = z.object({ registries: registryConfigSchema.optional() }).safeParse(configResult);
	if (!registriesConfig.success) throw new ConfigParseError(cwd, registriesConfig.error);
	return { registries: {
		...BUILTIN_REGISTRIES,
		...registriesConfig.data.registries || {}
	} };
}
async function getShadcnRegistryIndex() {
	try {
		const [result] = await fetchRegistry(["index.json"]);
		return registryIndexSchema.parse(result);
	} catch (error) {
		logger.error("\n");
		handleError(error);
	}
}
async function getRegistryStyles() {
	try {
		const [result] = await fetchRegistry(["styles/index.json"]);
		return stylesSchema.parse(result);
	} catch (error) {
		logger.error("\n");
		handleError(error);
		return [];
	}
}
async function getRegistryIcons() {
	try {
		const [result] = await fetchRegistry(["icons/index.json"]);
		return iconsSchema.parse(result);
	} catch (error) {
		handleError(error);
		return {};
	}
}
async function getRegistryBaseColors() {
	return BASE_COLORS;
}
/**
* Get available component library bases (e.g., Reka UI).
*/
function getRegistryBases() {
	return BASES;
}
/**
* Get available visual styles.
*/
function getRegistryVisualStyles() {
	return STYLES;
}
/**
* Get available icon libraries.
*/
function getRegistryIconLibraries() {
	return ICON_LIBRARIES;
}
/**
* Get available fonts.
*/
function getRegistryFonts() {
	return FONTS;
}
/**
* Get a specific base by name.
*/
function getRegistryBase(name) {
	return BASES.find((base) => base.name === name);
}
/**
* Get a specific visual style by name.
*/
function getRegistryVisualStyle(name) {
	return STYLES.find((style) => style.name === name);
}
/**
* Get a specific icon library by name.
*/
function getRegistryIconLibrary(name) {
	return ICON_LIBRARIES.find((lib) => lib.name === name);
}
/**
* Get a specific font by name.
*/
function getRegistryFont(name) {
	return FONTS.find((font) => font.name === name);
}
/**
* Get available presets (predefined combinations of base, style, icons, and font).
*/
function getRegistryPresets() {
	return Object.entries(DEFAULT_PRESETS).map(([name, preset]) => ({
		name,
		...preset
	}));
}
/**
* Get a specific preset by name.
*/
function getRegistryPreset(name) {
	const preset = DEFAULT_PRESETS[name];
	if (!preset) return void 0;
	return {
		name,
		...preset
	};
}
async function getRegistryBaseColor(baseColor) {
	try {
		const [result] = await fetchRegistry([`colors/${baseColor}.json`]);
		return registryBaseColorSchema.parse(result);
	} catch (error) {
		if (error instanceof RegistryNotFoundError) return;
		handleError(error);
	}
}
/**
* @deprecated This function is deprecated and will be removed in a future version.
*/
async function resolveTree(index, names) {
	const tree = [];
	for (const name of names) {
		const entry = index.find((entry$1) => entry$1.name === name);
		if (!entry) continue;
		tree.push(entry);
		if (entry.registryDependencies) {
			const dependencies = await resolveTree(index, entry.registryDependencies);
			tree.push(...dependencies);
		}
	}
	return tree.filter((component, index$1, self) => self.findIndex((c) => c.name === component.name) === index$1);
}
/**
* @deprecated This function is deprecated and will be removed in a future version.
*/
async function fetchTree(style, tree) {
	try {
		return (await fetchRegistry(tree.map((item) => `styles/${style}/${item.name}.json`))).map((result) => registryItemSchema.parse(result));
	} catch (error) {
		handleError(error);
		return [];
	}
}
/**
* @deprecated This function is deprecated and will be removed in a future version.
*/
async function getItemTargetPath(config, item, override) {
	if (override) return override;
	if (item.type === "registry:ui") return config.resolvedPaths.ui ?? config.resolvedPaths.components;
	const [parent, type] = item.type?.split(":") ?? [];
	if (!(parent in config.resolvedPaths)) return null;
	return path.join(config.resolvedPaths[parent], type);
}
async function getRegistriesIndex(options) {
	options = {
		useCache: true,
		...options
	};
	const [data] = await fetchRegistry([`${REGISTRY_URL}/registries.json`], { useCache: options.useCache });
	try {
		return registriesIndexSchema.parse(data);
	} catch (error) {
		if (error instanceof z.ZodError) throw new RegistriesIndexParseError(error);
		throw error;
	}
}

//#endregion
//#region src/utils/frameworks.ts
const FRAMEWORKS = {
	vite: {
		name: "vite",
		label: "Vite",
		links: {
			installation: "https://shadcn-vue.com/docs/installation/vite",
			tailwind: "https://tailwindcss.com/docs/guides/vite"
		}
	},
	nuxt3: {
		name: "nuxt3",
		label: "Nuxt 3",
		links: {
			installation: "https://shadcn-vue.com/docs/installation/nuxt",
			tailwind: "https://tailwindcss.com/docs/guides/nuxtjs"
		}
	},
	nuxt4: {
		name: "nuxt4",
		label: "Nuxt 4",
		links: {
			installation: "https://shadcn-vue.com/docs/installation/nuxt",
			tailwind: "https://tailwindcss.com/docs/guides/nuxtjs"
		}
	},
	astro: {
		name: "astro",
		label: "Astro",
		links: {
			installation: "https://shadcn-vue.com/docs/installation/astro",
			tailwind: "https://tailwindcss.com/docs/guides/astro"
		}
	},
	laravel: {
		name: "laravel",
		label: "Laravel",
		links: {
			installation: "https://shadcn-vue.com/docs/installation/laravel",
			tailwind: "https://tailwindcss.com/docs/guides/laravel"
		}
	},
	manual: {
		name: "manual",
		label: "Manual",
		links: {
			installation: "https://shadcn-vue.com/docs/installation/manual",
			tailwind: "https://tailwindcss.com/docs/installation"
		}
	},
	inertia: {
		name: "inertia",
		label: "Inertia",
		links: {
			installation: "https://shadcn-vue.com/docs/installation/manual",
			tailwind: "https://tailwindcss.com/docs/installation"
		}
	}
};

//#endregion
//#region src/utils/get-package-info.ts
function getPackageInfo(cwd = "", shouldThrow = true) {
	const packageJsonPath = path.join(cwd, "package.json");
	return fsExtra.readJSONSync(packageJsonPath, { throws: shouldThrow });
}

//#endregion
//#region src/utils/updaters/update-dependencies.ts
/**
* Detect the package manager in use for `cwd`. Mirrors shadcn-ui's
* `getPackageManager` API.
*
* Detection order:
* 1. `nypm.detectPackageManager` (lockfile / `packageManager` field)
* 2. `node_modules/.pnpm` directory — strong signal pnpm was used even when
*    the lockfile is missing (e.g. interrupted prior install)
* 3. With `withFallback: true`, the parent process' `npm_config_user_agent`
*    (e.g. `pnpm dlx` exports `pnpm/<version> ...`)
* 4. Fallback to `'npm'`
*/
async function getPackageManager(cwd, { withFallback } = { withFallback: false }) {
	const detected = await detectPackageManager(cwd);
	if (detected?.name) return detected.name;
	if (fs.existsSync(path$1.join(cwd, "node_modules", ".pnpm"))) return "pnpm";
	if (!withFallback) return "npm";
	return getPackageManagerFromUserAgent() ?? "npm";
}
function getPackageManagerFromUserAgent(userAgent = process.env.npm_config_user_agent ?? "") {
	if (userAgent.startsWith("pnpm")) return "pnpm";
	if (userAgent.startsWith("yarn")) return "yarn";
	if (userAgent.startsWith("bun")) return "bun";
	if (userAgent.startsWith("deno")) return "deno";
	if (userAgent.startsWith("npm")) return "npm";
	return null;
}
async function updateDependencies(dependencies, devDependencies, config, options) {
	dependencies = Array.from(new Set(dependencies));
	devDependencies = Array.from(new Set(devDependencies));
	if (!dependencies?.length && !devDependencies?.length) return;
	options = {
		silent: false,
		...options
	};
	const dependenciesSpinner = spinner(`Installing dependencies.`, { silent: options.silent })?.start();
	await installWithPackageManager(await getPackageManager(config.resolvedPaths.cwd, { withFallback: true }), dependencies, devDependencies, config.resolvedPaths.cwd, options.silent);
	dependenciesSpinner?.succeed();
}
async function installWithPackageManager(packageManager, dependencies, devDependencies, cwd, silent) {
	const options = {
		cwd,
		packageManager,
		silent
	};
	if (dependencies?.length) await addDependency(dependencies, options);
	if (devDependencies?.length) await addDevDependency(devDependencies, options);
}

//#endregion
//#region src/utils/get-project-info.ts
const PROJECT_SHARED_IGNORE = [
	"**/node_modules/**",
	".nuxt",
	"public",
	"dist",
	"build"
];
const TS_CONFIG_SCHEMA = z.object({ compilerOptions: z.object({ paths: z.record(z.string().or(z.array(z.string()))) }) });
async function detectFrameworkConfigFiles(cwd) {
	const packageInfo = await getPackageInfo(cwd, false);
	const configFiles = await glob("**/{nuxt,vite,astro,wxt}.config.*|composer.json", {
		cwd,
		deep: 3,
		ignore: PROJECT_SHARED_IGNORE
	});
	if (configFiles.find((file) => file.startsWith("nuxt.config."))) {
		const nuxtPkg = packageInfo?.dependencies?.nuxt || packageInfo?.devDependencies?.nuxt;
		const nuxtVersion = nuxtPkg && coerce(nuxtPkg)?.version || "4.0.0";
		if (nuxtVersion.startsWith("4")) return FRAMEWORKS.nuxt4;
		else if (nuxtVersion.startsWith("3")) return FRAMEWORKS.nuxt3;
		return null;
	}
	if (configFiles.find((file) => file.startsWith("astro.config."))) return FRAMEWORKS.astro;
	if (configFiles.find((file) => file.startsWith("composer.json"))) return FRAMEWORKS.laravel;
	if (packageInfo?.dependencies?.["@inertiajs/vue3"] || packageInfo?.devDependencies?.["@inertiajs/vue3"] || await fsExtra.pathExists(path.join(cwd, "resources/js"))) return FRAMEWORKS.inertia;
	if (configFiles.find((file) => file.startsWith("wxt.config."))) return FRAMEWORKS.vite;
	if (configFiles.find((file) => file.startsWith("vite.config."))) return FRAMEWORKS.vite;
	return null;
}
async function isTypeScriptProject(cwd) {
	return (await glob("tsconfig.*", {
		cwd,
		deep: 1,
		ignore: PROJECT_SHARED_IGNORE
	})).length > 0;
}
async function getProjectInfo(cwd) {
	const [detectedFramework, typescript, isSrcDir, tailwindConfigFile, tailwindCssFile, tailwindVersion, aliasPrefix, packageJson, packageManager] = await Promise.all([
		detectFrameworkConfigFiles(cwd),
		isTypeScriptProject(cwd),
		fsExtra.pathExists(path.resolve(cwd, "src")),
		getTailwindConfigFile(cwd),
		getTailwindCssFile(cwd),
		getTailwindVersion(cwd),
		getTsConfigAliasPrefix(cwd),
		getPackageInfo(cwd, false),
		getPackageManager(cwd, { withFallback: true })
	]);
	return {
		framework: detectedFramework || FRAMEWORKS.manual,
		typescript,
		isSrcDir,
		tailwindConfigFile,
		tailwindCssFile,
		tailwindVersion,
		aliasPrefix,
		packageManager
	};
}
async function getTailwindVersion(cwd) {
	const [packageInfo, config] = await Promise.all([getPackageInfo(cwd, false), getConfig(cwd)]);
	if (config?.tailwind?.config === "") return "v4";
	const hasNuxtTailwind = !!(packageInfo?.dependencies?.["@nuxtjs/tailwindcss"] || packageInfo?.devDependencies?.["@nuxtjs/tailwindcss"]);
	if (!!!(packageInfo?.dependencies?.tailwindcss || packageInfo?.devDependencies?.tailwindcss) && !hasNuxtTailwind) return null;
	if (/^(?:\^|~)?3(?:\.\d+)*(?:-.*)?$/.test(packageInfo?.dependencies?.tailwindcss || packageInfo?.devDependencies?.tailwindcss || "")) return "v3";
	return "v4";
}
async function getTailwindCssFile(cwd) {
	const [files, tailwindVersion] = await Promise.all([glob(["**/*.css", "**/*.scss"], {
		cwd,
		deep: 5,
		ignore: PROJECT_SHARED_IGNORE
	}), getTailwindVersion(cwd)]);
	if (!files.length) return null;
	for (const file of files) {
		const contents = await fsExtra.readFile(path.resolve(cwd, file), "utf8");
		if (contents.includes(`@import "tailwindcss"`) || contents.includes(`@import 'tailwindcss'`) || contents.includes(`@tailwind base`)) return file;
	}
	return null;
}
async function getTailwindConfigFile(cwd) {
	const files = await glob("tailwind.config.*", {
		cwd,
		deep: 3,
		ignore: PROJECT_SHARED_IGNORE
	});
	if (!files.length) return null;
	return files[0];
}
async function getTsConfigAliasPrefix(cwd) {
	const detectedFramework = await detectFrameworkConfigFiles(cwd);
	const isTypeScript = await isTypeScriptProject(cwd);
	const tsConfig = await getTsconfig(cwd, detectedFramework?.name === "nuxt4" ? "./.nuxt/tsconfig.app.json" : detectedFramework?.name === "nuxt3" ? "./.nuxt/tsconfig.json" : detectedFramework?.name === "inertia" ? "./inertia/tsconfig.json" : isTypeScript ? "./tsconfig.json" : "./jsconfig.json");
	if (tsConfig === null || !Object.entries(tsConfig.config.compilerOptions?.paths ?? {}).length) return null;
	const aliasPaths = tsConfig.config.compilerOptions?.paths ?? {};
	for (const [alias, paths] of Object.entries(aliasPaths)) if (paths.includes("./*") || paths.includes("./src/*") || paths.includes("./app/*") || paths.includes("./resources/js/*")) {
		const cleanAlias = alias.replace(/\/\*$/, "") ?? null;
		return cleanAlias === "#build" ? "@" : cleanAlias;
	}
	return Object.keys(aliasPaths)?.[0]?.replace(/\/\*$/, "") ?? null;
}
async function getProjectConfig(cwd, defaultProjectInfo = null) {
	const [existingConfig, projectInfo] = await Promise.all([getConfig(cwd), !defaultProjectInfo ? getProjectInfo(cwd) : Promise.resolve(defaultProjectInfo)]);
	if (existingConfig) return existingConfig;
	if (!projectInfo || !projectInfo.tailwindCssFile || projectInfo.tailwindVersion === "v3" && !projectInfo.tailwindConfigFile) return null;
	return await resolveConfigPaths(cwd, {
		$schema: "https://shadcn-vue.com/schema.json",
		typescript: projectInfo.typescript,
		style: "new-york",
		tailwind: {
			config: projectInfo.tailwindConfigFile ?? "",
			baseColor: "zinc",
			css: projectInfo.tailwindCssFile,
			cssVariables: true,
			prefix: ""
		},
		iconLibrary: "lucide",
		aliases: {
			components: `${projectInfo.aliasPrefix}/components`,
			ui: `${projectInfo.aliasPrefix}/components/ui`,
			composables: `${projectInfo.aliasPrefix}/composables`,
			lib: `${projectInfo.aliasPrefix}/lib`,
			utils: `${projectInfo.aliasPrefix}/lib/utils`
		}
	});
}
async function getProjectTailwindVersionFromConfig(config) {
	if (!config.resolvedPaths?.cwd) return "v3";
	const projectInfo = await getProjectInfo(config.resolvedPaths.cwd);
	if (!projectInfo?.tailwindVersion) return null;
	return projectInfo.tailwindVersion;
}
/**
* Returns the list of installed UI component names for a project by scanning
* the resolved `ui` directory in `components.json` and cross-referencing
* discovered names against the shadcn-vue registry index.
*
* shadcn-vue stores components as folders (`ui/button/Button.vue`), so
* candidates come from immediate subdirectories of `ui/` that contain at
* least one renderable `.vue`/`.tsx`/`.jsx` file, plus any flat `.vue`
* single-file components. Names are then filtered against the registry to
* avoid picking up helper artefacts like `utils.ts` or `lib/` — mirrors
* shadcn-ui's `getProjectComponents`.
*/
async function getProjectComponents(cwd) {
	const existingConfig = await getConfig(cwd);
	if (!existingConfig) return [];
	const uiDir = existingConfig.resolvedPaths.ui;
	if (!uiDir) return [];
	let entries;
	try {
		entries = await fsExtra.readdir(uiDir, { withFileTypes: true });
	} catch {
		return [];
	}
	const candidates = /* @__PURE__ */ new Set();
	for (const entry of entries) {
		if (entry.name.startsWith(".") || entry.name === "index.ts" || entry.name === "index.js") continue;
		if (entry.isDirectory()) {
			const componentDir = path.join(uiDir, entry.name);
			let dirEntries;
			try {
				dirEntries = await fsExtra.readdir(componentDir, { withFileTypes: true });
			} catch {
				continue;
			}
			if (dirEntries.some((e) => e.isFile() && /\.(?:vue|tsx|jsx)$/.test(e.name))) candidates.add(entry.name);
			continue;
		}
		if (entry.isFile() && /\.(?:vue|tsx|jsx)$/.test(entry.name)) candidates.add(path.basename(entry.name, path.extname(entry.name)));
	}
	const registryIndex = await getShadcnRegistryIndex();
	const registryNames = new Set(registryIndex?.map((item) => item.name) ?? []);
	return Array.from(candidates).filter((name) => registryNames.has(name)).sort();
}

//#endregion
//#region src/registry/search.ts
async function searchRegistries(registries, options) {
	const { query, limit, offset, config, useCache } = options || {};
	let allItems = [];
	for (const registry of registries) {
		const itemsWithRegistry = ((await getRegistry(registry, {
			config,
			useCache
		})).items || []).map((item) => ({
			name: item.name,
			type: item.type,
			description: item.description,
			registry,
			addCommandArgument: buildRegistryItemNameFromRegistry(item.name, registry)
		}));
		allItems = allItems.concat(itemsWithRegistry);
	}
	if (query) allItems = searchItems(allItems, {
		query,
		limit: allItems.length,
		keys: ["name", "description"]
	});
	const paginationOffset = offset || 0;
	const paginationLimit = limit || allItems.length;
	const totalItems = allItems.length;
	const result = {
		pagination: {
			total: totalItems,
			offset: paginationOffset,
			limit: paginationLimit,
			hasMore: paginationOffset + paginationLimit < totalItems
		},
		items: allItems.slice(paginationOffset, paginationOffset + paginationLimit)
	};
	return searchResultsSchema.parse(result);
}
const searchableItemSchema = z.object({
	name: z.string(),
	type: z.string().optional(),
	description: z.string().optional(),
	registry: z.string().optional(),
	addCommandArgument: z.string().optional()
}).passthrough();
function searchItems(items, options) {
	options = {
		limit: 100,
		threshold: -1e4,
		...options
	};
	const results = fuzzysort.go(options.query, items, {
		keys: options.keys,
		threshold: options.threshold,
		limit: options.limit
	}).map((result) => result.obj);
	return z.array(searchableItemSchema).parse(results);
}
function isUrl(string) {
	try {
		new URL(string);
		return true;
	} catch {
		return false;
	}
}
function buildRegistryItemNameFromRegistry(name, registry) {
	if (!isUrl(registry)) return `${registry}/${name}`;
	const protocolEnd = registry.indexOf("://") + 3;
	const hostEnd = registry.indexOf("/", protocolEnd);
	if (hostEnd === -1) {
		const queryStart = registry.indexOf("?", protocolEnd);
		if (queryStart !== -1) return registry.substring(0, queryStart) + registry.substring(queryStart).replace(/\bregistry\b/g, name);
		return registry;
	}
	const hostPart = registry.substring(0, hostEnd);
	const pathAndQuery = registry.substring(hostEnd);
	const pathEnd = pathAndQuery.includes("?") ? pathAndQuery.indexOf("?") : pathAndQuery.length;
	const pathOnly = pathAndQuery.substring(0, pathEnd);
	const queryAndAfter = pathAndQuery.substring(pathEnd);
	const lastIndex = pathOnly.lastIndexOf("registry");
	let updatedPath = pathOnly;
	if (lastIndex !== -1) updatedPath = pathOnly.substring(0, lastIndex) + name + pathOnly.substring(lastIndex + 8);
	const updatedQuery = queryAndAfter.replace(/\bregistry\b/g, name);
	return hostPart + updatedPath + updatedQuery;
}

//#endregion
export { findExistingEnvFile as $, resolveRegistryItems as A, _getQuoteChar as B, getRegistryItems as C, RegistryParseError as Ct, getRegistryVisualStyle as D, getRegistryStyles as E, resolveInitUrl as F, clearRegistryContext as G, handleError as H, resolveRegistryBaseConfig as I, updateFiles as J, isUniversalRegistryItem as K, ensureRegistriesInConfig as L, DEFAULT_PRESETS as M, promptForBase as N, getRegistryVisualStyles as O, promptForPreset as P, logger as Q, resolveRegistryTree as R, getRegistryIcons as S, RegistryNotFoundError as St, getRegistryPresets as T, highlighter as Tt, buildUrlAndHeadersForRegistryItem as U, updateTailwindConfig as V, validateRegistryConfigForItems as W, transformRtl as X, transform$2 as Y, spinner as Z, getRegistryBases as _, RegistryForbiddenError as _t, getProjectTailwindVersionFromConfig as a, DEFAULT_TAILWIND_CONFIG as at, getRegistryIconLibraries as b, RegistryMissingEnvironmentVariablesError as bt, getPackageInfo as c, createConfig as ct, getRegistriesConfig as d, getConfig as dt, getNewEnvKeys as et, getRegistriesIndex as f, getWorkspaceConfig as ft, getRegistryBaseColors as g, RegistryFetchError as gt, getRegistryBaseColor as h, RegistryError as ht, getProjectInfo as i, DEFAULT_COMPONENTS as it, resolveTree as j, getShadcnRegistryIndex as k, fetchTree as l, findCommonRoot$1 as lt, getRegistryBase as m, RegistriesIndexParseError as mt, getProjectComponents as n, composeStyleId as nt, getPackageManager as o, DEFAULT_TAILWIND_CSS as ot, getRegistry as p, resolveConfigPaths as pt, isUrl$1 as q, getProjectConfig as r, configWithDefaults as rt, updateDependencies as s, DEFAULT_UTILS as st, searchRegistries as t, mergeEnvContent as tt, getItemTargetPath as u, findPackageRoot as ut, getRegistryFont as v, RegistryInvalidNamespaceError as vt, getRegistryPreset as w, RegistryUnauthorizedError as wt, getRegistryIconLibrary as x, RegistryNotConfiguredError as xt, getRegistryFonts as y, RegistryLocalFileError as yt, _createSourceFile as z };
