// shadcn-vue@2.7.4/dist/schema-CjAXtlQ7.js
import { z } from "zod";

//#region src/registry/schema.ts
const registryConfigItemSchema = z.union([z.string().refine((s) => s.includes("{name}"), { message: "Registry URL must include {name} placeholder" }), z.object({
	url: z.string().refine((s) => s.includes("{name}"), { message: "Registry URL must include {name} placeholder" }),
	params: z.record(z.string(), z.string()).optional(),
	headers: z.record(z.string(), z.string()).optional()
})]);
const registryConfigSchema = z.record(z.string().refine((key) => key.startsWith("@"), { message: "Registry names must start with @ (e.g., @v0, @acme)" }), registryConfigItemSchema);
const rawConfigSchema = z.object({
	$schema: z.string().optional(),
	style: z.string(),
	font: z.string().optional(),
	fontHeading: z.string().optional(),
	typescript: z.coerce.boolean().default(true),
	tailwind: z.object({
		config: z.string().optional(),
		css: z.string(),
		baseColor: z.string(),
		cssVariables: z.boolean().default(true),
		prefix: z.string().default("").optional()
	}),
	iconLibrary: z.string().optional(),
	rtl: z.boolean().default(false).optional(),
	pointer: z.boolean().default(false).optional(),
	menuColor: z.enum([
		"default",
		"inverted",
		"default-translucent",
		"inverted-translucent"
	]).default("default").optional(),
	menuAccent: z.enum(["subtle", "bold"]).default("subtle").optional(),
	aliases: z.object({
		components: z.string(),
		utils: z.string(),
		ui: z.string().optional(),
		lib: z.string().optional(),
		hooks: z.string().optional(),
		composables: z.string().optional()
	}),
	registries: registryConfigSchema.optional()
}).strict();
const configSchema = rawConfigSchema.extend({ resolvedPaths: z.object({
	cwd: z.string(),
	tailwindConfig: z.string(),
	tailwindCss: z.string(),
	utils: z.string(),
	components: z.string(),
	lib: z.string(),
	hooks: z.string(),
	ui: z.string(),
	composables: z.string()
}) });
const workspaceConfigSchema = z.record(configSchema);
const registryItemTypeSchema = z.enum([
	"registry:lib",
	"registry:block",
	"registry:component",
	"registry:ui",
	"registry:hook",
	"registry:composable",
	"registry:page",
	"registry:file",
	"registry:theme",
	"registry:style",
	"registry:item",
	"registry:base",
	"registry:font",
	"registry:example",
	"registry:internal"
]);
const registryItemFileSchema = z.discriminatedUnion("type", [z.object({
	path: z.string(),
	content: z.string().optional(),
	type: z.enum(["registry:file", "registry:page"]),
	target: z.string()
}), z.object({
	path: z.string(),
	content: z.string().optional(),
	type: registryItemTypeSchema.exclude(["registry:file", "registry:page"]),
	target: z.string().optional()
})]);
const registryItemTailwindSchema = z.object({ config: z.object({
	content: z.array(z.string()).optional(),
	theme: z.record(z.string(), z.any()).optional(),
	plugins: z.array(z.string()).optional()
}).optional() });
const registryItemCssVarsSchema = z.object({
	theme: z.record(z.string(), z.string()).optional(),
	light: z.record(z.string(), z.string()).optional(),
	dark: z.record(z.string(), z.string()).optional()
});
const cssValueSchema = z.lazy(() => z.union([
	z.string(),
	z.array(z.union([z.string(), z.record(z.string(), z.string())])),
	z.record(z.string(), cssValueSchema)
]));
const registryItemCssSchema = z.record(z.string(), cssValueSchema);
const registryItemEnvVarsSchema = z.record(z.string(), z.string());
const registryItemFontSchema = z.object({
	family: z.string(),
	provider: z.literal("google"),
	import: z.string(),
	variable: z.string(),
	weight: z.array(z.string()).optional(),
	subsets: z.array(z.string()).optional()
});
const registryItemCommonSchema = z.object({
	$schema: z.string().optional(),
	extends: z.string().optional(),
	name: z.string(),
	title: z.string().optional(),
	author: z.string().min(2).optional(),
	description: z.string().optional(),
	dependencies: z.array(z.string()).optional(),
	devDependencies: z.array(z.string()).optional(),
	registryDependencies: z.array(z.string()).optional(),
	files: z.array(registryItemFileSchema).optional(),
	tailwind: registryItemTailwindSchema.optional(),
	cssVars: registryItemCssVarsSchema.optional(),
	css: registryItemCssSchema.optional(),
	envVars: registryItemEnvVarsSchema.optional(),
	meta: z.record(z.string(), z.any()).optional(),
	docs: z.string().optional(),
	categories: z.array(z.string()).optional()
});
const registryItemSchema = z.discriminatedUnion("type", [
	registryItemCommonSchema.extend({
		type: z.literal("registry:base"),
		config: rawConfigSchema.deepPartial().optional()
	}),
	registryItemCommonSchema.extend({
		type: z.literal("registry:font"),
		font: registryItemFontSchema
	}),
	registryItemCommonSchema.extend({ type: registryItemTypeSchema.exclude(["registry:base", "registry:font"]) })
]);
const registrySchema = z.object({
	name: z.string(),
	homepage: z.string(),
	items: z.array(registryItemSchema)
});
const registryIndexSchema = z.array(registryItemSchema);
const stylesSchema = z.array(z.object({
	name: z.string(),
	label: z.string()
}));
const iconsSchema = z.record(z.string(), z.record(z.string(), z.string()));
const registryBaseColorSchema = z.object({
	inlineColors: z.object({
		light: z.record(z.string(), z.string()),
		dark: z.record(z.string(), z.string())
	}),
	cssVars: registryItemCssVarsSchema,
	cssVarsV4: registryItemCssVarsSchema.optional(),
	inlineColorsTemplate: z.string(),
	cssVarsTemplate: z.string()
});
const registryResolvedItemsTreeSchema = registryItemCommonSchema.pick({
	dependencies: true,
	devDependencies: true,
	files: true,
	tailwind: true,
	cssVars: true,
	css: true,
	envVars: true,
	docs: true
}).extend({ fonts: z.array(registryItemCommonSchema.extend({
	type: z.literal("registry:font"),
	font: registryItemFontSchema
})).optional() });
const searchResultItemSchema = z.object({
	name: z.string(),
	type: z.string().optional(),
	description: z.string().optional(),
	registry: z.string(),
	addCommandArgument: z.string()
});
const searchResultsSchema = z.object({
	pagination: z.object({
		total: z.number(),
		offset: z.number(),
		limit: z.number(),
		hasMore: z.boolean()
	}),
	items: z.array(searchResultItemSchema)
});
const registriesIndexSchema = z.record(z.string().regex(/^@[a-z0-9][\w-]*$/i), z.string());
const presetSchema = z.object({
	name: z.string(),
	title: z.string(),
	description: z.string(),
	base: z.string(),
	style: z.string(),
	baseColor: z.string(),
	theme: z.string(),
	iconLibrary: z.string(),
	font: z.string(),
	menuAccent: z.enum(["subtle", "bold"]),
	menuColor: z.enum([
		"default",
		"inverted",
		"default-translucent",
		"inverted-translucent"
	]),
	radius: z.string()
});
const configJsonSchema = z.object({ presets: z.array(presetSchema) });

//#endregion
export { searchResultsSchema as C, searchResultItemSchema as S, workspaceConfigSchema as T, registryItemSchema as _, rawConfigSchema as a, registryResolvedItemsTreeSchema as b, registryConfigItemSchema as c, registryItemCommonSchema as d, registryItemCssSchema as f, registryItemFontSchema as g, registryItemFileSchema as h, presetSchema as i, registryConfigSchema as l, registryItemEnvVarsSchema as m, configSchema as n, registriesIndexSchema as o, registryItemCssVarsSchema as p, iconsSchema as r, registryBaseColorSchema as s, configJsonSchema as t, registryIndexSchema as u, registryItemTailwindSchema as v, stylesSchema as w, registrySchema as x, registryItemTypeSchema as y };
