// reka-ui@2.10.1/dist/Drawer/DrawerPortal.cjs
const require_rolldown_runtime = require('../rolldown-runtime.cjs');
const require_Teleport_Teleport = require('../Teleport/Teleport.cjs');
const vue = require_rolldown_runtime.__toESM(require("vue"));

//#region src/Drawer/DrawerPortal.vue?vue&type=script&setup=true&lang.ts
var DrawerPortal_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ (0, vue.defineComponent)({
	__name: "DrawerPortal",
	props: {
		to: {
			type: null,
			required: false
		},
		disabled: {
			type: Boolean,
			required: false
		},
		defer: {
			type: Boolean,
			required: false
		},
		forceMount: {
			type: Boolean,
			required: false
		}
	},
	setup(__props) {
		return (_ctx, _cache) => {
			return (0, vue.openBlock)(), (0, vue.createBlock)((0, vue.unref)(require_Teleport_Teleport.Teleport_default), (0, vue.normalizeProps)((0, vue.guardReactiveProps)(_ctx.$props)), {
				default: (0, vue.withCtx)(() => [(0, vue.renderSlot)(_ctx.$slots, "default")]),
				_: 3
			}, 16);
		};
	}
});

//#endregion
//#region src/Drawer/DrawerPortal.vue
var DrawerPortal_default = DrawerPortal_vue_vue_type_script_setup_true_lang_default;

//#endregion
Object.defineProperty(exports, 'DrawerPortal_default', {
  enumerable: true,
  get: function () {
    return DrawerPortal_default;
  }
});
