// @tanstack/vue-query@5.101.2/build/legacy/index.js
// src/index.ts
export * from "@tanstack/query-core";
import { useQueryClient } from "./useQueryClient.js";
import { VueQueryPlugin } from "./vueQueryPlugin.js";
import { QueryClient } from "./queryClient.js";
import { QueryCache } from "./queryCache.js";
import { queryOptions } from "./queryOptions.js";
import { infiniteQueryOptions } from "./infiniteQueryOptions.js";
import { MutationCache } from "./mutationCache.js";
import { mutationOptions } from "./mutationOptions.js";
import { useQuery } from "./useQuery.js";
import { useQueries } from "./useQueries.js";
import { useInfiniteQuery } from "./useInfiniteQuery.js";
import { usePrefetchQuery } from "./usePrefetchQuery.js";
import { usePrefetchInfiniteQuery } from "./usePrefetchInfiniteQuery.js";
import { useMutation } from "./useMutation.js";
import { useIsFetching } from "./useIsFetching.js";
import { useIsMutating, useMutationState } from "./useMutationState.js";
import { VUE_QUERY_CLIENT } from "./utils.js";
export {
  MutationCache,
  QueryCache,
  QueryClient,
  VUE_QUERY_CLIENT,
  VueQueryPlugin,
  infiniteQueryOptions,
  mutationOptions,
  queryOptions,
  useInfiniteQuery,
  useIsFetching,
  useIsMutating,
  useMutation,
  useMutationState,
  usePrefetchInfiniteQuery,
  usePrefetchQuery,
  useQueries,
  useQuery,
  useQueryClient
};
